{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "edf9b481-8650-4505-b1c8-23a14de78ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from tqdm import tqdm as tqdm\n",
    "import pandas as pd\n",
    "import bz2\n",
    "import subprocess\n",
    "import pickle as pkl\n",
    "from itertools import chain\n",
    "from glob import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "faaa1eff-e7bf-4f10-a5e8-fe67425f1681",
   "metadata": {},
   "outputs": [],
   "source": [
    "working_path = '/cluster/raid/home/f80878961/beestrong/'\n",
    "count_path = '{}data/allele_freqs/sync_files/'.format(working_path)\n",
    "queen_geno_path = '{}data/queen_genotypes/All/'.format(working_path)\n",
    "\n",
    "tlf_path = '{}tlf/'.format(working_path)\n",
    "pheno_df = pd.read_csv('{}BeeStrong_metadata.csv'.format(tlf_path))\n",
    "\n",
    "script_path = \"/cluster/raid/home/f80878961/scripts/tmp_scripts/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "faf43f06-0ec3-4e12-969b-17255a018f84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bs_ids = pheno_df['Colony'].to_list()\n",
    "# snp_nr = 7023976 # will need sanity check here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d409b7-31bd-4f5b-9dae-70b34ec06c83",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "42d775bb-783a-40c8-a80f-89e5ef376fe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1442\n",
      "1513\n"
     ]
    }
   ],
   "source": [
    "snp_nr = 7023976 # wc -l BS17_0071.count -1\n",
    "\n",
    "## All\n",
    "filtered_count_path = '{}data/allele_freqs/All/'.format(working_path)\n",
    "\n",
    "bs_ids = sorted([x.split('/')[-1].split('.')[0] for x in glob('{}*.count.bz2'.format(count_path))])\n",
    "\n",
    "chunk_size = 101 # 15 chunks\n",
    "batch_file = '{}bsid_batch.txt'.format(filtered_count_path)\n",
    "\n",
    "# bs_ids = pheno_df['Colony'].to_list()\n",
    "print(len(set(bs_ids).intersection(pheno_df['Colony'].to_list())))\n",
    "print(len(bs_ids))\n",
    "\n",
    "# for SNP filtering\n",
    "min_prev = 0.1 # prevalence threshold for common alleles (if >2 common allele, then real tri-allelic)\n",
    "min_depth = 10 # both mean and median depth \n",
    "max_depth = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b06aa8-69bb-4b83-951c-7cb63e00c9e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c71d2c00-42b6-47cc-9cb4-10c85c716518",
   "metadata": {},
   "source": [
    "## Merge allele frequencies .sync files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42e45547-d8bd-40a2-bbe7-7249db379db9",
   "metadata": {},
   "source": [
    "### Write per-allele count matrices (SNP x colony for A, T, C, G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "15b9cec1-d70e-426d-85f4-018666e4fbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_merge_sync_script(script_fn, array_str):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --array=ARRAY_STRING\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=20g\n",
    "#SBATCH --time=4:00:00\n",
    "#SBATCH --job-name=merge_sync\n",
    "#SBATCH --output=%x_%A_%a.out\n",
    "#SBATCH --error=%x_%A_%a.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "# file listing batches of BeeStrong ids (bs_id) for parallel computing\n",
    "batch_file=$1\n",
    "snp_nr=$2\n",
    "count_path=$3\n",
    "out_path=$4\n",
    "\n",
    "batch_id=$SLURM_ARRAY_TASK_ID\n",
    "\n",
    "# getting bs_ids\n",
    "bs_ids=$(sed -n ${batch_id}'{p;q}' ${batch_file})\n",
    "\n",
    "python ~/scripts/plan-b-omics/bin/merge_sync.py --batch_id ${batch_id} --snp_nr ${snp_nr} --bs_ids ${bs_ids} --count_path ${count_path} --out_path ${out_path}\n",
    "\n",
    "echo DONE\"\"\".replace(\"ARRAY_STRING\", array_str)\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c51fc6da-d8c3-4a76-aad4-69131f6c73f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create batch file for first time\n",
    "with open(batch_file, \"w\") as outf:\n",
    "    for i in range(0, len(bs_ids), chunk_size):\n",
    "        chunk = bs_ids[i:i+chunk_size]\n",
    "        outf.write(\",\".join(chunk) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "36329bf3-41ed-4dbc-9ced-c443339c9499",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_str = '1-15'\n",
    "script_fn = '{}merge_sync.run'.format(script_path)\n",
    "write_merge_sync_script(script_fn, array_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1f6b1d44-ed80-4424-8d0b-f25574d5129e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 1388883\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$script_fn\" \"$batch_file\" \"$snp_nr\" \"$count_path\" \"$filtered_count_path\"\n",
    "cd $5\n",
    "sbatch $1 $2 $3 $4 $5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "393d9d30-7544-4338-8ca3-c438f19c60b9",
   "metadata": {},
   "source": [
    "### Merge into 3-D matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "db125385-1ec0-4c05-bb94-585eafb3bb3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "assert sum([x == bs_ids[i] for i, x in enumerate(list(chain(*batch_bs_ids)))]) == len(bs_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "308c7820-70f6-4dbf-a3f0-7b97e20f1919",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:27, 79911.85it/s]\n",
      "7023976it [01:12, 96871.43it/s]\n",
      "7023976it [01:13, 95607.20it/s]\n",
      "7023976it [01:12, 96395.80it/s]\n",
      "7023976it [01:13, 95634.25it/s]\n",
      "7023976it [01:13, 95841.95it/s]\n",
      "7023976it [01:12, 96750.18it/s]\n",
      "7023976it [01:13, 96216.97it/s]\n",
      "7023976it [01:13, 96170.55it/s]\n",
      "7023976it [01:13, 94955.64it/s]\n",
      "7023976it [01:10, 99338.00it/s] \n",
      "7023976it [01:10, 100071.38it/s]\n",
      "7023976it [01:10, 99098.12it/s] \n",
      "7023976it [01:20, 86892.53it/s]\n",
      "7023976it [01:19, 88464.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:20, 87638.04it/s]\n",
      "7023976it [01:19, 88887.25it/s]\n",
      "7023976it [01:20, 87229.69it/s]\n",
      "7023976it [01:21, 86617.28it/s]\n",
      "7023976it [01:21, 86346.12it/s]\n",
      "7023976it [01:20, 87418.80it/s]\n",
      "7023976it [01:20, 87219.67it/s]\n",
      "7023976it [01:22, 85541.27it/s]\n",
      "7023976it [01:22, 85528.61it/s]\n",
      "7023976it [01:22, 85412.89it/s]\n",
      "7023976it [01:21, 86602.88it/s]\n",
      "7023976it [01:21, 86004.87it/s]\n",
      "7023976it [01:20, 86980.82it/s]\n",
      "7023976it [01:21, 86634.61it/s]\n",
      "7023976it [01:19, 87840.42it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:22, 85362.98it/s]\n",
      "7023976it [01:21, 86695.99it/s]\n",
      "7023976it [01:21, 86272.75it/s]\n",
      "7023976it [01:21, 85767.50it/s]\n",
      "7023976it [01:23, 83933.48it/s]\n",
      "7023976it [01:21, 85708.65it/s]\n",
      "7023976it [01:22, 85253.64it/s]\n",
      "7023976it [01:22, 84884.01it/s]\n",
      "7023976it [01:23, 84503.23it/s]\n",
      "7023976it [01:23, 84575.33it/s]\n",
      "7023976it [01:23, 84403.22it/s]\n",
      "7023976it [01:22, 85003.84it/s]\n",
      "7023976it [01:22, 84759.25it/s]\n",
      "7023976it [01:23, 84321.06it/s]\n",
      "7023976it [01:21, 86137.15it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:21, 86597.14it/s]\n",
      "7023976it [01:21, 86523.52it/s]\n",
      "7023976it [01:23, 84421.79it/s]\n",
      "7023976it [01:21, 85817.88it/s]\n",
      "7023976it [01:22, 85413.62it/s]\n",
      "7023976it [01:22, 85604.58it/s]\n",
      "7023976it [01:22, 85174.04it/s]\n",
      "7023976it [01:23, 84585.02it/s]\n",
      "7023976it [01:24, 83101.66it/s]\n",
      "7023976it [01:22, 84669.79it/s]\n",
      "7023976it [01:22, 84762.85it/s]\n",
      "7023976it [01:22, 85111.30it/s]\n",
      "7023976it [01:21, 86124.85it/s]\n",
      "7023976it [01:22, 85098.11it/s]\n",
      "7023976it [01:20, 86877.37it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1h 9min 1s, sys: 10min 24s, total: 1h 19min 25s\n",
      "Wall time: 1h 20min 4s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "allele2jj = {'A': 0, 'T': 1, 'C':2, 'G':3}\n",
    "count_mat = np.zeros((snp_nr, len(bs_ids), len(allele2jj)), dtype=np.uint16)\n",
    "\n",
    "for allele, jj in allele2jj.items():\n",
    "    print(allele, jj)\n",
    "    \n",
    "    for j, b_bs_ids in enumerate(batch_bs_ids):\n",
    "        batch_id = j + 1\n",
    "        batch_size = len(b_bs_ids)\n",
    "        with open('{}count_{}_{}.txt'.format(filtered_count_path, allele, batch_id), 'r') as inf:\n",
    "            i = 0\n",
    "            for l in tqdm(inf):                    \n",
    "                count_mat[i, j * chunk_size : j * chunk_size + batch_size, jj] = np.array(l.split()[2:], dtype=np.int16)\n",
    "                i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f3ff2627-20b8-4001-98c8-1ab5aa5f12b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 909 ms, sys: 2min 15s, total: 2min 16s\n",
      "Wall time: 2min 33s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "with open('{}count_mat.pkl'.format(filtered_count_path), 'wb') as outf:\n",
    "    pkl.dump(count_mat, outf, protocol=pkl.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5170bfda-6e58-4b04-a29a-5b7204eee9a3",
   "metadata": {},
   "source": [
    "## Filter SNPs and recode alleles"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2dfe5850-b452-4d42-8541-14e93c667c9b",
   "metadata": {},
   "source": [
    "filtering \n",
    "1. monoallelic filtering : 3/4 tables empty\n",
    "2. mitochondrial SNP: contig NC_001566.1\n",
    "3. tri-allelic SNP filtering: >2 allele present in min. 10% colonies, else delete variants from third and fourth alleles\n",
    "4. min-max depth (10/50, on mean AND median) (after tri-allelic variants deletions, also not including N and deletions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6ad2f606-4a24-4e4a-816d-97f89aad017b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85.018205504\n",
      "CPU times: user 72.6 ms, sys: 2min 34s, total: 2min 34s\n",
      "Wall time: 3min 12s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "with open('{}count_mat.pkl'.format(filtered_count_path), 'rb') as inf:\n",
    "    count_mat = pkl.load(inf)\n",
    "print(count_mat.nbytes/1000000000)\n",
    "\n",
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "assert sum([x == bs_ids[i] for i, x in enumerate(list(chain(*batch_bs_ids)))]) == len(bs_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "cb149492-4685-4689-997d-cb55aa1240ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_i = 0\n",
    "batch_ii = len(bs_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "cddd8f11-a4e8-42e7-8ae0-e348ae8ae3e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 54s, sys: 13.8 s, total: 3min 7s\n",
      "Wall time: 3min 8s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[   1, 1513,    3,    9],\n",
       "       [   4, 1513,   27,    2],\n",
       "       [1513,   11,    2,   22],\n",
       "       ...,\n",
       "       [1437,  849,  118,  337],\n",
       "       [ 397, 1364, 1058,  218],\n",
       "       [ 335, 1504,  310,  170]])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# counts the number of A, T, C, G allele per SNP # should take 7min\n",
    "allele_prev = np.sum(count_mat[:, batch_i:batch_ii, :] > 0, axis=1)\n",
    "allele_prev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "183ab84d-5cef-473d-bf56-57e8b6053d89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3261\n",
      "6963287\n",
      "17207\n",
      "CPU times: user 408 ms, sys: 63.9 ms, total: 472 ms\n",
      "Wall time: 476 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# if count 3/4 alleles == 0, then monoallelic\n",
    "mono_mask = np.sum(allele_prev > 0, axis=1) == 1\n",
    "print(np.sum(mono_mask))\n",
    "\n",
    "# tri-allelic stringent\n",
    "tri_stringent_mask = np.sum(allele_prev > 0, axis=1) > 2\n",
    "print(np.sum(tri_stringent_mask))\n",
    "\n",
    "# tri-allelic relaxed (filter SNP only when >2 common alleles, i.e., an allele present in >= 10% of colonies)\n",
    "tri_mask = np.sum(allele_prev >= ((batch_ii - batch_i) * min_prev), axis=1) > 2\n",
    "print(np.sum(tri_mask))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "14024d33-6f34-4d7c-9476-c7865ffa6c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # SNP with ambiguous alternative alleles\n",
    "# i = 134\n",
    "# i = 157\n",
    "# i = 159"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "95d9a0c3-b7d7-42e9-b38f-63eb48c1b8ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6955445/6955445 [01:08<00:00, 101246.59it/s]\n"
     ]
    }
   ],
   "source": [
    "# iterate through stringent tri-allelic SNPs: filter if ambigous else recode (delete 3rd and fourth)\n",
    "ambi_mask = np.full(snp_nr, False)\n",
    "for i in tqdm(np.argwhere(tri_stringent_mask).flatten()):\n",
    "\n",
    "    # get number of alleles and sort by prevalence \n",
    "    atcg_prev = allele_prev[i, :]\n",
    "    alleles_sorted_by_prevalence = np.argsort(atcg_prev)\n",
    "\n",
    "    # if 2nd and 3rd alleles are equally common --> ambiguous --> filter\n",
    "    if atcg_prev[alleles_sorted_by_prevalence[1]] == atcg_prev[alleles_sorted_by_prevalence[2]]:\n",
    "        ambi_mask[i] = True\n",
    "\n",
    "    # else delete 3rd and 4rth alleles\n",
    "    else:\n",
    "        count_mat[i, batch_i:batch_ii, alleles_sorted_by_prevalence[:2]] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0ca9fd4-4447-4f42-9b51-ee96bfed6601",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "126577"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(ambi_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3aaa7293-7cdd-4005-929e-d0392c9b25da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 41s, sys: 4min 21s, total: 7min 2s\n",
      "Wall time: 7min 5s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# depth filtering --> if to, long I could consider iterating ...to have an overview of the time it takes # 8 min\n",
    "depth  = np.sum(count_mat[:, batch_i:batch_ii, :], axis = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "64581fe7-fa03-4d60-8709-50f32ea97a98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.3 s, sys: 286 ms, total: 24.6 s\n",
      "Wall time: 24.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# kernel died while computing these...\n",
    "mean_depth = np.mean(depth, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f9cb1c87-5898-4f05-b89a-bdf0dd46f50b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# median is the issue..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e97605ba-fa61-48cb-af38-f0397cd1c3d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7023976/7023976 [02:54<00:00, 40315.44it/s]\n"
     ]
    }
   ],
   "source": [
    "median_depth = np.full(snp_nr, -1)\n",
    "for i in tqdm(range(snp_nr)):\n",
    "    median_depth[i] =  np.median(depth[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f17a385d-4df1-496b-8957-1321392fbc3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.6 ms, sys: 9.71 ms, total: 34.3 ms\n",
      "Wall time: 36.7 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "depth_mask = (mean_depth < min_depth) | (mean_depth > max_depth) | (median_depth < min_depth) | (median_depth > max_depth) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "624e9a11-5912-4725-86a7-65e3f1c1eddc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "120911"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(depth_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f890a3e1-0bcd-481d-8aca-534faebc82ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [00:25, 279103.26it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 23.4 s, sys: 1.64 s, total: 25.1 s\n",
      "Wall time: 25.2 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "chr_ids = [''] * snp_nr\n",
    "snp_pos = [''] * snp_nr\n",
    "mito_mask = np.full(snp_nr, False)\n",
    "with open('{}count_A_1.txt'.format(out_path), 'r') as inf:\n",
    "    i = 0\n",
    "    for l in tqdm(inf):\n",
    "        ch = l.split()[0]\n",
    "        pos = l.split()[1]\n",
    "        chr_ids[i] = ch\n",
    "        snp_pos[i] = pos\n",
    "        if ch == 'NC_001566.1':\n",
    "            mito_mask[i] = True\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2f77e46c-789c-46b1-9395-6e9de0b34104",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "287"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(mito_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a50bfbf9-63ef-43a1-9ce4-7f87cf90f2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNP filtering\n",
    "mask = mono_mask + tri_mask + depth_mask + mito_mask + ambi_mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "714fa3dc-88e6-4b0b-8ccf-a1b07ed25bb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "259604\n"
     ]
    }
   ],
   "source": [
    "print(np.sum(mask))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5ce3488d-99b3-4269-a208-6b7e515e15b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6764372\n"
     ]
    }
   ],
   "source": [
    "print(snp_nr - np.sum(mask))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "68837dde-434a-4da7-a4d4-7dcf1bb79a1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 34s, sys: 1min 3s, total: 5min 38s\n",
      "Wall time: 5min 39s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# getting the ref allele counts 5-8min...\n",
    "count_ref  = np.max(count_mat[:, batch_i:batch_ii, :], axis = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "10362de9-1c78-4d11-b2ab-5640089a205d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.23 s, sys: 25.7 s, total: 29.9 s\n",
      "Wall time: 30.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# creating reference allele count dataframe\n",
    "count_ref_df = pd.DataFrame(count_ref[~mask])\n",
    "count_ref_df.insert(0, 'SNP_position', np.array(snp_pos)[~mask])\n",
    "count_ref_df.insert(0, 'Chromosome', np.array(chr_ids)[~mask])\n",
    "count_ref_df.columns = ['Chromosome', 'SNP_position'] + bs_ids[batch_i: batch_ii]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d54dd69a-6799-42ca-bdd0-6c9f918fc185",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 21min 20s, sys: 33.5 s, total: 21min 53s\n",
      "Wall time: 22min 11s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# 22min\n",
    "count_ref_df.to_csv('{}count_ref.txt'.format(out_path), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b433a9eb-134a-4aec-9221-7ef815578a26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.3 s, sys: 31min 29s, total: 31min 41s\n",
      "Wall time: 31min 49s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# creating depth dataframe # 30min...\n",
    "depth_df = pd.DataFrame(depth[~mask])\n",
    "depth_df.insert(0, 'SNP_position', np.array(snp_pos)[~mask])\n",
    "depth_df.insert(0, 'Chromosome',  np.array(chr_ids)[~mask])\n",
    "depth_df.columns = ['Chromosome', 'SNP_position'] + bs_ids[batch_i: batch_ii]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "772e295c-4312-46f3-a949-41277f6b7180",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 21min 19s, sys: 48.2 s, total: 22min 7s\n",
      "Wall time: 22min 41s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "depth_df.to_csv('{}depth.txt'.format(out_path), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f97615a7-8ef6-42e1-ada8-ea8dd5ec4a0f",
   "metadata": {},
   "source": [
    "## Queen genotype reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e422096e-1170-45e5-93a6-849d2ef25fb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f675d0fb-1963-46e8-b80f-642815a39ced",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_ref_df = pd.read_csv('{}count_ref.txt'.format(out_path), header=None, sep='\\t', nrows=10)\n",
    "depth_df = pd.read_csv('{}depth.txt'.format(out_path), header=None, sep='\\t', nrows=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "16e4b7b9-1fb1-42f7-921a-1601b530ec95",
   "metadata": {},
   "outputs": [],
   "source": [
    "header = ['Chromosome', 'SNP_position'] + bs_ids\n",
    "count_ref_df.columns = header\n",
    "depth_df.columns = header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9e0b05ed-e3d2-4eb8-b63b-8ab6edc1e944",
   "metadata": {},
   "outputs": [],
   "source": [
    "header_df = pd.DataFrame(columns=header)\n",
    "header_df.to_csv('{}header.txt'.format(out_path), header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "21f493d6-5801-40aa-bfae-071d2c75b256",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>BS16_0001</th>\n",
       "      <th>BS16_0002</th>\n",
       "      <th>BS16_0003</th>\n",
       "      <th>BS16_0004</th>\n",
       "      <th>BS16_0005</th>\n",
       "      <th>BS16_0006</th>\n",
       "      <th>BS16_0007</th>\n",
       "      <th>BS16_0008</th>\n",
       "      <th>...</th>\n",
       "      <th>BS18_0670</th>\n",
       "      <th>BS18_0671</th>\n",
       "      <th>BS18_0672</th>\n",
       "      <th>BS18_0673</th>\n",
       "      <th>BS18_0674</th>\n",
       "      <th>BS18_0675</th>\n",
       "      <th>BS18_0681</th>\n",
       "      <th>BS18_0682</th>\n",
       "      <th>BS18_0684</th>\n",
       "      <th>BS18_0686</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 1444 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, SNP_position, BS16_0001, BS16_0002, BS16_0003, BS16_0004, BS16_0005, BS16_0006, BS16_0007, BS16_0008, BS16_0009, BS16_0010, BS16_0011, BS16_0012, BS16_0013, BS16_0014, BS16_0015, BS16_0016, BS16_0017, BS16_0018, BS16_0019, BS16_0020, BS16_0021, BS16_0022, BS16_0023, BS16_0024, BS16_0025, BS16_0026, BS16_0027, BS16_0028, BS16_0029, BS16_0030, BS16_0031, BS16_0032, BS16_0033, BS16_0034, BS16_0035, BS16_0036, BS16_0037, BS16_0038, BS16_0039, BS16_0040, BS16_0041, BS16_0042, BS16_0043, BS16_0044, BS16_0045, BS16_0046, BS16_0047, BS16_0048, BS16_0049, BS16_0050, BS16_0051, BS16_0052, BS16_0053, BS16_0054, BS16_0055, BS16_0056, BS16_0057, BS16_0058, BS16_0059, BS16_0060, BS16_0061, BS16_0062, BS16_0063, BS16_0064, BS16_0065, BS16_0066, BS16_0067, BS16_0068, BS16_0069, BS16_0070, BS16_0071, BS16_0072, BS16_0074, BS16_0075, BS16_0076, BS16_0077, BS16_0078, BS16_0079, BS16_0080, BS16_0081, BS16_0082, BS16_0083, BS16_0084, BS16_0085, BS16_0086, BS16_0087, BS16_0088, BS16_0089, BS16_0090, BS16_0091, BS16_0092, BS16_0093, BS16_0094, BS16_0095, BS16_0096, BS16_0097, BS16_0098, BS16_0099, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 1444 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ab56c0ca-c7de-4ad0-ab8c-008d3afa2d2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>BS16_0001</th>\n",
       "      <th>BS16_0002</th>\n",
       "      <th>BS16_0003</th>\n",
       "      <th>BS16_0004</th>\n",
       "      <th>BS16_0005</th>\n",
       "      <th>BS16_0006</th>\n",
       "      <th>BS16_0007</th>\n",
       "      <th>BS16_0008</th>\n",
       "      <th>...</th>\n",
       "      <th>BS18_0670</th>\n",
       "      <th>BS18_0671</th>\n",
       "      <th>BS18_0672</th>\n",
       "      <th>BS18_0673</th>\n",
       "      <th>BS18_0674</th>\n",
       "      <th>BS18_0675</th>\n",
       "      <th>BS18_0681</th>\n",
       "      <th>BS18_0682</th>\n",
       "      <th>BS18_0684</th>\n",
       "      <th>BS18_0686</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>5671</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>72</td>\n",
       "      <td>31</td>\n",
       "      <td>47</td>\n",
       "      <td>44</td>\n",
       "      <td>18</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>46</td>\n",
       "      <td>33</td>\n",
       "      <td>27</td>\n",
       "      <td>39</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>5698</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>74</td>\n",
       "      <td>22</td>\n",
       "      <td>50</td>\n",
       "      <td>44</td>\n",
       "      <td>18</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>41</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>27</td>\n",
       "      <td>45</td>\n",
       "      <td>31</td>\n",
       "      <td>23</td>\n",
       "      <td>39</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>6621</td>\n",
       "      <td>48</td>\n",
       "      <td>30</td>\n",
       "      <td>113</td>\n",
       "      <td>23</td>\n",
       "      <td>50</td>\n",
       "      <td>39</td>\n",
       "      <td>40</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>71</td>\n",
       "      <td>61</td>\n",
       "      <td>41</td>\n",
       "      <td>53</td>\n",
       "      <td>34</td>\n",
       "      <td>73</td>\n",
       "      <td>65</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>6717</td>\n",
       "      <td>30</td>\n",
       "      <td>28</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>58</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>44</td>\n",
       "      <td>49</td>\n",
       "      <td>48</td>\n",
       "      <td>41</td>\n",
       "      <td>61</td>\n",
       "      <td>46</td>\n",
       "      <td>33</td>\n",
       "      <td>40</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>6823</td>\n",
       "      <td>39</td>\n",
       "      <td>29</td>\n",
       "      <td>61</td>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>39</td>\n",
       "      <td>31</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>50</td>\n",
       "      <td>45</td>\n",
       "      <td>27</td>\n",
       "      <td>38</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>6832</td>\n",
       "      <td>44</td>\n",
       "      <td>29</td>\n",
       "      <td>57</td>\n",
       "      <td>19</td>\n",
       "      <td>38</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>43</td>\n",
       "      <td>44</td>\n",
       "      <td>23</td>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "      <td>46</td>\n",
       "      <td>38</td>\n",
       "      <td>37</td>\n",
       "      <td>28</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>6835</td>\n",
       "      <td>45</td>\n",
       "      <td>29</td>\n",
       "      <td>64</td>\n",
       "      <td>21</td>\n",
       "      <td>39</td>\n",
       "      <td>38</td>\n",
       "      <td>36</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>44</td>\n",
       "      <td>47</td>\n",
       "      <td>25</td>\n",
       "      <td>39</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>27</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>6914</td>\n",
       "      <td>41</td>\n",
       "      <td>25</td>\n",
       "      <td>67</td>\n",
       "      <td>26</td>\n",
       "      <td>45</td>\n",
       "      <td>39</td>\n",
       "      <td>46</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>44</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>54</td>\n",
       "      <td>33</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>7034</td>\n",
       "      <td>21</td>\n",
       "      <td>13</td>\n",
       "      <td>65</td>\n",
       "      <td>19</td>\n",
       "      <td>29</td>\n",
       "      <td>33</td>\n",
       "      <td>18</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>35</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>33</td>\n",
       "      <td>26</td>\n",
       "      <td>47</td>\n",
       "      <td>29</td>\n",
       "      <td>25</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>7092</td>\n",
       "      <td>24</td>\n",
       "      <td>21</td>\n",
       "      <td>61</td>\n",
       "      <td>28</td>\n",
       "      <td>35</td>\n",
       "      <td>34</td>\n",
       "      <td>29</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>33</td>\n",
       "      <td>24</td>\n",
       "      <td>42</td>\n",
       "      <td>24</td>\n",
       "      <td>40</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 1444 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome  SNP_position  BS16_0001  BS16_0002  BS16_0003  BS16_0004  \\\n",
       "0  NC_037638.1          5671         18         19         72         31   \n",
       "1  NC_037638.1          5698         20         16         74         22   \n",
       "2  NC_037638.1          6621         48         30        113         23   \n",
       "3  NC_037638.1          6717         30         28        105         25   \n",
       "4  NC_037638.1          6823         39         29         61         21   \n",
       "5  NC_037638.1          6832         44         29         57         19   \n",
       "6  NC_037638.1          6835         45         29         64         21   \n",
       "7  NC_037638.1          6914         41         25         67         26   \n",
       "8  NC_037638.1          7034         21         13         65         19   \n",
       "9  NC_037638.1          7092         24         21         61         28   \n",
       "\n",
       "   BS16_0005  BS16_0006  BS16_0007  BS16_0008  ...  BS18_0670  BS18_0671  \\\n",
       "0         47         44         18         34  ...         37         36   \n",
       "1         50         44         18         34  ...         39         41   \n",
       "2         50         39         40         36  ...         71         61   \n",
       "3         58         37         37         34  ...         58         44   \n",
       "4         40         39         31         24  ...         50         45   \n",
       "5         38         40         33         22  ...         43         44   \n",
       "6         39         38         36         22  ...         44         47   \n",
       "7         45         39         46         23  ...         37         36   \n",
       "8         29         33         18         24  ...         35         30   \n",
       "9         35         34         29         25  ...         34         33   \n",
       "\n",
       "   BS18_0672  BS18_0673  BS18_0674  BS18_0675  BS18_0681  BS18_0682  \\\n",
       "0         24         31         31         46         33         27   \n",
       "1         26         25         27         45         31         23   \n",
       "2         41         53         34         73         65         28   \n",
       "3         49         48         41         61         46         33   \n",
       "4         27         38         30         46         38         38   \n",
       "5         23         38         33         46         38         37   \n",
       "6         25         39         33         49         38         38   \n",
       "7         44         43         43         54         33         37   \n",
       "8         29         33         26         47         29         25   \n",
       "9         24         42         24         40         22         24   \n",
       "\n",
       "   BS18_0684  BS18_0686  \n",
       "0         39         38  \n",
       "1         39         33  \n",
       "2         47         48  \n",
       "3         40         48  \n",
       "4         27         32  \n",
       "5         28         35  \n",
       "6         27         36  \n",
       "7         36         42  \n",
       "8         27         32  \n",
       "9         33         16  \n",
       "\n",
       "[10 rows x 1444 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4993d488-1295-44d1-9fa2-36f9e351b18d",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_ref_df.to_csv('{}count_ref_test.txt'.format(out_path), header=True, index=False, sep='\\t')\n",
    "depth_df.to_csv('{}depth_test.txt'.format(out_path), header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7ea86c7a-57de-4e1c-be93-3b885b6d4345",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_run_beethoven_script(script_fn):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --mem=40g\n",
    "#SBATCH --time=48:00:00\n",
    "#SBATCH --job-name=beethoven\n",
    "#SBATCH --output=%x_%j.out\n",
    "#SBATCH --error=%x_%j.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "dir_in=$1\n",
    "depth_file=$2\n",
    "count_ref_file=$3\n",
    "prefix=$4\n",
    "\n",
    "genoqueen_hom ${dir_in} ${depth_file} ${count_ref_file} 2 10 ${prefix} 1000\n",
    "\n",
    "echo DONE\"\"\"\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7676fcb5-8d46-4847-80e2-cfce1b5d6c00",
   "metadata": {},
   "outputs": [],
   "source": [
    "script_fn = '{}beethoven.run'.format(script_path)\n",
    "write_run_beethoven_script(script_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ca5e6af6-2af7-4914-a7a4-e67e5bd6bb2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_in = out_path\n",
    "depth_file = 'depth_header.txt'\n",
    "count_ref_file = 'count_ref_header.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0c950efa-5c2b-46ef-9858-642e62fc611b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 1369553\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$script_fn\" \"$dir_in\" \"$depth_file\" \"$count_ref_file\"\n",
    "cd $2\n",
    "sbatch $1 $2 $3 $4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d0aa7b0-d630-41a4-accb-837c1108b87b",
   "metadata": {},
   "source": [
    "### by population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7255baac-a984-4f6a-8cf5-fc0d3ad469f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_in = allele_count_path\n",
    "depth_file = 'depth_header.txt'\n",
    "count_ref_file = 'count_ref_header.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1ce83c76-3a65-440b-b786-30b417cb800f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(pheno_df['Genetic group'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7bb498d2-234e-4d55-a305-703dad07b1e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6min 59s, sys: 5min 11s, total: 12min 11s\n",
      "Wall time: 12min 15s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "depth_df = pd.read_csv('{}{}'.format(dir_in, depth_file), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b99f4e84-0c33-41be-945e-bd598cb49a77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caucasica\n",
      "Mellifera\n",
      "Ligustica_Carnica\n",
      "hybrid\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for gg in set(pheno_df['Genetic group']):\n",
    "    print(gg)\n",
    "    depth_df[['Chromosome', 'SNP_position'] + list(pheno_df[pheno_df['Genetic group'] == gg]['Colony'])].to_csv('{}depth_{}.txt'.format(allele_count_path, gg), header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9e208260-babe-48b5-b38e-fecc53be08fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 28s, sys: 3min 19s, total: 10min 48s\n",
      "Wall time: 11min 18s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "count_ref_df = pd.read_csv('{}{}'.format(dir_in, count_ref_file), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "40be25e3-37a5-456b-bb99-cae9971a2758",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mellifera\n",
      "Caucasica\n",
      "hybrid\n",
      "Ligustica_Carnica\n",
      "CPU times: user 20min 51s, sys: 1min 14s, total: 22min 5s\n",
      "Wall time: 22min 12s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for gg in set(pheno_df['Genetic group']):\n",
    "    print(gg)\n",
    "    count_ref_df[['Chromosome', 'SNP_position'] + list(pheno_df[pheno_df['Genetic group'] == gg]['Colony'])].to_csv('{}count_ref_{}.txt'.format(allele_count_path, gg), header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a585b191-f428-43ac-870d-673fde9501fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "script_fn = '{}beethoven.run'.format(script_path)\n",
    "write_run_beethoven_script(script_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9e2df110-77d3-43db-a0b3-1f7a4ac48658",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_in = '/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d3ebda3-e357-4863-a874-3d495a671526",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gg in set(pheno_df['Genetic group']):\n",
    "    command = ['sbatch', script_fn, dir_in, 'depth_{}.txt'.format(gg), 'count_ref_{}.txt'.format(gg), 'queen_{}'.format(gg)]\n",
    "    result = subprocess.run(command, cwd=dir_in, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af8a7cd9-db02-4971-bb89-4b40a6586ac7",
   "metadata": {},
   "source": [
    "## PLINK\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16de778-6585-435d-a8fe-bb006369c02d",
   "metadata": {},
   "source": [
    "### .bim_like"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8baf3105-ed80-48dd-a4dc-e31d7c520d9f",
   "metadata": {},
   "source": [
    "I need to create bim-like files but be careful they are overwritten with make-bed command\n",
    "\n",
    "1. Chromosome code\n",
    "2. Variant ID\n",
    "3. Position in centimorgans (safe to use dummy value of '0')\n",
    "4. Base-pair coordinate (1-based; limited to 231-2)\n",
    "5. ALT ('A1' in PLINK 1.x) allele code\n",
    "6. REF ('A2' in PLINK 1.x) allele code\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "e78177aa-b0f1-4f67-942b-a5d4c9ab5d1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "81.028587136\n",
      "CPU times: user 0 ns, sys: 1min 33s, total: 1min 33s\n",
      "Wall time: 1min 52s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "with open('{}count_mat.pkl'.format(allele_count_path), 'rb') as inf:\n",
    "    count_mat = pkl.load(inf)\n",
    "print(count_mat.nbytes/1000000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "ab217aa9-7920-483c-a524-5a30cdf3091c",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_snp_nr = 6764372\n",
    "snp_nr = 7023976\n",
    "assert snp_nr == count_mat.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "6ca86c4d-f3aa-4dfd-ac6d-e9f957856237",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM POS REF ALT BS17-0071\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [00:05, 1403874.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.31 s, sys: 685 ms, total: 5 s\n",
      "Wall time: 5.08 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# load original SNP ids\n",
    "chr_ids = [''] * snp_nr\n",
    "snp_pos = [''] * snp_nr\n",
    "with open('{}data/gwas/BS17_0071.count'.format(working_path), 'r') as inf:\n",
    "    print(inf.readline())\n",
    "    i = 0\n",
    "    for l in tqdm(inf):\n",
    "        ch = l.split()[0]\n",
    "        pos = l.split()[1]\n",
    "        chr_ids[i] = ch\n",
    "        snp_pos[i] = pos\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "616695fb-3100-4116-8871-d96002a6bd5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6764372it [00:04, 1543827.37it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.7 s, sys: 676 ms, total: 4.37 s\n",
      "Wall time: 4.45 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# load filtered SNPs\n",
    "f_chr_ids = [''] * f_snp_nr\n",
    "f_snp_pos = [''] * f_snp_nr\n",
    "with open('{}filtered_snps.txt'.format(allele_count_path), 'r') as inf:\n",
    "    inf.readline()\n",
    "    i = 0\n",
    "    for l in tqdm(inf):\n",
    "        ch = l.split()[0]\n",
    "        pos = l.split()[1]\n",
    "        f_chr_ids[i] = ch\n",
    "        f_snp_pos[i] = pos\n",
    "        i += 1\n",
    "assert f_snp_nr == i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "49518cfa-38ec-48e7-845d-8c82f7939c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_path = '{}data/queen_genotypes/hom_pop/'.format(working_path)\n",
    "bim = pd.read_csv('{}{}.bim'.format(plink_path, 'queen'), sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "a0c2aa05-01df-433d-9807-37607ae8b59c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# good\n",
    "bim[3].to_list() == [int(x) for x in f_snp_pos]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "c96187ca-d725-44bd-a177-f3ae3f0b014b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get back mask by comparing original snps and filtered snps. TO DO: if rerun filtering, include this part directly\n",
    "snp_ids = ['{}_{}'.format(x, snp_pos[i]) for i, x in enumerate(chr_ids)]\n",
    "f_snp_ids = ['{}_{}'.format(x, f_snp_pos[i]) for i, x in enumerate(f_chr_ids)]\n",
    "\n",
    "assert len(snp_ids) == len(set(snp_ids))\n",
    "assert len(f_snp_ids) == len(set(f_snp_ids))\n",
    "\n",
    "snp_mask = ~np.isin(np.array(snp_ids), np.array(f_snp_ids))\n",
    "\n",
    "assert np.sum(~snp_mask) == f_snp_nr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "5c85e1db-7dcf-4236-979a-e4b88319de33",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6764372/6764372 [03:30<00:00, 32146.91it/s]\n"
     ]
    }
   ],
   "source": [
    "# find reference and alternative allele per SNP (REF: the one with the most counts. not sure if correct but probably REF and ALT are interchangeable) \n",
    "jj2allele = np.array(['A', 'T', 'C', 'G'])\n",
    "ref_alt = np.full((snp_nr, 2), '.')\n",
    "for i in tqdm(np.argwhere(~snp_mask).flatten()):\n",
    "    ref_alt[i] = jj2allele[np.argsort(np.sum(count_mat[i, :, :], axis=0))[::-1][:2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "d0630f4a-293d-40ca-b525-4eb52f7355ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_like = pd.DataFrame({\n",
    "    'CHR': f_chr_ids,\n",
    "    'SNP_ID': f_snp_ids,\n",
    "    'GEN_DIST': ['0'] * f_snp_nr,\n",
    "    'BP_POS': f_snp_pos,\n",
    "    'ALT': ref_alt[~snp_mask][:, 1],\n",
    "    'REF': ref_alt[~snp_mask][:, 0]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "22e8c407-a0ec-483b-8cb4-9d21008fe934",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3354437/2140746017.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  bim_like['CHR'] = bim_like['CHR'].replace(chr2int)\n"
     ]
    }
   ],
   "source": [
    "# recode snp id\n",
    "chr2int = dict(zip(list(dict.fromkeys(bim_like['CHR'])), range(1, 1 + len(set(bim_like['CHR'])))))\n",
    "bim_like['CHR'] = bim_like['CHR'].replace(chr2int)\n",
    "\n",
    "def recode_SNP_ids(old_id):\n",
    "    chrom, pos = old_id.split('.1_')\n",
    "    return '{}:{}'.format(chr2int['{}.1'.format(chrom)], pos)\n",
    "\n",
    "bim_like['SNP_ID'] = bim_like['SNP_ID'].map(recode_SNP_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "c260a07c-ffec-4348-8417-06c6b55296c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GEN_DIST</th>\n",
       "      <th>BP_POS</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR      SNP_ID GEN_DIST   BP_POS ALT REF\n",
       "0          1      1:5671        0     5671   G   T\n",
       "1          1      1:5698        0     5698   C   T\n",
       "2          1      1:6621        0     6621   G   A\n",
       "3          1      1:6717        0     6717   T   C\n",
       "4          1      1:6823        0     6823   C   G\n",
       "...      ...         ...      ...      ...  ..  ..\n",
       "6764367   16  16:7225908        0  7225908   C   T\n",
       "6764368   16  16:7225911        0  7225911   T   C\n",
       "6764369   16  16:7226186        0  7226186   A   G\n",
       "6764370   16  16:7226346        0  7226346   A   G\n",
       "6764371   16  16:7227034        0  7227034   A   G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "8853381a-a668-4b1b-a21b-55aa4ee0c057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# good\n",
    "bim[1].to_list() == bim_like['SNP_ID'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "f398c788-f3cf-4d90-bd85-78ca6fcf559d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.47 s, sys: 509 ms, total: 6.98 s\n",
      "Wall time: 7.11 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "bim_like.to_csv('{}queen.bim_like'.format(hom_pop_path), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "0b0b63a9-5207-4e62-be80-91410803f736",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/'"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hom_pop_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2e4db9d-571f-4dd9-ba4f-c10f3043dcf1",
   "metadata": {},
   "source": [
    "### .ped and .bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ec2c9b-6092-4147-8b44-71de61bd9bdc",
   "metadata": {},
   "source": [
    "Family ID (FID)\n",
    "\n",
    "Individual ID (IID)\n",
    "\n",
    "Paternal ID\n",
    "\n",
    "Maternal ID\n",
    "\n",
    "Sex\n",
    "\n",
    "Phenotype\n",
    "\n",
    "genotypes 2 columns per SNP when diploid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c34aae5b-9d0b-48dd-9f50-4312a56c7998",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_j = 3\n",
    "chunk_bs_ids = batch_bs_ids[batch_j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e5c869f-b162-4f16-a01c-1085ab665ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_ped(bim_file, chunk_bs_ids, snp_nr, bgs_file, batch_j, chunk_size, plink_path, prefix):\n",
    "\n",
    "    # contains REF and ALT alleles\n",
    "    bim_df = pd.read_csv(bim_file, sep='\\t')\n",
    "    \n",
    "    ped_mat = np.full((len(chunk_bs_ids), 2 * snp_nr), '', dtype='U1')\n",
    "    \n",
    "    with open(bgs_file, 'r') as inf:\n",
    "        \n",
    "        # index of the first colony of the batch/chunk\n",
    "        start_j = batch_j * chunk_size\n",
    "    \n",
    "        # check we have the same bs ids in header than in batch\n",
    "        header_bsids = inf.readline()\n",
    "        assert header_bsids.rstrip().split(',')[3:][start_j: start_j + len(chunk_bs_ids)] == chunk_bs_ids, 'bs_id mismatch'\n",
    "        \n",
    "        for i in tqdm(range(snp_nr)):\n",
    "            # parse number of ref allele per SNP (0, 1, 2 as queen is diploid)\n",
    "            ref_numbers = inf.readline().rstrip().split(',')[3:]\n",
    "    \n",
    "            for j in range(start_j, start_j + len(chunk_bs_ids)):\n",
    "                ref_nr = ref_numbers[j]\n",
    "                if ref_nr == '2':\n",
    "                    ped_mat[j - start_j, 2 * i] = bim_df.iloc[i, 5]\n",
    "                    ped_mat[j - start_j, 2 * i + 1] = bim_df.iloc[i, 5]\n",
    "                elif ref_nr == '1': \n",
    "                    ped_mat[j - start_j, 2 * i] = bim_df.iloc[i, 5]\n",
    "                    ped_mat[j - start_j, 2 * i + 1] = bim_df.iloc[i, 4]\n",
    "                elif ref_nr == '0': \n",
    "                    ped_mat[j - start_j, 2 * i] = bim_df.iloc[i, 4]\n",
    "                    ped_mat[j - start_j, 2 * i + 1] = bim_df.iloc[i, 4]\n",
    "                else: \n",
    "                    print('ERROR')\n",
    "\n",
    "    ped_df = pd.DataFrame({\n",
    "        'FID': chunk_bs_ids, \n",
    "        'IID': chunk_bs_ids,\n",
    "        'PID': ['0'] * len(chunk_bs_ids),\n",
    "        'MID': ['0'] * len(chunk_bs_ids),\n",
    "        'SEX': ['0'] * len(chunk_bs_ids), \n",
    "        'PHENOTYPE': ['-9'] * len(chunk_bs_ids)\n",
    "    })\n",
    "    ped_df = ped_df.join(pd.DataFrame(ped_mat))\n",
    "    ped_df.to_csv('{}{}_{}.ped'.format(plink_path, prefix, batch_j), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8dd7acf3-4fbf-4a5d-b08a-7d10ee7a57cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_write_ped_script(script_fn, array_str):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --array=ARRAY_STRING\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=100g\n",
    "#SBATCH --time=6:00:00\n",
    "#SBATCH --job-name=write_ped\n",
    "#SBATCH --output=%x_%A_%a.out\n",
    "#SBATCH --error=%x_%A_%a.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "# file listing batches of BeeStrong ids (bs_id) for parallel computing\n",
    "bim_file=$1\n",
    "snp_nr=$2\n",
    "bgs_file=$3\n",
    "chunk_size=$4\n",
    "plink_path=$5\n",
    "prefix=$6\n",
    "batch_file=$7\n",
    "\n",
    "batch_id=$SLURM_ARRAY_TASK_ID\n",
    "\n",
    "# getting bs_ids\n",
    "chunk_bs_ids=$(sed -n ${batch_id}'{p;q}' ${batch_file})\n",
    "\n",
    "python ~/scripts/plan-b-omics/bin/write_ped.py --bim_file ${bim_file} --chunk_bs_ids ${chunk_bs_ids} --snp_nr ${snp_nr} --bgs_file ${bgs_file} --batch_id ${batch_id} --chunk_size ${chunk_size} --plink_path ${plink_path} --prefix ${prefix}\n",
    "\n",
    "echo DONE\"\"\".replace(\"ARRAY_STRING\", array_str)\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "64609dda-fe3d-4dd5-aacd-2dd57ec266e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'queen'\n",
    "bim_file = '{}{}.bim'.format(allele_count_path, prefix)\n",
    "snp_nr = 6764372\n",
    "bgs_file = '{}{}.bgs'.format(allele_count_path, prefix)\n",
    "chunk_size = 50\n",
    "batch_file = '{}bsid_batch.txt'.format(plink_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d32a1b5c-5fa9-4263-8052-9776617a5872",
   "metadata": {},
   "outputs": [],
   "source": [
    "# prod\n",
    "bs_ids = pheno_df['Colony'].to_list()\n",
    "batch_bs_ids = []\n",
    "with open(batch_file, \"w\") as outf:\n",
    "    for i in range(0, len(bs_ids), chunk_size):\n",
    "        chunk = bs_ids[i:i+chunk_size]\n",
    "        batch_bs_ids.append(chunk)\n",
    "        outf.write(\",\".join(chunk) + \"\\n\")\n",
    "assert sum([x == bs_ids[i] for i, x in enumerate(list(chain(*batch_bs_ids)))]) == len(bs_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0a6a449d-fd38-482b-b92f-ec3533d51996",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_str = '1-{}'.format(len(batch_bs_ids))\n",
    "script_fn = '{}write_ped.run'.format(script_path)\n",
    "write_write_ped_script(script_fn, array_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "91cb1198-0b9b-4535-9d35-de7503cf64b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 1378632\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$script_fn\" \"$bim_file\" \"$snp_nr\" \"$bgs_file\" \"$chunk_size\" \"$plink_path\" \"$prefix\" \"$batch_file\" \n",
    "cd $6\n",
    "sbatch $1 $2 $3 $4 $5 $6 $7 $8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2835342d-4609-4cd1-856d-79c10cececb0",
   "metadata": {},
   "source": [
    "a simple cat on all theose .ped files should work"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96e3a0f4-a66a-455f-b577-f0adcf7fb4df",
   "metadata": {},
   "source": [
    "### .ped with probs and by population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "ca56b034-d5b7-44bb-9f04-ebcece1527e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'queen'\n",
    "plink_path = '{}data/queen_genotypes/hom_pop/'.format(working_path)\n",
    "populations = ['Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid']\n",
    "chunk_size = 25\n",
    "prob_threshold = 0.95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "155a4732-09fe-4e61-b5e1-8ccea375f6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to split .prob files into 3: AA, AR, RR --> easier to read probability of these genotypes simulateously then...\n",
    "def split_prob_file(\n",
    "    plink_path, prefix, pop):\n",
    "\n",
    "    prob_file = '{}{}_{}.prob'.format(plink_path, prefix, pop)\n",
    "    aa_file = '{}{}_{}_AA.prob'.format(plink_path, prefix, pop)\n",
    "    ar_file = '{}{}_{}_AR.prob'.format(plink_path, prefix, pop)\n",
    "    rr_file = '{}{}_{}_RR.prob'.format(plink_path, prefix, pop)\n",
    "    \n",
    "    out_aa = open(aa_file, 'w')\n",
    "    out_ar = open(ar_file, 'w')\n",
    "    out_rr = open(rr_file, 'w')\n",
    "    \n",
    "    with open(prob_file, 'r') as inf:\n",
    "        headers = inf.readline()\n",
    "        out_aa.write(headers.rstrip(',geno\\n') + '\\n')\n",
    "        out_ar.write(headers.rstrip(',geno\\n') + '\\n')\n",
    "        out_rr.write(headers.rstrip(',geno\\n') + '\\n')\n",
    "\n",
    "        for l in tqdm(inf):\n",
    "            split_line = l.rstrip().split(',')\n",
    "            geno = split_line[-1]\n",
    "            if geno == 'AA':\n",
    "                out_aa.write(','.join(split_line[:-1]) + '\\n')\n",
    "            elif geno == 'AR':\n",
    "                out_ar.write(','.join(split_line[:-1]) + '\\n')\n",
    "            elif geno == 'RR':\n",
    "                out_rr.write(','.join(split_line[:-1]) + '\\n')\n",
    "            else: \n",
    "                print('GENO: {}'.format(geno))\n",
    "                      \n",
    "    out_aa.close()\n",
    "    out_ar.close()\n",
    "    out_rr.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "52986b4b-f649-482c-8249-75333eb0acdc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20293116it [00:46, 437729.20it/s]\n",
      "20293116it [24:53, 13589.27it/s]\n",
      "20293116it [15:02, 22474.99it/s]\n",
      "20293116it [14:04, 24026.43it/s]\n"
     ]
    }
   ],
   "source": [
    "for pop in populations:\n",
    "    split_prob_file(plink_path, prefix, pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "d0f67534-50ad-4f27-bc47-e801a1aea241",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n",
      "645\n",
      "405\n",
      "372\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "68e73475-e363-451a-b709-38fbcaf14c05",
   "metadata": {},
   "source": [
    "#### debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0bdf42e7-7775-4c83-bf69-9d8aeb92352a",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim = pd.read_csv('{}{}.bim'.format(plink_path, prefix), sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "32599aaa-8a69-47f9-a7b8-781b35bf91d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop = 'Mellifera'\n",
    "batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)\n",
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e5ab8bee-dbf1-42a7-af37-f1c72f28c3ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "batch_j = 0\n",
    "chunk_bs_ids = batch_bs_ids[batch_j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3681190f-48a4-4f7e-8327-f541f452209d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'BS16_0136'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chunk_bs_ids[7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "45c6fcf5-e6d9-48de-918c-2fba7ec08366",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim = pd.read_csv('{}{}.bim'.format(plink_path, prefix), sep='\\t', header=None)\n",
    "bgs_file = '{}{}_{}.bgs'.format(plink_path, prefix, pop)\n",
    "aa_prob_file = '{}{}_{}_AA.prob'.format(plink_path, prefix, pop)\n",
    "ar_prob_file = '{}{}_{}_AR.prob'.format(plink_path, prefix, pop)\n",
    "rr_prob_file = '{}{}_{}_RR.prob'.format(plink_path, prefix, pop)\n",
    "snp_nr = len(bim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "cc8fa192-559d-4a05-b0a2-d06e1c24d03b",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_j = batch_j * chunk_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "1eb61ccb-6315-48d8-a72c-847bf3246fc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "inf = open(bgs_file, 'r')\n",
    "inf_aa = open(aa_prob_file, 'r')\n",
    "inf_ar = open(ar_prob_file, 'r')\n",
    "inf_rr = open(rr_prob_file, 'r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "d6727d54-4284-4feb-8ef0-0c3ec8c2c908",
   "metadata": {},
   "outputs": [],
   "source": [
    "header_bsids = inf.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "a587764a-bf2f-4f21-ae58-b7efcd6cecbd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "',Chromosome,SNP_position,BS16_0004,BS16_0032,BS16_0040,BS16_0044,BS16_0045,BS16_0047,BS16_0134,BS16_0136,BS16_0137,BS16_0141,BS16_0144,BS16_0145,BS16_0146,BS16_0147,BS16_0148,BS16_0149,BS16_0150,BS16_0151,BS16_0152,BS16_0154,BS16_0155,BS16_0171,BS16_0172,BS16_0174,BS16_0175,BS16_0176,BS16_0216,BS16_0218,BS16_0276,BS16_0277,BS16_0279,BS16_0280,BS16_0300,BS16_0326,BS16_0382,BS16_0404,BS16_0407,BS16_0408,BS16_0409,BS16_0410,BS16_0411,BS16_0412,BS16_0413,BS16_0414,BS16_0415,BS16_0416,BS16_0417,BS16_0418,BS16_0419,BS16_0420,BS16_0421,BS16_0422,BS16_0423,BS16_0424,BS16_0425,BS16_0426,BS16_0427,BS16_0428,BS16_0429,BS16_0437,BS16_0438,BS16_0439,BS16_0440,BS16_0449,BS16_0496,BS16_0500,BS16_0501,BS16_0502,BS16_0503,BS16_0504,BS16_0511,BS16_0513,BS16_0514,BS16_0515,BS16_0516,BS16_0517,BS16_0518,BS16_0520,BS16_0521,BS16_0540,BS16_0541,BS16_0542,BS16_0543,BS16_0544,BS16_0545,BS16_0546,BS16_0547,BS16_0549,BS16_0550,BS16_0551,BS16_0553,BS16_0554,BS16_0570,BS16_0571,BS16_0572,BS16_0573,BS16_0574,BS17_0108,BS17_0110,BS17_0111,BS17_0112,BS17_0113,BS17_0114,BS17_0157,BS17_0171,BS17_0172,BS17_0173,BS17_0174,BS17_0176,BS17_0177,BS17_0178,BS17_0179,BS17_0180,BS17_0213,BS17_0218,BS17_0230,BS17_0410,BS17_0436,BS17_0452,BS17_0469,BS17_0474,BS17_0480,BS17_0484,BS17_0488,BS17_0495,BS17_0500,BS17_0509,BS17_0522,BS17_0703,BS17_0704,BS17_0705,BS17_0706,BS17_0707,BS17_0708,BS17_0709,BS17_0710,BS17_0711,BS17_0712,BS17_0717,BS17_0718,BS17_0719,BS17_0720,BS17_0722,BS17_0723,BS17_0724,BS17_0725,BS17_0726,BS17_0727,BS17_0728,BS17_0729,BS17_0730,BS17_0731,BS17_0732,BS17_0733,BS17_0734,BS17_0735,BS17_0736,BS17_0737,BS17_0738,BS17_0739,BS17_0740,BS17_0741,BS17_0742,BS17_0743,BS17_0744,BS17_0745,BS17_0746,BS17_0747,BS17_0748,BS17_0749,BS17_0750,BS17_0751,BS17_0752,BS18_0001,BS18_0002,BS18_0003,BS18_0004,BS18_0005,BS18_0006,BS18_0007,BS18_0008,BS18_0009,BS18_0010,BS18_0011,BS18_0012,BS18_0013,BS18_0014,BS18_0015,BS18_0016,BS18_0017,BS18_0018,BS18_0019,BS18_0020,BS18_0021,BS18_0022,BS18_0023,BS18_0024,BS18_0025,BS18_0026,BS18_0038,BS18_0039,BS18_0040,BS18_0041,BS18_0042,BS18_0043,BS18_0044,BS18_0045,BS18_0046,BS18_0047,BS18_0048,BS18_0049,BS18_0050,BS18_0051,BS18_0052,BS18_0053,BS18_0054,BS18_0055,BS18_0056,BS18_0057,BS18_0058,BS18_0059,BS18_0060,BS18_0061,BS18_0062,BS18_0063,BS18_0064,BS18_0065,BS18_0066,BS18_0067,BS18_0068,BS18_0069,BS18_0070,BS18_0071,BS18_0072,BS18_0073,BS18_0074,BS18_0075,BS18_0076,BS18_0077,BS18_0078,BS18_0079,BS18_0080,BS18_0081,BS18_0082,BS18_0083,BS18_0084,BS18_0085,BS18_0086,BS18_0087,BS18_0088,BS18_0089,BS18_0090,BS18_0091,BS18_0092,BS18_0093,BS18_0094,BS18_0095,BS18_0096,BS18_0097,BS18_0098,BS18_0099,BS18_0100,BS18_0101,BS18_0102,BS18_0103,BS18_0104,BS18_0105,BS18_0106,BS18_0107,BS18_0108,BS18_0109,BS18_0110,BS18_0111,BS18_0112,BS18_0113,BS18_0114,BS18_0115,BS18_0116,BS18_0117,BS18_0118,BS18_0119,BS18_0120,BS18_0121,BS18_0122,BS18_0123,BS18_0124,BS18_0125,BS18_0126,BS18_0127,BS18_0128,BS18_0129,BS18_0130,BS18_0131,BS18_0132,BS18_0133,BS18_0134,BS18_0135,BS18_0136,BS18_0137,BS18_0138,BS18_0140,BS18_0141,BS18_0142,BS18_0143,BS18_0144,BS18_0145,BS18_0146,BS18_0147,BS18_0148,BS18_0149,BS18_0150,BS18_0151,BS18_0152,BS18_0153,BS18_0154,BS18_0155,BS18_0156,BS18_0157,BS18_0158,BS18_0159,BS18_0160,BS18_0161,BS18_0162,BS18_0163,BS18_0164,BS18_0165,BS18_0166,BS18_0167,BS18_0168,BS18_0260,BS18_0261,BS18_0270,BS18_0282,BS18_0283,BS18_0284,BS18_0312,BS18_0313,BS18_0314,BS18_0315,BS18_0316,BS18_0318,BS18_0319,BS18_0321,BS18_0322,BS18_0323,BS18_0324,BS18_0325,BS18_0326,BS18_0328,BS18_0329,BS18_0524,BS18_0525,BS18_0529,BS18_0552,BS18_0553,BS18_0554,BS18_0555,BS18_0557,BS18_0558,BS18_0559,BS18_0560,BS18_0588,BS18_0601,BS18_0602,BS18_0603,BS18_0604,BS18_0605,BS18_0606,BS18_0607,BS18_0608,BS18_0609,BS18_0610,BS18_0621,BS18_0623,BS18_0624,BS18_0625,BS18_0626,BS18_0627,BS18_0628,BS18_0629,BS18_0630,BS18_0631,BS18_0633,BS18_0634,BS18_0635,BS18_0636,BS18_0637,BS18_0638,BS18_0639,BS18_0640,BS18_0651,BS18_0652,BS18_0653,BS18_0654,BS18_0655,BS18_0656,BS18_0657,BS18_0658,BS18_0659,BS18_0660,BS18_0661,BS18_0662,BS18_0663,BS18_0665,BS18_0666\\n'"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assert header_bsids.rstrip().split(',')[3:][start_j: start_j + len(chunk_bs_ids)] == chunk_bs_ids, 'bs_id mismatch'\n",
    "inf_aa.readline()\n",
    "inf_ar.readline()\n",
    "inf_rr.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "9cefb25e-ebbe-4338-ab01-0da83b50ee48",
   "metadata": {},
   "outputs": [],
   "source": [
    "inf.close()\n",
    "inf_aa.close()\n",
    "inf_ar.close()\n",
    "inf_rr.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "d4f0c300-5e9d-4bfb-b358-480166750291",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 92%|█████████▏| 6240306/6764372 [02:52<00:14, 36224.92it/s]\n"
     ]
    }
   ],
   "source": [
    "for i in tqdm(range(snp_nr)):\n",
    "    ref_numbers = inf.readline()\n",
    "    aa_probs = inf_aa.readline()\n",
    "    ar_probs = inf_ar.readline()\n",
    "    rr_probs = inf_rr.readline()\n",
    "    if i == 6240306: \n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0f6b95-58e9-478c-962e-c99bca4a5f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "9d8996ab-a716-4a50-a92d-3382c9a0e83a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'6240306,NC_037651.1,9770952,2,2,2,2,2,2,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2\\n'"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "2448b56b-883c-4705-bcda-8cad2c2a6902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'6240306,NC_037651.1,9770952,1.1253998361201744e-07,1.9988935406767335e-06,8.44031190971643e-08,4.742949587530005e-07,2.675415631799144e-10,1.9988935406767335e-06,0.9975190960441253,6.318023592626571e-06,6.330093744971093e-08,2.675415631799144e-10,1.9988935406767335e-06,8.44031190971643e-08,1.1253998361201744e-07,3.5537405396740974e-06,1.4991376729384742e-06,3.5537405396740974e-06,8.432281095893794e-07,1.1232484713080014e-05,2.6652486915791254e-06,4.7384201416619685e-06,3.5537405396740974e-06,1.500566321253565e-07,8.432281095893794e-07,3.5537405396740974e-06,4.7474651455224204e-08,3.5537405396740974e-06,2.6652486915791254e-06,6.318023592626571e-06,6.336118011097954e-09,1.1296899377085736e-11,4.751983177356673e-09,3.570694667313889e-11,8.42420045159261e-06,2.6652486915791254e-06,8.456384362775845e-10,3.5639083851319007e-09,4.751983177356673e-09,1.4976922194352454e-05,4.751983177356673e-09,3.557133760522609e-07,4.7384201416619685e-06,6.336118011097954e-09,4.751983177356673e-09,1.4976922194352454e-05,8.432281095893794e-07,1.1243287528229769e-06,6.311658370697072e-05,3.5639083851319007e-09,2.677961637410313e-11,1.1286159131746151e-10,8.432281095893794e-07,1.5019945075512058e-08,1.9969572552347172e-05,1.1232484713080014e-05,1.1243287528229769e-06,8.42420045159261e-06,3.560519974396102e-08,1.4976922194352454e-05,8.448344602871815e-09,0.00011220672558762844,3.550248479350091e-05,2.677961637410313e-11,8.42420045159261e-06,3.5639083851319007e-09,1.1232484713080014e-05,4.7384201416619685e-06,6.318023592626571e-06,1.1264709141086165e-08,1.4991376729384742e-06,6.336118011097954e-09,4.7384201416619685e-06,1.1243287528229769e-06,8.432281095893794e-07,4.751983177356673e-09,2.677961637410313e-11,3.550248479350091e-05,6.35422418491181e-12,4.770097549996675e-13,8.4805493909702e-13,6.35422418491181e-12,1.1286159131746151e-10,2.0122510825990464e-13,1.1307649842129329e-12,3.570694667313889e-11,8.432281095893794e-07,8.4805493909702e-13,8.464431715704131e-11,2.0141660004869664e-14,2.004609595332771e-09,3.57749385922517e-13,6.336118011097954e-09,2.683060918282969e-13,4.7610317597661644e-11,1.5062866397951075e-11,2.0084267067338393e-11,4.756505326007187e-10,1.5034238704500349e-09,1.500566321253565e-07,2.667791420002213e-07,1.1243287528229769e-06,8.432281095893794e-07,4.742949587530005e-07,1.4991376729384742e-06,2.0027037268645274e-08,8.432281095893794e-07,6.324071967785796e-07,4.742949587530005e-07,2.667791420002213e-07,1.9988935406767335e-06,4.7474651455224204e-08,3.557133760522609e-07,1.9969572552347172e-05,1.5019945075512058e-08,8.432281095893794e-07,2.004609595332771e-09,1.1243287528229769e-06,6.330093744971093e-08,8.44031190971643e-08,1.500566321253565e-07,4.7474651455224204e-08,6.342147685321855e-10,1.127542908713118e-09,2.0027037268645274e-08,1.9988935406767335e-06,2.667791420002213e-07,1.5019945075512058e-08,4.742949587530005e-07,8.448344602871815e-09,6.330093744971093e-08,1.5019945075512058e-08,2.0007993461967956e-07,2.0007993461967956e-07,1.9988935406767335e-06,1.500566321253565e-07,3.5639083851319007e-09,6.330093744971093e-08,2.0027037268645274e-08,3.560519974396102e-08,1.5019945075512058e-08,2.0007993461967956e-07,2.667791420002213e-07,6.330093744971093e-08,2.670330810624187e-08,3.560519974396102e-08,6.324071967785796e-07,6.330093744971093e-08,6.330093744971093e-08,8.464431715704131e-11,2.667791420002213e-07,8.432281095893794e-07,1.1243287528229769e-06,4.742949587530005e-07,4.742949587530005e-07,3.5639083851319007e-09,8.44031190971643e-08,3.557133760522609e-07,4.742949587530005e-07,3.557133760522609e-07,2.004609595332771e-09,1.5019945075512058e-08,8.44031190971643e-08,1.1253998361201744e-07,2.670330810624187e-08,1.5019945075512058e-08,3.560519974396102e-08,8.44031190971643e-08,6.336118011097954e-09,4.7474651455224204e-08,3.560519974396102e-08,1.500566321253565e-07,4.7474651455224204e-08,8.432281095893794e-07,3.560519974396102e-08,1.5019945075512058e-08,2.0065172449737883e-10,2.004609595332771e-09,4.756505326007187e-10,6.324071967785796e-07,4.751983177356673e-09,3.570694667313889e-11,6.330093744971093e-08,1.127542908713118e-09,2.667791420002213e-07,1.5019945075512058e-08,2.004609595332771e-09,4.756505326007187e-10,1.5034238704500349e-09,8.472486720936424e-12,1.1264709141086165e-08,3.5639083851319007e-09,3.557133760522609e-07,6.348183065023614e-11,2.675415631799144e-10,2.0007993461967956e-07,2.677961637410313e-11,4.7610317597661644e-11,2.0027037268645274e-08,1.1253998361201744e-07,6.336118011097954e-09,4.742949587530005e-07,3.560519974396102e-08,3.560519974396102e-08,2.004609595332771e-09,8.44031190971643e-08,6.330093744971093e-08,3.5639083851319007e-09,3.560519974396102e-08,2.670330810624187e-08,8.456384362775845e-10,4.7474651455224204e-08,2.675415631799144e-10,2.004609595332771e-09,2.004609595332771e-09,2.672872040954052e-09,2.672872040954052e-09,2.0027037268645274e-08,1.5034238704500349e-09,1.1264709141086165e-08,1.1264709141086165e-08,2.672872040954052e-09,3.5639083851319007e-09,2.672872040954052e-09,2.0027037268645274e-08,1.1296899377085736e-11,2.672872040954052e-09,1.127542908713118e-09,2.0065172449737883e-10,2.0027037268645274e-08,1.1253998361201744e-07,6.324071967785796e-07,4.7474651455224204e-08,2.0065172449737883e-10,1.1253998361201744e-07,1.127542908713118e-09,3.560519974396102e-08,3.56729991762872e-10,2.677961637410313e-11,1.1264709141086165e-08,8.50478346615514e-16,2.0027037268645274e-08,4.7474651455224204e-08,6.342147685321855e-10,2.0065172449737883e-10,8.464431715704131e-11,1.5034238704500349e-09,6.348183065023614e-11,2.0027037268645274e-08,8.464431715704131e-11,8.44031190971643e-08,6.330093744971093e-08,8.44031190971643e-08,2.675415631799144e-10,8.448344602871815e-09,2.0027037268645274e-08,1.1253998361201744e-07,4.770097549996675e-13,2.004609595332771e-09,8.456384362775845e-10,8.456384362775845e-10,4.742949587530005e-07,4.7474651455224204e-08,8.448344602871815e-09,1.5034238704500349e-09,6.342147685321855e-10,8.448344602871815e-09,4.7474651455224204e-08,1.500566321253565e-07,1.5019945075512058e-08,2.0007993461967956e-07,8.456384362775845e-10,1.1253998361201744e-07,8.448344602871815e-09,4.751983177356673e-09,1.1264709141086165e-08,2.675415631799144e-10,2.667791420002213e-07,2.667791420002213e-07,8.44031190971643e-08,6.336118011097954e-09,4.742949587530005e-07,1.127542908713118e-09,2.0007993461967956e-07,4.742949587530005e-07,1.127542908713118e-09,4.751983177356673e-09,2.0007993461967956e-07,1.1253998361201744e-07,2.0007993461967956e-07,2.6652486915791254e-06,4.756505326007187e-10,4.7474651455224204e-08,6.330093744971093e-08,4.742949587530005e-07,2.0007993461967956e-07,2.0027037268645274e-08,1.127542908713118e-09,1.1264709141086165e-08,3.5639083851319007e-09,2.0027037268645274e-08,2.0007993461967956e-07,1.5019945075512058e-08,1.1253998361201744e-07,2.0007993461967956e-07,3.560519974396102e-08,2.0007993461967956e-07,1.1253998361201744e-07,8.448344602871815e-09,2.0065172449737883e-10,2.667791420002213e-07,4.7474651455224204e-08,6.330093744971093e-08,8.456384362775845e-10,4.756505326007187e-10,1.5019945075512058e-08,1.1264709141086165e-08,4.742949587530005e-07,6.324071967785796e-07,3.560519974396102e-08,2.0007993461967956e-07,2.670330810624187e-08,1.1264709141086165e-08,1.1253998361201744e-07,1.500566321253565e-07,6.330093744971093e-08,2.0027037268645274e-08,2.667791420002213e-07,1.1264709141086165e-08,1.5034238704500349e-09,1.1253998361201744e-07,1.5034238704500349e-09,1.5019945075512058e-08,2.670330810624187e-08,3.560519974396102e-08,1.4991376729384742e-06,8.448344602871815e-09,2.0027037268645274e-08,2.667791420002213e-07,3.5639083851319007e-09,6.330093744971093e-08,1.1264709141086165e-08,6.330093744971093e-08,8.44031190971643e-08,4.751983177356673e-09,1.1264709141086165e-08,2.672872040954052e-09,2.667791420002213e-07,3.570694667313889e-11,2.672872040954052e-09,3.560519974396102e-08,8.448344602871815e-09,8.448344602871815e-09,4.7474651455224204e-08,6.342147685321855e-10,8.448344602871815e-09,2.0027037268645274e-08,3.56729991762872e-10,6.348183065023614e-11,2.667791420002213e-07,3.56729991762872e-10,2.004609595332771e-09,2.0027037268645274e-08,3.557133760522609e-07,1.1253998361201744e-07,4.751983177356673e-09,3.5639083851319007e-09,3.5639083851319007e-09,1.127542908713118e-09,6.342147685321855e-10,8.456384362775845e-10,6.336118011097954e-09,1.1264709141086165e-08,1.127542908713118e-09,8.448344602871815e-09,4.7474651455224204e-08,8.448344602871815e-09,8.448344602871815e-09,2.0027037268645274e-08,8.44031190971643e-08,6.342147685321855e-10,4.7474651455224204e-08,2.004609595332771e-09,4.751983177356673e-09,2.667791420002213e-07,8.44031190971643e-08,4.751983177356673e-09,1.5019945075512058e-08,8.44031190971643e-08,6.324071967785796e-07,4.742949587530005e-07,2.0007993461967956e-07,2.004609595332771e-09,4.7474651455224204e-08,1.5019945075512058e-08,4.7610317597661644e-11,2.667791420002213e-07,4.7474651455224204e-08,3.570694667313889e-11,1.5019945075512058e-08,1.5034238704500349e-09,4.7610317597661644e-11,4.751983177356673e-09,3.557133760522609e-07,1.127542908713118e-09,2.0007993461967956e-07,4.751983177356673e-09\\n'"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ar_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f1cb04f1-23e8-48eb-9447-b50789d8c152",
   "metadata": {},
   "outputs": [],
   "source": [
    "i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "74483ac8-23db-45bc-838b-b1996f664fe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6240308"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c86f4446-39e3-47bf-b6ed-b1791ee1026a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_numbers = ref_numbers.rstrip().split(',')[3:]\n",
    "aa_probs = aa_probs.rstrip().split(',')[3:]\n",
    "ar_probs = ar_probs.rstrip().split(',')[3:]\n",
    "rr_probs = rr_probs.rstrip().split(',')[3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "afc69bf2-d16a-4178-9fb7-f01e82195cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_numbers = inf.readline().rstrip().split(',')[3:]\n",
    "aa_probs = inf_aa.readline().rstrip().split(',')[3:]\n",
    "ar_probs = inf_ar.readline().rstrip().split(',')[3:]\n",
    "rr_probs = inf_rr.readline().rstrip().split(',')[3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ca60c4e6-d034-44b4-a132-bdd08276cc92",
   "metadata": {},
   "outputs": [],
   "source": [
    "j = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c42b8059-21aa-405f-9e80-1f12534d8339",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'BS16_0134'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_bsids.rstrip().split(',')[3:][j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "f32f3468-3661-4d20-a800-e3838d861342",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2'"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_nr = ref_numbers[j]\n",
    "ref_nr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "45429fe5-e62d-4515-a2f6-5cffa9bf6b08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.95"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prob_threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "772edaa6-58de-4c83-9032-853a3d6a6a85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(ref_nr == '2') and (float(rr_probs[j]) >= prob_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a820c1e1-ee60-4376-9ea6-d08a1986dc5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ar_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "a66aa6b3-d577-4431-b81f-71433c47e39f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.9895146463136785'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ar_probs[j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "0b16d6d1-b462-485e-b162-1eee0cff0f39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7023687"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "dec8c438-6b93-41df-a669-8d63904c8aa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHR                 14\n",
       "SNP_ID      14:9770952\n",
       "GEN_DIST             0\n",
       "BP_POS         9770952\n",
       "ALT                  T\n",
       "REF                  C\n",
       "Name: 6240306, dtype: object"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_like.iloc[6240306, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "1945ffe9-b9f9-47ce-ad74-e06e75a4887b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim.iloc[i, 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "0982b335-2256-423b-be35-799af1bf4fc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'T'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim.iloc[i, 4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "60fd8f1d-495e-4997-9c19-588af4c5129b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ped_file = '{}{}_{}.ped'.format(plink_path, prefix, pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "fa7e64e8-c490-4f4c-99d9-04dd63b0d486",
   "metadata": {},
   "outputs": [],
   "source": [
    "inf_ped = open(ped_file, 'r')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "3905f24b-64c6-4f0e-a6e9-3183961db502",
   "metadata": {},
   "outputs": [],
   "source": [
    "for l in inf_ped:\n",
    "    if l.split()[1] == 'BS16_0134': break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "377e5932-cbdf-412e-b988-b9c6b8a889a1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "l_split = l.split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "5a547e16-bde0-4495-968a-fec9d52fc501",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6764372.0"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(l_split[6:]) / 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "e5ff1630-7f88-4e20-a01d-85e5f8b59af1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['C',\n",
       " '0',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'T',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'T',\n",
       " 'T',\n",
       " 'A',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'T',\n",
       " 'T',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'T',\n",
       " 'T',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " '0',\n",
       " '0',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'A',\n",
       " 'G',\n",
       " 'A',\n",
       " 'A',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " '0',\n",
       " '0',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'A',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'T',\n",
       " 'T',\n",
       " '0',\n",
       " '0',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'A',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'C',\n",
       " 'T',\n",
       " 'T',\n",
       " 'T',\n",
       " 'T',\n",
       " 'G',\n",
       " 'G',\n",
       " '0',\n",
       " '0',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'G',\n",
       " 'C',\n",
       " 'C',\n",
       " 'A',\n",
       " 'G',\n",
       " 'C',\n",
       " 'T',\n",
       " ...]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_split[(6 + 2*6240306):]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "4655da57-3d9d-4629-8946-8e93f75d4166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1  2        3  4  5\n",
       "0         1      1:5671  0     5671  0  T\n",
       "1         1      1:5698  0     5698  C  T\n",
       "2         1      1:6621  0     6621  G  A\n",
       "3         1      1:6717  0     6717  0  C\n",
       "4         1      1:6823  0     6823  C  G\n",
       "...      ..         ... ..      ... .. ..\n",
       "6764367  16  16:7225908  0  7225908  C  T\n",
       "6764368  16  16:7225911  0  7225911  T  C\n",
       "6764369  16  16:7226186  0  7226186  A  G\n",
       "6764370  16  16:7226346  0  7226346  0  G\n",
       "6764371  16  16:7227034  0  7227034  0  G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "96fe5910-5dd1-46d8-8607-bbfb6b5cc520",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0            14\n",
       "1    14:9770952\n",
       "2             0\n",
       "3       9770952\n",
       "4             0\n",
       "5             C\n",
       "Name: 6240306, dtype: object"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim.loc[6240306]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "dbf0f95c-90c0-4a7b-a9ae-62c1c4a059cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_ped_pop(\n",
    "    plink_path, prefix, pop, chunk_bs_ids, batch_j, chunk_size):\n",
    "\n",
    "    # bim like because not output of --make-bed (which filters some alternative alleles)\n",
    "    bim_like = pd.read_csv('{}{}.bim_like'.format(plink_path, prefix), sep='\\t', header=None)\n",
    "    \n",
    "    bgs_file = '{}{}_{}.bgs'.format(plink_path, prefix, pop)\n",
    "    aa_prob_file = '{}{}_{}_AA.prob'.format(plink_path, prefix, pop)\n",
    "    ar_prob_file = '{}{}_{}_AR.prob'.format(plink_path, prefix, pop)\n",
    "    rr_prob_file = '{}{}_{}_RR.prob'.format(plink_path, prefix, pop)\n",
    "    snp_nr = len(bim_like)\n",
    "    # snp_nr = 10\n",
    "    \n",
    "    start_j = batch_j * chunk_size\n",
    "    ped_mat = np.full((len(chunk_bs_ids), 2 * snp_nr), '', dtype='U1')\n",
    "    \n",
    "    inf = open(bgs_file, 'r')\n",
    "    inf_aa = open(aa_prob_file, 'r')\n",
    "    inf_ar = open(ar_prob_file, 'r')\n",
    "    inf_rr = open(rr_prob_file, 'r')\n",
    "\n",
    "    # check we have the same bs ids in header than in batch\n",
    "    header_bsids = inf.readline()\n",
    "    assert header_bsids.rstrip().split(',')[3:][start_j: start_j + len(chunk_bs_ids)] == chunk_bs_ids, 'bs_id mismatch'\n",
    "    inf_aa.readline()\n",
    "    inf_ar.readline()\n",
    "    inf_rr.readline()\n",
    "\n",
    "    # iterate snps and colonies\n",
    "    for i in tqdm(range(snp_nr)):\n",
    "        ref_numbers = inf.readline().rstrip().split(',')[3:]\n",
    "        aa_probs = inf_aa.readline().rstrip().split(',')[3:]\n",
    "        ar_probs = inf_ar.readline().rstrip().split(',')[3:]\n",
    "        rr_probs = inf_rr.readline().rstrip().split(',')[3:]\n",
    "\n",
    "        for j in range(start_j, start_j + len(chunk_bs_ids)):\n",
    "            ref_nr = ref_numbers[j]\n",
    "\n",
    "            # homozygote reference\n",
    "            if (ref_nr == '2') and (float(rr_probs[j]) >= prob_threshold):\n",
    "                ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 5]\n",
    "                ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 5]\n",
    "            # heterozygote\n",
    "            elif (ref_nr == '1') and (float(ar_probs[j]) >= prob_threshold):\n",
    "                ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 5]\n",
    "                ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 4]\n",
    "            # homozygote alternative\n",
    "            elif (ref_nr == '0') and (float(aa_probs[j]) >= prob_threshold):\n",
    "                ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 4]\n",
    "                ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 4]\n",
    "            # below prob_threshold\n",
    "            elif ref_nr in {'0', '1', '2'}:\n",
    "                ped_mat[j - start_j, 2 * i] = '0'\n",
    "                ped_mat[j - start_j, 2 * i + 1] = '0'\n",
    "            else: \n",
    "                print('ERROR')\n",
    "            \n",
    "    inf.close()\n",
    "    inf_aa.close()\n",
    "    inf_ar.close()\n",
    "    inf_rr.close()\n",
    "\n",
    "    ped_df = pd.DataFrame({\n",
    "        'FID': ['0'] * len(chunk_bs_ids), \n",
    "        'IID': chunk_bs_ids,\n",
    "        'PID': ['0'] * len(chunk_bs_ids),\n",
    "        'MID': ['0'] * len(chunk_bs_ids),\n",
    "        'SEX': ['0'] * len(chunk_bs_ids), \n",
    "        'PHENOTYPE': ['-9'] * len(chunk_bs_ids)\n",
    "    })\n",
    "    ped_df = ped_df.join(pd.DataFrame(ped_mat))\n",
    "    ped_df.to_csv('{}{}_{}_{}.ped'.format(plink_path, prefix, pop, batch_j), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "518c3447-f515-461b-97bf-6678606f04fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_j = 0\n",
    "chunk_bs_ids = batch_bs_ids[batch_j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "7930bebb-a396-481d-bbd9-902208b293c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10/10 [00:00<00:00, 1696.11it/s]\n"
     ]
    }
   ],
   "source": [
    "write_ped_pop(\n",
    "    plink_path, prefix, pop, chunk_bs_ids, batch_j, chunk_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc66cbaf-e400-48c1-a115-857e96ec5bad",
   "metadata": {},
   "source": [
    "prod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "39a83de7-0970-4c8b-a0fe-37d6084aa2d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_write_ped_script(script_fn, array_str):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --array=ARRAY_STRING\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=100g\n",
    "#SBATCH --time=6:00:00\n",
    "#SBATCH --job-name=write_ped\n",
    "#SBATCH --output=%x_%A_%a.out\n",
    "#SBATCH --error=%x_%A_%a.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "# file listing batches of BeeStrong ids (bs_id) for parallel computing\n",
    "batch_file=$1\n",
    "chunk_size=$2\n",
    "plink_path=$3\n",
    "prefix=$4\n",
    "pop=$5\n",
    "prob_threshold=$6\n",
    "    \n",
    "batch_id=$SLURM_ARRAY_TASK_ID\n",
    "\n",
    "# getting bs_ids\n",
    "chunk_bs_ids=$(sed -n ${batch_id}'{p;q}' ${batch_file})\n",
    "    \n",
    "python ~/scripts/plan-b-omics/bin/write_ped.py --chunk_bs_ids ${chunk_bs_ids} --batch_id ${batch_id} --chunk_size ${chunk_size} --plink_path ${plink_path} --prefix ${prefix} --pop ${pop} --prob_threshold ${prob_threshold}\n",
    "\n",
    "echo DONE\"\"\".replace(\"ARRAY_STRING\", array_str)\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "d1d8ad32-38d4-4bdd-82f8-7f35734ddb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "chunk_size = 25 # should match the\n",
    "plink_path = '{}data/queen_genotypes/hom_pop/'.format(working_path)\n",
    "prefix = 'queen'\n",
    "# populations = ['Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid']\n",
    "\n",
    "prob_threshold = 0.95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "7d2f6a54-ffdf-463d-a579-c5c4d01129af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n",
      "645\n",
      "405\n",
      "372\n"
     ]
    }
   ],
   "source": [
    "# write batch files\n",
    "for pop in populations:\n",
    "    batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)\n",
    "    bs_ids = pheno_df[pheno_df['Genetic group'] == pop]['Colony'].to_list()\n",
    "    print(len(bs_ids))\n",
    "    \n",
    "    batch_bs_ids = []\n",
    "    with open(batch_file, \"w\") as outf:\n",
    "        for i in range(0, len(bs_ids), chunk_size):\n",
    "            chunk = bs_ids[i:i+chunk_size]\n",
    "            batch_bs_ids.append(chunk)\n",
    "            outf.write(\",\".join(chunk) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "cfd82126-2442-4fb7-b99f-3c03f7c6e553",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop = 'Ligustica_Carnica'\n",
    "# pop = 'Caucasica'\n",
    "# pop = 'Mellifera'\n",
    "# pop = 'hybrid'\n",
    "\n",
    "batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "be1b9a64-fed0-4895-8a91-c76946948663",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "895eebd7-e16c-4a03-a6b2-d75ace9c61cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_str = '1-{}'.format(len(batch_bs_ids)) if len(batch_bs_ids) > 1 else '1'\n",
    "script_fn = '{}write_ped.run'.format(script_path)\n",
    "write_write_ped_script(script_fn, array_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "df9e5aaa-9cfd-46e5-bbb4-46535d13aad0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 1380901\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$script_fn\" \"$batch_file\" \"$chunk_size\" \"$plink_path\" \"$prefix\" \"$pop\" \"$prob_threshold\" \n",
    "cd $4\n",
    "sbatch $1 $2 $3 $4 $5 $6 $7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0989631e-9e1b-4571-8f53-c845276a430f",
   "metadata": {},
   "source": [
    "cat the ped files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d34c9503-9bea-49c3-a46d-48b987b9bb74",
   "metadata": {},
   "source": [
    "### .map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fc0666-f437-4124-b391-ca3b929179ae",
   "metadata": {},
   "source": [
    ".map file is also needed to convert ped to bed (only map + bed)\n",
    "\n",
    "CHR   SNP_ID   GEN_DIST   BP_POS\n",
    "\n",
    "basically bim wihout REF and ALT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab5ab449-661f-462f-b1ec-929f47713227",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'queen'\n",
    "plink_path = '{}data/queen_genotypes/hom_pop/'.format(working_path)\n",
    "populations = ['Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "d004a53f-cd6d-466c-8286-b35055261521",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_like = pd.read_csv('{}{}.bim_like'.format(plink_path, prefix), sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "57c70009-fdfa-4eda-8eda-fdefefe17794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1  2        3  4  5\n",
       "0         1      1:5671  0     5671  G  T\n",
       "1         1      1:5698  0     5698  C  T\n",
       "2         1      1:6621  0     6621  G  A\n",
       "3         1      1:6717  0     6717  T  C\n",
       "4         1      1:6823  0     6823  C  G\n",
       "...      ..         ... ..      ... .. ..\n",
       "6764367  16  16:7225908  0  7225908  C  T\n",
       "6764368  16  16:7225911  0  7225911  T  C\n",
       "6764369  16  16:7226186  0  7226186  A  G\n",
       "6764370  16  16:7226346  0  7226346  A  G\n",
       "6764371  16  16:7227034  0  7227034  A  G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "889979b2-8304-451b-b1bd-f378766bf7c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caucasica\n",
      "Ligustica_Carnica\n",
      "Mellifera\n",
      "hybrid\n"
     ]
    }
   ],
   "source": [
    "for pop in populations:\n",
    "    print(pop)\n",
    "    bim[[0, 1, 2, 3]].to_csv('{}{}_{}.map'.format(plink_path, prefix, pop), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad7fbd2d-332e-486c-a100-ef37b876cb94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ea91ad53-f6f8-4fa9-acbb-ee97b23852e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1  2        3  4  5\n",
       "0         1      1:5671  0     5671  0  T\n",
       "1         1      1:5698  0     5698  C  T\n",
       "2         1      1:6621  0     6621  G  A\n",
       "3         1      1:6717  0     6717  0  C\n",
       "4         1      1:6823  0     6823  C  G\n",
       "...      ..         ... ..      ... .. ..\n",
       "6764367  16  16:7225908  0  7225908  C  T\n",
       "6764368  16  16:7225911  0  7225911  T  C\n",
       "6764369  16  16:7226186  0  7226186  A  G\n",
       "6764370  16  16:7226346  0  7226346  0  G\n",
       "6764371  16  16:7227034  0  7227034  0  G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "15fd1362-57a6-4a81-b187-e23f256923f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'queen'\n",
    "snp_nr = 6764372\n",
    "bs_ids = pheno_df['Colony'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b4d921ee-f6be-4e71-bb3c-de857e907001",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GEN_DIST</th>\n",
       "      <th>BP_POS</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>NC_037638.1_5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>NC_037638.1_5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>NC_037638.1_6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>NC_037638.1_6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_037638.1</td>\n",
       "      <td>NC_037638.1_6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>NC_037653.1</td>\n",
       "      <td>NC_037653.1_7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>NC_037653.1</td>\n",
       "      <td>NC_037653.1_7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>NC_037653.1</td>\n",
       "      <td>NC_037653.1_7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>NC_037653.1</td>\n",
       "      <td>NC_037653.1_7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>NC_037653.1</td>\n",
       "      <td>NC_037653.1_7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 CHR               SNP_ID  GEN_DIST   BP_POS ALT REF\n",
       "0        NC_037638.1     NC_037638.1_5671         0     5671   G   T\n",
       "1        NC_037638.1     NC_037638.1_5698         0     5698   C   T\n",
       "2        NC_037638.1     NC_037638.1_6621         0     6621   G   A\n",
       "3        NC_037638.1     NC_037638.1_6717         0     6717   T   C\n",
       "4        NC_037638.1     NC_037638.1_6823         0     6823   C   G\n",
       "...              ...                  ...       ...      ...  ..  ..\n",
       "6764367  NC_037653.1  NC_037653.1_7225908         0  7225908   C   T\n",
       "6764368  NC_037653.1  NC_037653.1_7225911         0  7225911   T   C\n",
       "6764369  NC_037653.1  NC_037653.1_7226186         0  7226186   A   G\n",
       "6764370  NC_037653.1  NC_037653.1_7226346         0  7226346   A   G\n",
       "6764371  NC_037653.1  NC_037653.1_7227034         0  7227034   A   G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read bim file\n",
    "bim_file = '{}{}.bim'.format(allele_count_path, prefix)\n",
    "bim_df = pd.read_csv(bim_file, sep='\\t')\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b88d885e-b756-4f3c-9023-a41225861584",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write map file\n",
    "bim_df[['CHR', 'SNP_ID', 'GEN_DIST', 'BP_POS']][:snp_nr].to_csv('{}{}.map'.format(plink_path, prefix), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c64f6b0f-554a-408c-995b-813588ec124f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and fam file\n",
    "fam_df = pd.DataFrame({\n",
    "    'FID': bs_ids, \n",
    "    'IID': bs_ids,\n",
    "    'PID ': ['0'] * len(bs_ids),\n",
    "    'MID': ['0'] * len(bs_ids),\n",
    "    'SEX': ['0'] * len(bs_ids), \n",
    "    'PHENOTYPE': ['-9'] * len(bs_ids)\n",
    "})\n",
    "fam_df.to_csv('{}{}.fam'.format(plink_path, prefix), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8b5f7f4b-f1d8-45d6-a13a-152feacec9db",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1  2        3  4  5\n",
       "0         1      1:5671  0     5671  G  T\n",
       "1         1      1:5698  0     5698  C  T\n",
       "2         1      1:6621  0     6621  G  A\n",
       "3         1      1:6717  0     6717  T  C\n",
       "4         1      1:6823  0     6823  C  G\n",
       "...      ..         ... ..      ... .. ..\n",
       "6764367  16  16:7225908  0  7225908  C  T\n",
       "6764368  16  16:7225911  0  7225911  T  C\n",
       "6764369  16  16:7226186  0  7226186  A  G\n",
       "6764370  16  16:7226346  0  7226346  A  G\n",
       "6764371  16  16:7227034  0  7227034  A  G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_file = '{}{}.bim'.format(plink_path, prefix)\n",
    "bim_df.to_csv(bim_file, header=False, index=False, sep='\\t')\n",
    "bim_df = pd.read_csv(bim_file, sep='\\t', header=None)\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd7cd40-befc-4bfe-9101-07934e49a434",
   "metadata": {},
   "source": [
    "convert to bed in CLI. Not sure how long it will take for convertion to bed...\n",
    "\n",
    "ml PLINK\n",
    "\n",
    "plink --file queen --make-bed --out queen --allow-extra-chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "2e6f20d2-12d5-4edb-a4f9-b7439d817025",
   "metadata": {},
   "outputs": [],
   "source": [
    "# recode bs_ids of suiss colonies (_ instead of -)\n",
    "with open('{}suisse/suisse_original.fam'.format(plink_path), 'r') as inf:\n",
    "    with open('{}suisse/suisse.fam'.format(plink_path), 'w') as outf:\n",
    "        for l in inf:\n",
    "            outf.write(l.replace('BS18-', 'BS18_'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff623da-bdac-4952-b70b-c30900dcccbc",
   "metadata": {},
   "source": [
    "recode queen_phased (for comparison with queen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "455c93e7-651a-4ce0-974c-18f3420d0acb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>snp1</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>snp2</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>snp3</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>snp4</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>snp5</td>\n",
       "      <td>0</td>\n",
       "      <td>6832</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944678</th>\n",
       "      <td>16</td>\n",
       "      <td>snp5944679</td>\n",
       "      <td>0</td>\n",
       "      <td>7225891</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944679</th>\n",
       "      <td>16</td>\n",
       "      <td>snp5944680</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944680</th>\n",
       "      <td>16</td>\n",
       "      <td>snp5944681</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944681</th>\n",
       "      <td>16</td>\n",
       "      <td>snp5944682</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944682</th>\n",
       "      <td>16</td>\n",
       "      <td>snp5944683</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5944683 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1  2        3  4  5\n",
       "0         1        snp1  0     5671  .  T\n",
       "1         1        snp2  0     5698  C  T\n",
       "2         1        snp3  0     6621  .  A\n",
       "3         1        snp4  0     6823  .  G\n",
       "4         1        snp5  0     6832  .  G\n",
       "...      ..         ... ..      ... .. ..\n",
       "5944678  16  snp5944679  0  7225891  A  G\n",
       "5944679  16  snp5944680  0  7225908  .  T\n",
       "5944680  16  snp5944681  0  7225911  T  C\n",
       "5944681  16  snp5944682  0  7226346  .  G\n",
       "5944682  16  snp5944683  0  7227034  .  G\n",
       "\n",
       "[5944683 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_file = '{}suisse_queen_phased/queen_phased.bim'.format(plink_path)\n",
    "bim_df = pd.read_csv(bim_file, sep='\\t', header=None)\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a7209ff7-5cdf-4918-a744-7f401f5b52a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5944683"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bim_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f3ea33d9-1d3f-44af-bc44-92a70c9e5618",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "5944683it [03:23, 29222.47it/s]\n"
     ]
    }
   ],
   "source": [
    "new_snp_ids = [''] * len(bim_df)\n",
    "i= 0\n",
    "for row in tqdm(bim_df.iterrows()):\n",
    "    new_snp_ids[i] = '{}:{}'.format(row[1][0], row[1][3])\n",
    "    i +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c684bfd-fe8e-41bd-93a3-946110bc667b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5944683"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(new_snp_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3c6d2e97-4f59-441c-b8e3-695660e99b49",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_df[1] = new_snp_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "01388c87-acc3-4d7a-be8f-6245586ea917",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6832</td>\n",
       "      <td>0</td>\n",
       "      <td>6832</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944678</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225891</td>\n",
       "      <td>0</td>\n",
       "      <td>7225891</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944679</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944680</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944681</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5944682</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5944683 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1  2        3  4  5\n",
       "0         1      1:5671  0     5671  .  T\n",
       "1         1      1:5698  0     5698  C  T\n",
       "2         1      1:6621  0     6621  .  A\n",
       "3         1      1:6823  0     6823  .  G\n",
       "4         1      1:6832  0     6832  .  G\n",
       "...      ..         ... ..      ... .. ..\n",
       "5944678  16  16:7225891  0  7225891  A  G\n",
       "5944679  16  16:7225908  0  7225908  .  T\n",
       "5944680  16  16:7225911  0  7225911  T  C\n",
       "5944681  16  16:7226346  0  7226346  .  G\n",
       "5944682  16  16:7227034  0  7227034  .  G\n",
       "\n",
       "[5944683 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a88c575b-ba22-4de2-ae04-7eaef3c14102",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_df.to_csv(bim_file, header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "65cdf84e-d4b6-4c3d-85e9-867a0238cffe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/plink/suisse_queen_phased/queen_phased.bim'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5a564d0b-7a3f-419e-aeb5-3e0a72103db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_df[[0, 1, 2, 3]].to_csv('/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/plink/suisse_queen_phased/queen_phased.map', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec4c5a6-a121-483c-abeb-62fd76bce5ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# recode snp names of queen_phased\n",
    "\n",
    "bim_df[['CHR', 'SNP_ID', 'GEN_DIST', 'BP_POS']].to_csv('{}{}.map'.format(plink_path, prefix), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "96518382-c824-4518-b596-573a37c8819a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_304589/947095543.py:4: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  tfam_df = pd.read_csv('{}suisse_queen_phased/suisse.tfam'.format(plink_path), sep='\\s', header=None)\n"
     ]
    }
   ],
   "source": [
    "# recode tfam \n",
    "\n",
    "\n",
    "tfam_df = pd.read_csv('{}suisse_queen_phased/suisse.tfam'.format(plink_path), sep='\\s', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "21cc0fa2-bcf6-4280-88bc-5e3e086e7455",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfam_df[0] = len(tfam_df) * [0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "48e694ba-13b4-485a-89f7-2fce0a246b77",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3354437/187792405.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  fam = pd.read_csv('/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/suisse/queen_suisse/queen_suisse.fam', sep='\\s', header=None)\n"
     ]
    }
   ],
   "source": [
    "fam = pd.read_csv('/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/suisse/queen_suisse/queen_suisse.fam', sep='\\s', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "c50ec0f4-114f-4069-a4c5-5cec48111680",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam[0] = ['0'] * len(fam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "097e2162-3415-47a9-83cf-1bc2c1bfb2f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam.to_csv('/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/suisse/queen_suisse_suiss_conc/queen_suisse.fam', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3010fc2d-e483-4294-84b9-f38039a79959",
   "metadata": {},
   "outputs": [],
   "source": [
    "def recode_suisse_bsids(old_id):\n",
    "    return '{}_{}'.format(old_id, old_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4555a0b3-5d8b-4d50-82bd-fd48f2dee6dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfam_df[1] = tfam_df[1].map(recode_suisse_bsids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0747ec24-6524-4ea3-8a56-d3d16b914a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfam_df.to_csv('{}suisse_queen_phased/suisse.tfam'.format(plink_path), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d40b2ea6-8e77-488d-9d19-ce429fa12b57",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43023e95-6bbe-44b3-949e-6da09385e604",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "41b26baa-caeb-48f0-8f42-ad6861dd824d",
   "metadata": {},
   "source": [
    "## GREML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "d1b6f7fe-e606-4bd1-b61b-3aa12d346db8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_path = '/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "fff10525-496c-4313-9ff4-e33b8f091aa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3354437/1480174743.py:3: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  fam_df = pd.read_csv('{}{}.fam'.format(out_path, prefix), header=None, sep='\\s')\n"
     ]
    }
   ],
   "source": [
    "prefix = 'hom_pop'\n",
    "out_path = plink_path\n",
    "fam_df = pd.read_csv('{}{}.fam'.format(out_path, prefix), header=None, sep='\\s') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "3b997aa2-c54d-460d-b993-87594edaf984",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1  2  3  4  5\n",
       "0     0  BS16_0060  0  0  0 -9\n",
       "1     0  BS16_0061  0  0  0 -9\n",
       "2     0  BS16_0162  0  0  0 -9\n",
       "3     0  BS16_0163  0  0  0 -9\n",
       "4     0  BS16_0164  0  0  0 -9\n",
       "...  ..        ... .. .. .. ..\n",
       "1437  0  BS17_0506  0  0  0 -9\n",
       "1438  0  BS17_0511  0  0  0 -9\n",
       "1439  0  BS17_0512  0  0  0 -9\n",
       "1440  0  BS17_0513  0  0  0 -9\n",
       "1441  0  BS17_0517  0  0  0 -9\n",
       "\n",
       "[1442 rows x 6 columns]"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fam_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "5d4fe18c-30ea-426e-ae82-6690d9eb171d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Colony</th>\n",
       "      <th>date</th>\n",
       "      <th>Genetic group</th>\n",
       "      <th>percent_Ligustica_Carnica</th>\n",
       "      <th>percent_Mellifera</th>\n",
       "      <th>percent_Caucasica</th>\n",
       "      <th>bee_weigth</th>\n",
       "      <th>nbr_pho_varroa</th>\n",
       "      <th>nbr_pho_varroa_100bee</th>\n",
       "      <th>v_pho</th>\n",
       "      <th>...</th>\n",
       "      <th>Day of Year</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>v_meta</th>\n",
       "      <th>log_v_meta</th>\n",
       "      <th>log_v_pho</th>\n",
       "      <th>log_v_mito</th>\n",
       "      <th>Varroa group</th>\n",
       "      <th>Recapping group</th>\n",
       "      <th>DMR group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.118259</td>\n",
       "      <td>1.369552e-04</td>\n",
       "      <td>0.881604</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>188</td>\n",
       "      <td>47.00</td>\n",
       "      <td>2.583</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-8.156349</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.109495</td>\n",
       "      <td>9.200239e-05</td>\n",
       "      <td>0.890413</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>...</td>\n",
       "      <td>188</td>\n",
       "      <td>47.00</td>\n",
       "      <td>2.583</td>\n",
       "      <td>6.124765e-06</td>\n",
       "      <td>-12.003170</td>\n",
       "      <td>-1.312186</td>\n",
       "      <td>-4.984972</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.047143</td>\n",
       "      <td>1.376441e-05</td>\n",
       "      <td>0.952843</td>\n",
       "      <td>63.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.222222</td>\n",
       "      <td>...</td>\n",
       "      <td>244</td>\n",
       "      <td>47.00</td>\n",
       "      <td>2.583</td>\n",
       "      <td>5.620967e-05</td>\n",
       "      <td>-9.786422</td>\n",
       "      <td>1.828127</td>\n",
       "      <td>-2.992058</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.015421</td>\n",
       "      <td>1.019360e-05</td>\n",
       "      <td>0.984569</td>\n",
       "      <td>52.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>...</td>\n",
       "      <td>244</td>\n",
       "      <td>47.00</td>\n",
       "      <td>2.583</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>-0.619039</td>\n",
       "      <td>-8.046489</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.005619</td>\n",
       "      <td>4.687815e-08</td>\n",
       "      <td>0.994381</td>\n",
       "      <td>60.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.100000</td>\n",
       "      <td>...</td>\n",
       "      <td>244</td>\n",
       "      <td>47.00</td>\n",
       "      <td>2.583</td>\n",
       "      <td>2.652597e-05</td>\n",
       "      <td>-10.537386</td>\n",
       "      <td>0.741937</td>\n",
       "      <td>-3.941058</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Medium</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.418368</td>\n",
       "      <td>5.687723e-01</td>\n",
       "      <td>0.012860</td>\n",
       "      <td>47.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.294737</td>\n",
       "      <td>...</td>\n",
       "      <td>213</td>\n",
       "      <td>43.25</td>\n",
       "      <td>-0.833</td>\n",
       "      <td>8.441952e-06</td>\n",
       "      <td>-11.682297</td>\n",
       "      <td>-1.221672</td>\n",
       "      <td>-5.483170</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.044593</td>\n",
       "      <td>7.929777e-01</td>\n",
       "      <td>0.162429</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>213</td>\n",
       "      <td>43.25</td>\n",
       "      <td>-0.833</td>\n",
       "      <td>4.292615e-07</td>\n",
       "      <td>-14.661200</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-8.551460</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.113869</td>\n",
       "      <td>5.741510e-01</td>\n",
       "      <td>0.311980</td>\n",
       "      <td>39.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>213</td>\n",
       "      <td>43.25</td>\n",
       "      <td>-0.833</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-7.279276</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.554205</td>\n",
       "      <td>4.213804e-01</td>\n",
       "      <td>0.024415</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>213</td>\n",
       "      <td>43.25</td>\n",
       "      <td>-0.833</td>\n",
       "      <td>2.434765e-05</td>\n",
       "      <td>-10.623075</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-3.551751</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.378583</td>\n",
       "      <td>2.230246e-01</td>\n",
       "      <td>0.398393</td>\n",
       "      <td>56.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>...</td>\n",
       "      <td>213</td>\n",
       "      <td>43.25</td>\n",
       "      <td>-0.833</td>\n",
       "      <td>9.880694e-06</td>\n",
       "      <td>-11.524928</td>\n",
       "      <td>0.223144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low</td>\n",
       "      <td>Low</td>\n",
       "      <td>Medium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Colony        date Genetic group  percent_Ligustica_Carnica  \\\n",
       "0     BS16_0060  06/07/2016     Caucasica                   0.118259   \n",
       "1     BS16_0061  06/07/2016     Caucasica                   0.109495   \n",
       "2     BS16_0162  31/08/2016     Caucasica                   0.047143   \n",
       "3     BS16_0163  31/08/2016     Caucasica                   0.015421   \n",
       "4     BS16_0164  31/08/2016     Caucasica                   0.005619   \n",
       "...         ...         ...           ...                        ...   \n",
       "1437  BS17_0506  01/08/2017        hybrid                   0.418368   \n",
       "1438  BS17_0511  01/08/2017        hybrid                   0.044593   \n",
       "1439  BS17_0512  01/08/2017        hybrid                   0.113869   \n",
       "1440  BS17_0513  01/08/2017        hybrid                   0.554205   \n",
       "1441  BS17_0517  01/08/2017        hybrid                   0.378583   \n",
       "\n",
       "      percent_Mellifera  percent_Caucasica  bee_weigth  nbr_pho_varroa  \\\n",
       "0          1.369552e-04           0.881604        51.0             0.0   \n",
       "1          9.200239e-05           0.890413        52.0             1.0   \n",
       "2          1.376441e-05           0.952843        63.0            28.0   \n",
       "3          1.019360e-05           0.984569        52.0             2.0   \n",
       "4          4.687815e-08           0.994381        60.0             9.0   \n",
       "...                 ...                ...         ...             ...   \n",
       "1437       5.687723e-01           0.012860        47.5             1.0   \n",
       "1438       7.929777e-01           0.162429        41.0             0.0   \n",
       "1439       5.741510e-01           0.311980        39.2             0.0   \n",
       "1440       4.213804e-01           0.024415        43.0             0.0   \n",
       "1441       2.230246e-01           0.398393        56.0             5.0   \n",
       "\n",
       "      nbr_pho_varroa_100bee     v_pho  ...  Day of Year  Latitude  Longitude  \\\n",
       "0                       0.0  0.000000  ...          188     47.00      2.583   \n",
       "1                       0.0  0.269231  ...          188     47.00      2.583   \n",
       "2                       6.0  6.222222  ...          244     47.00      2.583   \n",
       "3                       1.0  0.538462  ...          244     47.00      2.583   \n",
       "4                       2.0  2.100000  ...          244     47.00      2.583   \n",
       "...                     ...       ...  ...          ...       ...        ...   \n",
       "1437                    0.0  0.294737  ...          213     43.25     -0.833   \n",
       "1438                    0.0  0.000000  ...          213     43.25     -0.833   \n",
       "1439                    0.0  0.000000  ...          213     43.25     -0.833   \n",
       "1440                    0.0  0.000000  ...          213     43.25     -0.833   \n",
       "1441                    1.0  1.250000  ...          213     43.25     -0.833   \n",
       "\n",
       "            v_meta  log_v_meta  log_v_pho  log_v_mito  Varroa group  \\\n",
       "0     0.000000e+00  -15.168735  -1.987004   -8.156349           Low   \n",
       "1     6.124765e-06  -12.003170  -1.312186   -4.984972           Low   \n",
       "2     5.620967e-05   -9.786422   1.828127   -2.992058        Medium   \n",
       "3     0.000000e+00  -15.168735  -0.619039   -8.046489           Low   \n",
       "4     2.652597e-05  -10.537386   0.741937   -3.941058        Medium   \n",
       "...            ...         ...        ...         ...           ...   \n",
       "1437  8.441952e-06  -11.682297  -1.221672   -5.483170           Low   \n",
       "1438  4.292615e-07  -14.661200  -1.987004   -8.551460           Low   \n",
       "1439  0.000000e+00  -15.168735  -1.987004   -7.279276           Low   \n",
       "1440  2.434765e-05  -10.623075  -1.987004   -3.551751        Medium   \n",
       "1441  9.880694e-06  -11.524928   0.223144         NaN           Low   \n",
       "\n",
       "      Recapping group  DMR group  \n",
       "0                 Low        NaN  \n",
       "1                 Low        Low  \n",
       "2              Medium     Medium  \n",
       "3                 Low        Low  \n",
       "4              Medium       High  \n",
       "...               ...        ...  \n",
       "1437              Low        Low  \n",
       "1438              Low       High  \n",
       "1439              Low       High  \n",
       "1440              Low        Low  \n",
       "1441              Low     Medium  \n",
       "\n",
       "[1442 rows x 44 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reorder pheno df according to fam_df\n",
    "pheno_df = pheno_df.set_index(\"Colony\").loc[fam_df[1].to_list()].reset_index()\n",
    "pheno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "e4ed0eaf-e96a-493e-8d96-62ec7fe8346f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check consistent colony order between pheno df and plink\n",
    "bs_ids = pheno_df['Colony'].to_list()\n",
    "assert fam_df[1].to_list() == bs_ids, 'messed up colony order'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "cc74acf6-5acb-4998-bd28-d09255716635",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['Colony', 'log_v_pho', 'log_v_meta', 'Apiary', 'date', 'Year', 'Day of Year']\n",
    "\n",
    "# Varroa-Associated families\n",
    "vaf_abund = [\n",
    "    'Bifidobacteriaceae',\n",
    "    'Lactobacillaceae',\n",
    "    'Pseudomonadaceae',   \n",
    "    'Debaryomycetaceae',\n",
    "    'Bartonellaceae',  \n",
    "    'Morganellaceae'\n",
    "]\n",
    "vaf_prev = [\n",
    "    'Oscillospiraceae',\n",
    "    'Corynebacteriaceae',\n",
    "    'Cannabaceae',\n",
    "    'Aeromonadaceae',\n",
    "    'Weeksellaceae',\n",
    "    'Araliaceae',\n",
    "    'Actinomycetaceae',\n",
    "    'Trypanosomatidae'\n",
    "]\n",
    "\n",
    "fn = '{}tlf/Core_Family_LRA.csv'.format(working_path)\n",
    "abund_df = pd.read_csv(fn) #, index_col='Colony').reset_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "ce3b5f9d-b606-47c4-8ba9-c6932fb7f95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gcta_pheno_df = pd.DataFrame(\n",
    "    {\n",
    "        '0' : [0] * len(bs_ids),\n",
    "        # '0': bs_ids,\n",
    "        'Colony': bs_ids\n",
    "    }\n",
    ")\n",
    "gcta_pheno_df = gcta_pheno_df.merge(pheno_df[columns], on='Colony', how='left')\n",
    "gcta_pheno_df = gcta_pheno_df.merge(abund_df[['Colony'] + vaf_abund + vaf_prev], on='Colony', how='left')\n",
    "\n",
    "# replace abundance values by presence-absence for prevalence associations\n",
    "lod = np.log(5e-06)\n",
    "gcta_pheno_df[gcta_pheno_df[vaf_prev] >= lod] = int(1)\n",
    "gcta_pheno_df[gcta_pheno_df[vaf_prev] < lod] = int(0)\n",
    "\n",
    "gcta_pheno_df = gcta_pheno_df.fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "f198551c-3e89-4031-8732-1b9934f15edd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>Colony</th>\n",
       "      <th>log_v_pho</th>\n",
       "      <th>log_v_meta</th>\n",
       "      <th>Apiary</th>\n",
       "      <th>date</th>\n",
       "      <th>Year</th>\n",
       "      <th>Day of Year</th>\n",
       "      <th>Bifidobacteriaceae</th>\n",
       "      <th>Lactobacillaceae</th>\n",
       "      <th>...</th>\n",
       "      <th>Bartonellaceae</th>\n",
       "      <th>Morganellaceae</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Corynebacteriaceae</th>\n",
       "      <th>Cannabaceae</th>\n",
       "      <th>Aeromonadaceae</th>\n",
       "      <th>Weeksellaceae</th>\n",
       "      <th>Araliaceae</th>\n",
       "      <th>Actinomycetaceae</th>\n",
       "      <th>Trypanosomatidae</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>88.0</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>188</td>\n",
       "      <td>-6.146180</td>\n",
       "      <td>-4.997377</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.744082</td>\n",
       "      <td>-9.018644</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>-1.312186</td>\n",
       "      <td>-12.003170</td>\n",
       "      <td>88.0</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>188</td>\n",
       "      <td>-6.114353</td>\n",
       "      <td>-4.710165</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.098915</td>\n",
       "      <td>-10.080948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>1.828127</td>\n",
       "      <td>-9.786422</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>244</td>\n",
       "      <td>-5.950136</td>\n",
       "      <td>-4.598848</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.069882</td>\n",
       "      <td>-7.210088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>-0.619039</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>244</td>\n",
       "      <td>-5.707229</td>\n",
       "      <td>-4.439819</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.104517</td>\n",
       "      <td>-10.880938</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>0.741937</td>\n",
       "      <td>-10.537386</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>244</td>\n",
       "      <td>-5.405451</td>\n",
       "      <td>-4.150527</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.050514</td>\n",
       "      <td>-9.954003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>-1.221672</td>\n",
       "      <td>-11.682297</td>\n",
       "      <td>128.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.955760</td>\n",
       "      <td>-4.546061</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.303314</td>\n",
       "      <td>-5.188458</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-14.661200</td>\n",
       "      <td>128.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.556695</td>\n",
       "      <td>-3.947906</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.801414</td>\n",
       "      <td>-8.585859</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>97.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.081466</td>\n",
       "      <td>-4.056083</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.312203</td>\n",
       "      <td>-9.878038</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-10.623075</td>\n",
       "      <td>97.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.038656</td>\n",
       "      <td>-4.301295</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.182466</td>\n",
       "      <td>-4.987978</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>0.223144</td>\n",
       "      <td>-11.524928</td>\n",
       "      <td>97.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-6.144559</td>\n",
       "      <td>-4.356450</td>\n",
       "      <td>...</td>\n",
       "      <td>-8.342813</td>\n",
       "      <td>-5.329028</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0     Colony log_v_pho  log_v_meta  Apiary        date  Year  \\\n",
       "0     0  BS16_0060 -1.987004  -15.168735    88.0  06/07/2016  2016   \n",
       "1     0  BS16_0061 -1.312186  -12.003170    88.0  06/07/2016  2016   \n",
       "2     0  BS16_0162  1.828127   -9.786422    88.0  31/08/2016  2016   \n",
       "3     0  BS16_0163 -0.619039  -15.168735    88.0  31/08/2016  2016   \n",
       "4     0  BS16_0164  0.741937  -10.537386    88.0  31/08/2016  2016   \n",
       "...  ..        ...       ...         ...     ...         ...   ...   \n",
       "1437  0  BS17_0506 -1.221672  -11.682297   128.0  01/08/2017  2017   \n",
       "1438  0  BS17_0511 -1.987004  -14.661200   128.0  01/08/2017  2017   \n",
       "1439  0  BS17_0512 -1.987004  -15.168735    97.0  01/08/2017  2017   \n",
       "1440  0  BS17_0513 -1.987004  -10.623075    97.0  01/08/2017  2017   \n",
       "1441  0  BS17_0517  0.223144  -11.524928    97.0  01/08/2017  2017   \n",
       "\n",
       "      Day of Year  Bifidobacteriaceae  Lactobacillaceae  ...  Bartonellaceae  \\\n",
       "0             188           -6.146180         -4.997377  ...       -6.744082   \n",
       "1             188           -6.114353         -4.710165  ...       -6.098915   \n",
       "2             244           -5.950136         -4.598848  ...       -7.069882   \n",
       "3             244           -5.707229         -4.439819  ...       -6.104517   \n",
       "4             244           -5.405451         -4.150527  ...       -7.050514   \n",
       "...           ...                 ...               ...  ...             ...   \n",
       "1437          213           -5.955760         -4.546061  ...       -7.303314   \n",
       "1438          213           -5.556695         -3.947906  ...       -5.801414   \n",
       "1439          213           -5.081466         -4.056083  ...       -9.312203   \n",
       "1440          213           -5.038656         -4.301295  ...       -7.182466   \n",
       "1441          213           -6.144559         -4.356450  ...       -8.342813   \n",
       "\n",
       "      Morganellaceae  Oscillospiraceae  Corynebacteriaceae  Cannabaceae  \\\n",
       "0          -9.018644               0.0                 0.0          0.0   \n",
       "1         -10.080948               0.0                 0.0          0.0   \n",
       "2          -7.210088               0.0                 0.0          1.0   \n",
       "3         -10.880938               0.0                 1.0          0.0   \n",
       "4          -9.954003               0.0                 0.0          1.0   \n",
       "...              ...               ...                 ...          ...   \n",
       "1437       -5.188458               0.0                 0.0          0.0   \n",
       "1438       -8.585859               0.0                 0.0          0.0   \n",
       "1439       -9.878038               0.0                 1.0          1.0   \n",
       "1440       -4.987978               0.0                 1.0          0.0   \n",
       "1441       -5.329028               0.0                 0.0          1.0   \n",
       "\n",
       "      Aeromonadaceae  Weeksellaceae  Araliaceae  Actinomycetaceae  \\\n",
       "0                1.0            1.0         1.0               0.0   \n",
       "1                1.0            1.0         1.0               0.0   \n",
       "2                0.0            1.0         1.0               0.0   \n",
       "3                1.0            1.0         1.0               0.0   \n",
       "4                0.0            1.0         0.0               0.0   \n",
       "...              ...            ...         ...               ...   \n",
       "1437             0.0            1.0         0.0               1.0   \n",
       "1438             0.0            1.0         0.0               1.0   \n",
       "1439             0.0            1.0         0.0               0.0   \n",
       "1440             1.0            1.0         0.0               1.0   \n",
       "1441             0.0            1.0         0.0               0.0   \n",
       "\n",
       "      Trypanosomatidae  \n",
       "0                  1.0  \n",
       "1                  1.0  \n",
       "2                  1.0  \n",
       "3                  0.0  \n",
       "4                  0.0  \n",
       "...                ...  \n",
       "1437               0.0  \n",
       "1438               1.0  \n",
       "1439               0.0  \n",
       "1440               0.0  \n",
       "1441               0.0  \n",
       "\n",
       "[1442 rows x 22 columns]"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gcta_pheno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "4f9e8843-f508-4f65-a3c8-38f7440251f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>-15.168735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>-12.003170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>-9.786422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>-15.168735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>-10.537386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>-11.682297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>-14.661200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>-15.168735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>-10.623075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>-11.524928</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2\n",
       "0     0  BS16_0060 -15.168735\n",
       "1     0  BS16_0061 -12.003170\n",
       "2     0  BS16_0162  -9.786422\n",
       "3     0  BS16_0163 -15.168735\n",
       "4     0  BS16_0164 -10.537386\n",
       "...  ..        ...        ...\n",
       "1437  0  BS17_0506 -11.682297\n",
       "1438  0  BS17_0511 -14.661200\n",
       "1439  0  BS17_0512 -15.168735\n",
       "1440  0  BS17_0513 -10.623075\n",
       "1441  0  BS17_0517 -11.524928\n",
       "\n",
       "[1442 rows x 3 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# phenotypes\n",
    "fn = '{}{}.v_pho'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'log_v_pho']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "\n",
    "fn = '{}{}.v_meta'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'log_v_meta']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "\n",
    "pd.read_table(fn, header=None) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "8f8cdf31-0bc3-4151-bedd-53067916c2d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1      2\n",
       "0     0  BS16_0060   88.0\n",
       "1     0  BS16_0061   88.0\n",
       "2     0  BS16_0162   88.0\n",
       "3     0  BS16_0163   88.0\n",
       "4     0  BS16_0164   88.0\n",
       "...  ..        ...    ...\n",
       "1437  0  BS17_0506  128.0\n",
       "1438  0  BS17_0511  128.0\n",
       "1439  0  BS17_0512   97.0\n",
       "1440  0  BS17_0513   97.0\n",
       "1441  0  BS17_0517   97.0\n",
       "\n",
       "[1442 rows x 3 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# categorical covariable\n",
    "fn = '{}{}.covar'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'Apiary']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony', 'Apiary', 'Year'] + vaf_prev].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony'] + vaf_prev].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony', 'Apiary']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "pd.read_table(fn, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "d7e9c698-ca53-4bf0-9ed7-b5aabc61181e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1    2\n",
       "0     0  BS16_0060  188\n",
       "1     0  BS16_0061  188\n",
       "2     0  BS16_0162  244\n",
       "3     0  BS16_0163  244\n",
       "4     0  BS16_0164  244\n",
       "...  ..        ...  ...\n",
       "1437  0  BS17_0506  213\n",
       "1438  0  BS17_0511  213\n",
       "1439  0  BS17_0512  213\n",
       "1440  0  BS17_0513  213\n",
       "1441  0  BS17_0517  213\n",
       "\n",
       "[1442 rows x 3 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# continuous covariable\n",
    "fn = '{}{}.qcovar'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'Day of Year']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony', 'Day of Year'] + vaf_abund].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "#gcta_pheno_df[['0', 'Colony'] + vaf_abund].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "pd.read_table(fn, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f2e0d0-7d47-4826-8d3a-493c2f070348",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
