{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "edf9b481-8650-4505-b1c8-23a14de78ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "from tqdm import tqdm as tqdm\n",
    "import pandas as pd\n",
    "import bz2\n",
    "import subprocess\n",
    "import pickle as pkl\n",
    "from itertools import chain\n",
    "from glob import glob\n",
    "from collections import defaultdict\n",
    "from IPython.display import HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "faaa1eff-e7bf-4f10-a5e8-fe67425f1681",
   "metadata": {},
   "outputs": [],
   "source": [
    "working_path = '/cluster/raid/home/f80878961/beestrong/'\n",
    "count_path = '{}data/allele_freqs/sync_files/'.format(working_path)\n",
    "tlf_path = '{}tlf/'.format(working_path)\n",
    "script_path = \"/cluster/raid/home/f80878961/scripts/tmp_scripts/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "faf43f06-0ec3-4e12-969b-17255a018f84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bs_ids = pheno_df['Colony'].to_list()\n",
    "# snp_nr = 7023976 # will need sanity check here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d409b7-31bd-4f5b-9dae-70b34ec06c83",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "42d775bb-783a-40c8-a80f-89e5ef376fe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1513\n",
      "1513\n"
     ]
    }
   ],
   "source": [
    "snp_nr = 7023976 # wc -l BS17_0071.count -1\n",
    "\n",
    "prefix = 'All'\n",
    "if prefix == 'All':\n",
    "    pheno_df = pd.read_table('/cluster/raid/home/f80878961/beestrong/data/metadata/whole_beestrong_data.csv').rename(\n",
    "        columns={\n",
    "            'num_bs':'Colony',\n",
    "            'group':'Genetic group'\n",
    "        })\n",
    "else: \n",
    "    pheno_df = pd.read_csv('{}BeeStrong_metadata.csv'.format(tlf_path))\n",
    "\n",
    "filtered_count_path = '{}data/allele_freqs/{}/'.format(working_path, prefix)\n",
    "queen_geno_path = '{}data/queen_genotypes/{}/'.format(working_path, prefix)\n",
    "\n",
    "bs_ids = sorted([x.split('/')[-1].split('.')[0] for x in glob('{}*.count.bz2'.format(count_path))])\n",
    "\n",
    "chunk_size = 101 # 15 chunks\n",
    "batch_file = '{}bsid_batch.txt'.format(filtered_count_path)\n",
    "\n",
    "# bs_ids = pheno_df['Colony'].to_list()\n",
    "print(len(set(bs_ids).intersection(pheno_df['Colony'].to_list())))\n",
    "print(len(bs_ids))\n",
    "assert pheno_df['Colony'].to_list() == bs_ids\n",
    "\n",
    "# for SNP filtering\n",
    "min_prev = 0.1 # prevalence threshold for common alleles (if >2 common allele, then real tri-allelic)\n",
    "min_depth = 10 # both mean and median depth \n",
    "max_depth = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c71d2c00-42b6-47cc-9cb4-10c85c716518",
   "metadata": {},
   "source": [
    "## Merge allele frequencies .sync files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42e45547-d8bd-40a2-bbe7-7249db379db9",
   "metadata": {},
   "source": [
    "### Write per-allele count matrices (SNP x colony for A, T, C, G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "15b9cec1-d70e-426d-85f4-018666e4fbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_merge_sync_script(script_fn, array_str):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --array=ARRAY_STRING\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=20g\n",
    "#SBATCH --time=4:00:00\n",
    "#SBATCH --job-name=merge_sync\n",
    "#SBATCH --output=%x_%A_%a.out\n",
    "#SBATCH --error=%x_%A_%a.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "# file listing batches of BeeStrong ids (bs_id) for parallel computing\n",
    "batch_file=$1\n",
    "snp_nr=$2\n",
    "count_path=$3\n",
    "out_path=$4\n",
    "\n",
    "batch_id=$SLURM_ARRAY_TASK_ID\n",
    "\n",
    "# getting bs_ids\n",
    "bs_ids=$(sed -n ${batch_id}'{p;q}' ${batch_file})\n",
    "\n",
    "python ~/scripts/plan-b-omics/bin/merge_sync.py --batch_id ${batch_id} --snp_nr ${snp_nr} --bs_ids ${bs_ids} --count_path ${count_path} --out_path ${out_path}\n",
    "\n",
    "echo DONE\"\"\".replace(\"ARRAY_STRING\", array_str)\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c51fc6da-d8c3-4a76-aad4-69131f6c73f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create batch file for first time\n",
    "with open(batch_file, \"w\") as outf:\n",
    "    for i in range(0, len(bs_ids), chunk_size):\n",
    "        chunk = bs_ids[i:i+chunk_size]\n",
    "        outf.write(\",\".join(chunk) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "36329bf3-41ed-4dbc-9ced-c443339c9499",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_str = '1-15'\n",
    "script_fn = '{}merge_sync.run'.format(script_path)\n",
    "write_merge_sync_script(script_fn, array_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1f6b1d44-ed80-4424-8d0b-f25574d5129e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 1388883\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$script_fn\" \"$batch_file\" \"$snp_nr\" \"$count_path\" \"$filtered_count_path\"\n",
    "cd $5\n",
    "sbatch $1 $2 $3 $4 $5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "393d9d30-7544-4338-8ca3-c438f19c60b9",
   "metadata": {},
   "source": [
    "### Merge into 3-D matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "db125385-1ec0-4c05-bb94-585eafb3bb3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "assert sum([x == bs_ids[i] for i, x in enumerate(list(chain(*batch_bs_ids)))]) == len(bs_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "308c7820-70f6-4dbf-a3f0-7b97e20f1919",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:27, 79911.85it/s]\n",
      "7023976it [01:12, 96871.43it/s]\n",
      "7023976it [01:13, 95607.20it/s]\n",
      "7023976it [01:12, 96395.80it/s]\n",
      "7023976it [01:13, 95634.25it/s]\n",
      "7023976it [01:13, 95841.95it/s]\n",
      "7023976it [01:12, 96750.18it/s]\n",
      "7023976it [01:13, 96216.97it/s]\n",
      "7023976it [01:13, 96170.55it/s]\n",
      "7023976it [01:13, 94955.64it/s]\n",
      "7023976it [01:10, 99338.00it/s] \n",
      "7023976it [01:10, 100071.38it/s]\n",
      "7023976it [01:10, 99098.12it/s] \n",
      "7023976it [01:20, 86892.53it/s]\n",
      "7023976it [01:19, 88464.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:20, 87638.04it/s]\n",
      "7023976it [01:19, 88887.25it/s]\n",
      "7023976it [01:20, 87229.69it/s]\n",
      "7023976it [01:21, 86617.28it/s]\n",
      "7023976it [01:21, 86346.12it/s]\n",
      "7023976it [01:20, 87418.80it/s]\n",
      "7023976it [01:20, 87219.67it/s]\n",
      "7023976it [01:22, 85541.27it/s]\n",
      "7023976it [01:22, 85528.61it/s]\n",
      "7023976it [01:22, 85412.89it/s]\n",
      "7023976it [01:21, 86602.88it/s]\n",
      "7023976it [01:21, 86004.87it/s]\n",
      "7023976it [01:20, 86980.82it/s]\n",
      "7023976it [01:21, 86634.61it/s]\n",
      "7023976it [01:19, 87840.42it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:22, 85362.98it/s]\n",
      "7023976it [01:21, 86695.99it/s]\n",
      "7023976it [01:21, 86272.75it/s]\n",
      "7023976it [01:21, 85767.50it/s]\n",
      "7023976it [01:23, 83933.48it/s]\n",
      "7023976it [01:21, 85708.65it/s]\n",
      "7023976it [01:22, 85253.64it/s]\n",
      "7023976it [01:22, 84884.01it/s]\n",
      "7023976it [01:23, 84503.23it/s]\n",
      "7023976it [01:23, 84575.33it/s]\n",
      "7023976it [01:23, 84403.22it/s]\n",
      "7023976it [01:22, 85003.84it/s]\n",
      "7023976it [01:22, 84759.25it/s]\n",
      "7023976it [01:23, 84321.06it/s]\n",
      "7023976it [01:21, 86137.15it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [01:21, 86597.14it/s]\n",
      "7023976it [01:21, 86523.52it/s]\n",
      "7023976it [01:23, 84421.79it/s]\n",
      "7023976it [01:21, 85817.88it/s]\n",
      "7023976it [01:22, 85413.62it/s]\n",
      "7023976it [01:22, 85604.58it/s]\n",
      "7023976it [01:22, 85174.04it/s]\n",
      "7023976it [01:23, 84585.02it/s]\n",
      "7023976it [01:24, 83101.66it/s]\n",
      "7023976it [01:22, 84669.79it/s]\n",
      "7023976it [01:22, 84762.85it/s]\n",
      "7023976it [01:22, 85111.30it/s]\n",
      "7023976it [01:21, 86124.85it/s]\n",
      "7023976it [01:22, 85098.11it/s]\n",
      "7023976it [01:20, 86877.37it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1h 9min 1s, sys: 10min 24s, total: 1h 19min 25s\n",
      "Wall time: 1h 20min 4s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "allele2jj = {'A': 0, 'T': 1, 'C':2, 'G':3}\n",
    "count_mat = np.zeros((snp_nr, len(bs_ids), len(allele2jj)), dtype=np.uint16)\n",
    "\n",
    "for allele, jj in allele2jj.items():\n",
    "    print(allele, jj)\n",
    "    \n",
    "    for j, b_bs_ids in enumerate(batch_bs_ids):\n",
    "        batch_id = j + 1\n",
    "        batch_size = len(b_bs_ids)\n",
    "        with open('{}count_{}_{}.txt'.format(filtered_count_path, allele, batch_id), 'r') as inf:\n",
    "            i = 0\n",
    "            for l in tqdm(inf):                    \n",
    "                count_mat[i, j * chunk_size : j * chunk_size + batch_size, jj] = np.array(l.split()[2:], dtype=np.int16)\n",
    "                i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f3ff2627-20b8-4001-98c8-1ab5aa5f12b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 909 ms, sys: 2min 15s, total: 2min 16s\n",
      "Wall time: 2min 33s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "with open('{}count_mat.pkl'.format(filtered_count_path), 'wb') as outf:\n",
    "    pkl.dump(count_mat, outf, protocol=pkl.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5170bfda-6e58-4b04-a29a-5b7204eee9a3",
   "metadata": {},
   "source": [
    "## Filter SNPs and recode alleles"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2dfe5850-b452-4d42-8541-14e93c667c9b",
   "metadata": {},
   "source": [
    "filtering \n",
    "1. monoallelic filtering : 3/4 tables empty\n",
    "2. mitochondrial SNP: contig NC_001566.1\n",
    "3. tri-allelic SNP filtering: >2 allele present in min. 10% colonies, else delete variants from third and fourth alleles\n",
    "4. min-max depth (10/50, on mean AND median) (after tri-allelic variants deletions, also not including N and deletions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "25f49154-3eb6-48e2-be44-43a9ffdf9025",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85.018205504\n",
      "CPU times: user 0 ns, sys: 1min 10s, total: 1min 10s\n",
      "Wall time: 1min 26s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "with open('{}count_mat.pkl'.format(filtered_count_path), 'rb') as inf:\n",
    "    count_mat = pkl.load(inf)\n",
    "print(count_mat.nbytes/1000000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cc168aea-5e7b-48ab-b27b-f38a82a65532",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "assert sum([x == bs_ids[i] for i, x in enumerate(list(chain(*batch_bs_ids)))]) == len(bs_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cb149492-4685-4689-997d-cb55aa1240ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1513\n"
     ]
    }
   ],
   "source": [
    "batch_i = 0\n",
    "batch_ii = len(bs_ids)\n",
    "print(batch_ii)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cddd8f11-a4e8-42e7-8ae0-e348ae8ae3e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 44s, sys: 5.78 s, total: 2min 50s\n",
      "Wall time: 2min 51s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[   1, 1513,    3,    9],\n",
       "       [   4, 1513,   27,    2],\n",
       "       [1513,   11,    2,   22],\n",
       "       ...,\n",
       "       [1437,  849,  118,  337],\n",
       "       [ 397, 1364, 1058,  218],\n",
       "       [ 335, 1504,  310,  170]])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "# counts the number of A, T, C, G allele per SNP # should take 7min\n",
    "allele_prev = np.sum(count_mat[:, batch_i:batch_ii, :] > 0, axis=1)\n",
    "allele_prev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "183ab84d-5cef-473d-bf56-57e8b6053d89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3261\n",
      "6963287\n",
      "17207\n",
      "CPU times: user 433 ms, sys: 41.7 ms, total: 475 ms\n",
      "Wall time: 479 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# if count 3/4 alleles == 0, then monoallelic\n",
    "mono_mask = np.sum(allele_prev > 0, axis=1) == 1\n",
    "print(np.sum(mono_mask))\n",
    "\n",
    "# tri-allelic stringent\n",
    "tri_stringent_mask = np.sum(allele_prev > 0, axis=1) > 2\n",
    "print(np.sum(tri_stringent_mask))\n",
    "\n",
    "# tri-allelic relaxed (filter SNP only when >2 common alleles, i.e., an allele present in >= 10% of colonies)\n",
    "tri_mask = np.sum(allele_prev >= ((batch_ii - batch_i) * min_prev), axis=1) > 2\n",
    "print(np.sum(tri_mask))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f9a4d2e1-4fe3-4c2a-9f39-cfdc387f68ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6963287/6963287 [00:50<00:00, 138304.88it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "122115"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# iterate through stringent tri-allelic SNPs: filter if ambigous else recode (delete 3rd and fourth)\n",
    "ambi_mask = np.full(snp_nr, False)\n",
    "for i in tqdm(np.argwhere(tri_stringent_mask).flatten()):\n",
    "\n",
    "    # get number of alleles and sort by prevalence \n",
    "    atcg_prev = allele_prev[i, :]\n",
    "    alleles_sorted_by_prevalence = np.argsort(atcg_prev)\n",
    "\n",
    "    # if 2nd and 3rd alleles are equally common --> ambiguous --> filter\n",
    "    if atcg_prev[alleles_sorted_by_prevalence[1]] == atcg_prev[alleles_sorted_by_prevalence[2]]:\n",
    "        ambi_mask[i] = True\n",
    "\n",
    "    # else delete 3rd and 4rth alleles\n",
    "    else:\n",
    "        count_mat[i, batch_i:batch_ii, alleles_sorted_by_prevalence[:2]] = 0\n",
    "\n",
    "np.sum(ambi_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3aaa7293-7cdd-4005-929e-d0392c9b25da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 23s, sys: 58 s, total: 3min 21s\n",
      "Wall time: 3min 22s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# depth filtering --> if to, long I could consider iterating ...to have an overview of the time it takes # 8 min\n",
    "depth  = np.sum(count_mat[:, batch_i:batch_ii, :], axis = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "df55178f-0037-48cc-8841-2fb18beed331",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7023976/7023976 [02:47<00:00, 41860.60it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 4s, sys: 686 ms, total: 3min 5s\n",
      "Wall time: 3min 7s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "123823"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "mean_depth = np.mean(depth, axis = 1)\n",
    "median_depth = np.full(snp_nr, -1) # np.median breaks the kernel somehow\n",
    "for i in tqdm(range(snp_nr)):\n",
    "    median_depth[i] =  np.median(depth[i])\n",
    "\n",
    "depth_mask = (mean_depth < min_depth) | (mean_depth > max_depth) | (median_depth < min_depth) | (median_depth > max_depth) \n",
    "np.sum(depth_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f890a3e1-0bcd-481d-8aca-534faebc82ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [00:25, 272771.65it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 23.4 s, sys: 2.03 s, total: 25.4 s\n",
      "Wall time: 25.8 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "287"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "chr_ids = [''] * snp_nr\n",
    "snp_pos = [''] * snp_nr\n",
    "mito_mask = np.full(snp_nr, False)\n",
    "with open('{}count_A_1.txt'.format(filtered_count_path), 'r') as inf:\n",
    "    i = 0\n",
    "    for l in tqdm(inf):\n",
    "        ch = l.split()[0]\n",
    "        pos = l.split()[1]\n",
    "        chr_ids[i] = ch\n",
    "        snp_pos[i] = pos\n",
    "        if ch == 'NC_001566.1':\n",
    "            mito_mask[i] = True\n",
    "        i += 1\n",
    "np.sum(mito_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a50bfbf9-63ef-43a1-9ce4-7f87cf90f2e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "257718\n",
      "6766258\n"
     ]
    }
   ],
   "source": [
    "# combine masks\n",
    "mask = mono_mask + tri_mask + depth_mask + mito_mask + ambi_mask\n",
    "print(np.sum(mask))\n",
    "print(snp_nr - np.sum(mask))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24c1c375-5874-4d09-82ec-26d447c4bc40",
   "metadata": {},
   "source": [
    "### Write PLINK bim_like files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eb269d9-ecc2-4f97-b394-2f9238b72828",
   "metadata": {},
   "source": [
    "1. Chromosome code\n",
    "2. Variant ID\n",
    "3. Position in centimorgans (safe to use dummy value of '0')\n",
    "4. Base-pair coordinate (1-based; limited to 231-2)\n",
    "5. ALT ('A1' in PLINK 1.x) allele code\n",
    "6. REF ('A2' in PLINK 1.x) allele code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d504d7c2-e797-45b5-9e20-dcc8c9046da8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6766258/6766258 [03:38<00:00, 30928.81it/s]\n"
     ]
    }
   ],
   "source": [
    "# find reference and alternative allele per SNP (REF: the one with the most counts)\n",
    "jj2allele = np.array(['A', 'T', 'C', 'G'])\n",
    "ref_alt = np.full((snp_nr, 2), '.')\n",
    "for i in tqdm(np.argwhere(~mask).flatten()):\n",
    "    ref_alt[i] = jj2allele[np.argsort(np.sum(count_mat[i, :, :], axis=0))[::-1][:2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4c2c584c-6b13-41ce-b994-4e099a09b412",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2652648/1469035007.py:12: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  bim_like['CHROM'] = bim_like['CHROM'].replace(chr2int)\n"
     ]
    }
   ],
   "source": [
    "bim_like = pd.DataFrame({\n",
    "    'CHROM': np.array(chr_ids)[~mask],\n",
    "    'SNP_ID': np.array(snp_pos)[~mask],\n",
    "    'GEN_DIST': ['0'] * np.sum(~mask),\n",
    "    'POS': np.array(snp_pos)[~mask],\n",
    "    'ALT': ref_alt[~mask][:, 1],\n",
    "    'REF': ref_alt[~mask][:, 0]\n",
    "})\n",
    "\n",
    "# recode chromosome and SNP identifiers into integers\n",
    "chr2int = dict(zip(list(dict.fromkeys(bim_like['CHROM'])), range(1, 1 + len(set(bim_like['CHROM'])))))\n",
    "bim_like['CHROM'] = bim_like['CHROM'].replace(chr2int)\n",
    "\n",
    "bim_like['CHROM'] = bim_like['CHROM'].astype(str)\n",
    "bim_like['GEN_DIST'] = bim_like['GEN_DIST'].astype(str)\n",
    "bim_like['SNP_ID'] = bim_like[[\"CHROM\", \"POS\"]].agg(\":\".join, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3a704779-ac66-4e3f-a0ee-6bcd567060e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GEN_DIST</th>\n",
       "      <th>POS</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6766253</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6766254</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6766255</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6766256</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6766257</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6766258 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM      SNP_ID GEN_DIST      POS ALT REF\n",
       "0           1      1:5671        0     5671   G   T\n",
       "1           1      1:5698        0     5698   C   T\n",
       "2           1      1:6621        0     6621   G   A\n",
       "3           1      1:6717        0     6717   A   C\n",
       "4           1      1:6823        0     6823   C   G\n",
       "...       ...         ...      ...      ...  ..  ..\n",
       "6766253    16  16:7225908        0  7225908   C   T\n",
       "6766254    16  16:7225911        0  7225911   T   C\n",
       "6766255    16  16:7226186        0  7226186   A   G\n",
       "6766256    16  16:7226346        0  7226346   A   G\n",
       "6766257    16  16:7227034        0  7227034   A   G\n",
       "\n",
       "[6766258 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a3c5d396-6af1-437c-8e1c-67f1b7b8a5d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.17 s, sys: 395 ms, total: 6.57 s\n",
      "Wall time: 6.73 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "bim_like.to_csv('{}{}.bim_like'.format(queen_geno_path, prefix), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c4ef8aa-b22e-4454-96b8-09960ee4dbad",
   "metadata": {},
   "source": [
    "### Export count_ref.txt and depth.txt files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61726916-cd32-4c5a-b2be-8f158d4c8738",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TO DO: insert also alleles! --> aim to have same .bgs as Sonia!\n",
    "# ,CHROM,POS,REF,ALT,BS16_0032..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4d523131-3938-4543-bba9-bce52c33accd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 8.46 s, sys: 15.3 s, total: 23.8 s\n",
      "Wall time: 23.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# creating depth dataframe # 30min...\n",
    "depth_df = pd.DataFrame(depth[~mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c571529-ce42-47c2-b149-fccd18e7fa43",
   "metadata": {},
   "outputs": [],
   "source": [
    "depth_df.insert(0, 'CHROM',  bim_like['CHROM'])\n",
    "depth_df.insert(1, 'POS', bim_like['POS'])\n",
    "depth_df.insert(2, 'REF',  bim_like['REF'])\n",
    "depth_df.insert(3, 'ALT', bim_like['ALT'])\n",
    "depth_df.columns = ['CHROM', 'POS', 'REF', 'ALT'] + bs_ids[batch_i: batch_ii]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "edd5211a-7a30-4e78-b383-da164d3ccec1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caucasica\n",
      "hybrid\n",
      "Mellifera\n",
      "Ligustica_Carnica\n",
      "CPU times: user 23min 46s, sys: 2min 26s, total: 26min 13s\n",
      "Wall time: 26min 31s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for gg in set(pheno_df['Genetic group']):\n",
    "    print(gg)\n",
    "    depth_df[['CHROM', 'POS', 'REF', 'ALT'] + list(pheno_df[pheno_df['Genetic group'] == gg]['Colony'])].to_csv('{}depth_{}.txt'.format(queen_geno_path, gg), header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "68837dde-434a-4da7-a4d4-7dcf1bb79a1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 58s, sys: 4min 6s, total: 9min 4s\n",
      "Wall time: 9min 5s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# getting the ref allele counts 5-8min...\n",
    "count_ref  = np.max(count_mat[:, batch_i:batch_ii, :], axis = 2)\n",
    "\n",
    "# creating reference allele count dataframe\n",
    "count_ref_df = pd.DataFrame(count_ref[~mask])\n",
    "count_ref_df.insert(0, 'CHROM',  bim_like['CHROM'])\n",
    "count_ref_df.insert(1, 'POS', bim_like['POS'])\n",
    "count_ref_df.insert(2, 'REF',  bim_like['REF'])\n",
    "count_ref_df.insert(3, 'ALT', bim_like['ALT'])\n",
    "count_ref_df.columns = ['CHROM', 'POS', 'REF', 'ALT'] + bs_ids[batch_i: batch_ii]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5bf21168-b156-4be0-8dc7-6d38fc714399",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caucasica\n",
      "hybrid\n",
      "Mellifera\n",
      "Ligustica_Carnica\n",
      "CPU times: user 22min 17s, sys: 1min 5s, total: 23min 22s\n",
      "Wall time: 23min 39s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for gg in set(pheno_df['Genetic group']):\n",
    "    print(gg)\n",
    "    count_ref_df[['CHROM', 'POS', 'REF', 'ALT'] + list(pheno_df[pheno_df['Genetic group'] == gg]['Colony'])].to_csv('{}count_ref_{}.txt'.format(queen_geno_path, gg), header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f97615a7-8ef6-42e1-ada8-ea8dd5ec4a0f",
   "metadata": {},
   "source": [
    "## Queen genotype reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7ea86c7a-57de-4e1c-be93-3b885b6d4345",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_run_beethoven_script(script_fn):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --mem=40g\n",
    "#SBATCH --time=48:00:00\n",
    "#SBATCH --job-name=beethoven\n",
    "#SBATCH --output=%x_%j.out\n",
    "#SBATCH --error=%x_%j.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "dir_in=$1\n",
    "depth_file=$2\n",
    "count_ref_file=$3\n",
    "prefix=$4\n",
    "\n",
    "genoqueen_hom ${dir_in} ${depth_file} ${count_ref_file} 4 10 ${prefix} 1000\n",
    "\n",
    "echo DONE\"\"\"\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a585b191-f428-43ac-870d-673fde9501fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "script_fn = '{}beethoven.run'.format(script_path)\n",
    "write_run_beethoven_script(script_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "5b338e4c-a19b-4746-8981-ab550abdb938",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/scripts/tmp_scripts/beethoven.run'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "script_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "12de406c-f4a7-46a4-9a5b-e3f6121323f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gg = 'Caucasica'\n",
    "command = ['sbatch', script_fn, queen_geno_path, 'depth_{}.txt'.format(gg), 'count_ref_{}.txt'.format(gg), '{}_{}'.format(prefix, gg)]\n",
    "result = subprocess.run(command, cwd=queen_geno_path, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6d3ebda3-e357-4863-a874-3d495a671526",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hybrid\n",
      "Mellifera\n",
      "Ligustica_Carnica\n"
     ]
    }
   ],
   "source": [
    "for gg in set(pheno_df['Genetic group']):\n",
    "    if gg == 'Caucasica':\n",
    "        continue\n",
    "    print(gg)\n",
    "    command = ['sbatch', script_fn, queen_geno_path, 'depth_{}.txt'.format(gg), 'count_ref_{}.txt'.format(gg), '{}_{}'.format(prefix, gg)]\n",
    "    result = subprocess.run(command, cwd=queen_geno_path, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af8a7cd9-db02-4971-bb89-4b40a6586ac7",
   "metadata": {},
   "source": [
    "## Write PLINK .ped\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa044a26-e444-4390-993d-bdad83f8970c",
   "metadata": {},
   "source": [
    "### Split beethoven .prob files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "ca56b034-d5b7-44bb-9f04-ebcece1527e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'queen'\n",
    "plink_path = '{}data/queen_genotypes/hom_pop/'.format(working_path)\n",
    "populations = ['Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid']\n",
    "chunk_size = 25\n",
    "prob_threshold = 0.95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "155a4732-09fe-4e61-b5e1-8ccea375f6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to split .prob files into 3: AA, AR, RR --> easier to read probability of these genotypes simulateously then...\n",
    "def split_prob_file(\n",
    "    plink_path, prefix, pop):\n",
    "\n",
    "    prob_file = '{}{}_{}.prob'.format(plink_path, prefix, pop)\n",
    "    aa_file = '{}{}_{}_AA.prob'.format(plink_path, prefix, pop)\n",
    "    ar_file = '{}{}_{}_AR.prob'.format(plink_path, prefix, pop)\n",
    "    rr_file = '{}{}_{}_RR.prob'.format(plink_path, prefix, pop)\n",
    "    \n",
    "    out_aa = open(aa_file, 'w')\n",
    "    out_ar = open(ar_file, 'w')\n",
    "    out_rr = open(rr_file, 'w')\n",
    "    \n",
    "    with open(prob_file, 'r') as inf:\n",
    "        headers = inf.readline()\n",
    "        out_aa.write(headers.rstrip(',geno\\n') + '\\n')\n",
    "        out_ar.write(headers.rstrip(',geno\\n') + '\\n')\n",
    "        out_rr.write(headers.rstrip(',geno\\n') + '\\n')\n",
    "\n",
    "        for l in tqdm(inf):\n",
    "            split_line = l.rstrip().split(',')\n",
    "            geno = split_line[-1]\n",
    "            if geno == 'AA':\n",
    "                out_aa.write(','.join(split_line[:-1]) + '\\n')\n",
    "            elif geno == 'AR':\n",
    "                out_ar.write(','.join(split_line[:-1]) + '\\n')\n",
    "            elif geno == 'RR':\n",
    "                out_rr.write(','.join(split_line[:-1]) + '\\n')\n",
    "            else: \n",
    "                print('GENO: {}'.format(geno))\n",
    "                      \n",
    "    out_aa.close()\n",
    "    out_ar.close()\n",
    "    out_rr.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "52986b4b-f649-482c-8249-75333eb0acdc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20293116it [00:46, 437729.20it/s]\n",
      "20293116it [24:53, 13589.27it/s]\n",
      "20293116it [15:02, 22474.99it/s]\n",
      "20293116it [14:04, 24026.43it/s]\n"
     ]
    }
   ],
   "source": [
    "for pop in populations:\n",
    "    split_prob_file(plink_path, prefix, pop)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a1c8314-eed1-4f2d-9f95-1cfe832e031e",
   "metadata": {},
   "source": [
    "### Write .ped by population with probability threshold"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d2103cc-601d-402d-bdc9-59d1d5dc61ff",
   "metadata": {},
   "source": [
    "debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e2efca3-7743-42eb-8c6c-47bc221391eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_ped_woprob(\n",
    "    plink_path, prefix, pop, chunk_bs_ids, batch_j, chunk_size):\n",
    "    \n",
    "    # index of column with first bs_id (skipping metadata columns)\n",
    "    bgs_col_idx = 5\n",
    "    \n",
    "    bim_like = pd.read_csv('{}{}.bim_like'.format(plink_path, prefix), sep='\\t', header=None)\n",
    "    bgs_file = '{}{}_{}.bgs'.format(plink_path, prefix, pop)\n",
    "    snp_nr = len(bim_like)\n",
    "    \n",
    "    start_j = batch_j * chunk_size\n",
    "    ped_mat = np.full((len(chunk_bs_ids), 2 * snp_nr), '', dtype='U1')\n",
    "    \n",
    "    inf = open(bgs_file, 'r')\n",
    "\n",
    "    # check we have the same bs ids in header than in batch\n",
    "    header_bsids = inf.readline()\n",
    "    assert header_bsids.rstrip().split(',')[bgs_col_idx:][start_j: start_j + len(chunk_bs_ids)] == chunk_bs_ids, 'bs_id mismatch'\n",
    "\n",
    "    # iterate snps and colonies\n",
    "    for i in tqdm(range(snp_nr)):\n",
    "        ref_numbers = inf.readline().rstrip().split(',')[bgs_col_idx:]\n",
    "\n",
    "        for j in range(start_j, start_j + len(chunk_bs_ids)):\n",
    "            ref_nr = ref_numbers[j]\n",
    "\n",
    "            # homozygote reference\n",
    "            if (ref_nr == '2'):\n",
    "                ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 5]\n",
    "                ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 5]\n",
    "            # heterozygote\n",
    "            elif (ref_nr == '1'):\n",
    "                ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 5]\n",
    "                ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 4]\n",
    "            # homozygote alternative\n",
    "            elif (ref_nr == '0'):\n",
    "                ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 4]\n",
    "                ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 4]\n",
    "            # below prob_threshold\n",
    "            elif ref_nr in {'0', '1', '2'}:\n",
    "                ped_mat[j - start_j, 2 * i] = '0'\n",
    "                ped_mat[j - start_j, 2 * i + 1] = '0'\n",
    "            else: \n",
    "                print('ERROR')\n",
    "            \n",
    "    inf.close()\n",
    "\n",
    "    ped_df = pd.DataFrame({\n",
    "        'FID': ['0'] * len(chunk_bs_ids), \n",
    "        'IID': chunk_bs_ids,\n",
    "        'PID': ['0'] * len(chunk_bs_ids),\n",
    "        'MID': ['0'] * len(chunk_bs_ids),\n",
    "        'SEX': ['0'] * len(chunk_bs_ids), \n",
    "        'PHENOTYPE': ['-9'] * len(chunk_bs_ids)\n",
    "    })\n",
    "    ped_df = ped_df.join(pd.DataFrame(ped_mat))\n",
    "    ped_df.to_csv('{}{}_{}_{}.ped'.format(plink_path, prefix, pop, batch_j), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "979d063f-a4b4-4c1c-bf77-e523ba4fe12f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "queen_geno_path = '{}data/queen_genotypes/{}/'.format(working_path, prefix)\n",
    "plink_path = queen_geno_path\n",
    "populations = ['Ligustica_Carnica', 'Mellifera']\n",
    "populations = ['Ligustica_Carnica']\n",
    "prob_threshold = 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "27748959-0846-4879-bd30-790b12f453bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'Sonia_All'\n",
    "plink_path = '{}data/queen_genotypes/{}/'.format(working_path, prefix)\n",
    "pop = 'Ligustica_Carnica'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "043f5b24-5289-44cc-981c-28268693b5ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/Sonia_All/bsid_batch_Ligustica_Carnica.txt'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)\n",
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "batch_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "eb4d46b2-8346-4d98-84f6-cfcf40f0da8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/Sonia_All/bsid_batch_Ligustica_Carnica.txt'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a871939b-b2f3-4e10-9929-663a4dbf6d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_j = 0\n",
    "chunk_bs_ids = batch_bs_ids[batch_j]\n",
    "chunk_size = 25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b2a79415-a4c5-4e79-867f-759a71c1dd94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHROM</td>\n",
       "      <td>SNP_ID</td>\n",
       "      <td>GEN_DIST</td>\n",
       "      <td>POS</td>\n",
       "      <td>ALT</td>\n",
       "      <td>REF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831070</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831071</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225986</td>\n",
       "      <td>0</td>\n",
       "      <td>7225986</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831072</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831073</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831074</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6831075 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0           1         2        3    4    5\n",
       "0        CHROM      SNP_ID  GEN_DIST      POS  ALT  REF\n",
       "1            1      1:5671         0     5671    C    T\n",
       "2            1      1:5698         0     5698    C    T\n",
       "3            1      1:6621         0     6621    G    A\n",
       "4            1      1:6717         0     6717    T    C\n",
       "...        ...         ...       ...      ...  ...  ...\n",
       "6831070     16  16:7225911         0  7225911    T    C\n",
       "6831071     16  16:7225986         0  7225986    C    G\n",
       "6831072     16  16:7226186         0  7226186    C    G\n",
       "6831073     16  16:7226346         0  7226346    A    G\n",
       "6831074     16  16:7227034         0  7227034    A    G\n",
       "\n",
       "[6831075 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_like\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b2e2217a-92c0-4635-812f-699b958f68e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1824108/1242175706.py:4: DtypeWarning: Columns (0,2,3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  bim_like = pd.read_csv('{}{}.bim_like'.format(plink_path, prefix), sep='\\t', header=None)\n"
     ]
    }
   ],
   "source": [
    "# index of column with first bs_id (skipping metadata columns)\n",
    "bgs_col_idx = 5\n",
    "\n",
    "bim_like = pd.read_csv('{}{}.bim_like'.format(plink_path, prefix), sep='\\t', header=None)\n",
    "bgs_file = '{}{}_{}.bgs'.format(plink_path, prefix, pop)\n",
    "snp_nr = len(bim_like)\n",
    "\n",
    "start_j = batch_j * chunk_size\n",
    "ped_mat = np.full((len(chunk_bs_ids), 2 * snp_nr), '', dtype='U1')\n",
    "\n",
    "inf = open(bgs_file, 'r')\n",
    "\n",
    "# check we have the same bs ids in header than in batch\n",
    "header_bsids = inf.readline()\n",
    "assert header_bsids.rstrip().split(',')[bgs_col_idx:][start_j: start_j + len(chunk_bs_ids)] == chunk_bs_ids, 'bs_id mismatch'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5c027e6-9e8d-49f6-89e6-30f355a97c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate snps and colonies\n",
    "for i in tqdm(range(snp_nr)):\n",
    "    ref_numbers = inf.readline().rstrip().split(',')[bgs_col_idx:]\n",
    "\n",
    "    for j in range(start_j, start_j + len(chunk_bs_ids)):\n",
    "        ref_nr = ref_numbers[j]\n",
    "\n",
    "        # homozygote reference\n",
    "        if (ref_nr == '2'):\n",
    "            ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 5]\n",
    "            ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 5]\n",
    "        # heterozygote\n",
    "        elif (ref_nr == '1'):\n",
    "            ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 5]\n",
    "            ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 4]\n",
    "        # homozygote alternative\n",
    "        elif (ref_nr == '0'):\n",
    "            ped_mat[j - start_j, 2 * i] = bim_like.iloc[i, 4]\n",
    "            ped_mat[j - start_j, 2 * i + 1] = bim_like.iloc[i, 4]\n",
    "        # below prob_threshold\n",
    "        elif ref_nr in {'0', '1', '2'}:\n",
    "            ped_mat[j - start_j, 2 * i] = '0'\n",
    "            ped_mat[j - start_j, 2 * i + 1] = '0'\n",
    "        else: \n",
    "            print('ERROR')\n",
    "        \n",
    "inf.close()\n",
    "\n",
    "ped_df = pd.DataFrame({\n",
    "    'FID': ['0'] * len(chunk_bs_ids), \n",
    "    'IID': chunk_bs_ids,\n",
    "    'PID': ['0'] * len(chunk_bs_ids),\n",
    "    'MID': ['0'] * len(chunk_bs_ids),\n",
    "    'SEX': ['0'] * len(chunk_bs_ids), \n",
    "    'PHENOTYPE': ['-9'] * len(chunk_bs_ids)\n",
    "})\n",
    "ped_df = ped_df.join(pd.DataFrame(ped_mat))\n",
    "ped_df.to_csv('{}{}_{}_{}.ped'.format(plink_path, prefix, pop, batch_j), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "39a83de7-0970-4c8b-a0fe-37d6084aa2d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_write_ped_script(script_fn, array_str):\n",
    "    runstr=\"\"\"#!/bin/bash -l\n",
    "#SBATCH --array=ARRAY_STRING\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=50g\n",
    "#SBATCH --time=6:00:00\n",
    "#SBATCH --job-name=write_ped\n",
    "#SBATCH --output=%x_%A_%a.out\n",
    "#SBATCH --error=%x_%A_%a.err\n",
    "\n",
    "modulesld\n",
    "ebld\n",
    "module use /software/anaconda3/envs/eb/easybuild/modules/all\n",
    "conda activate planb\n",
    "\n",
    "# file listing batches of BeeStrong ids (bs_id) for parallel computing\n",
    "batch_file=$1\n",
    "chunk_size=$2\n",
    "plink_path=$3\n",
    "prefix=$4\n",
    "pop=$5\n",
    "prob_threshold=$6\n",
    "    \n",
    "batch_id=$SLURM_ARRAY_TASK_ID\n",
    "\n",
    "# getting bs_ids\n",
    "chunk_bs_ids=$(sed -n ${batch_id}'{p;q}' ${batch_file})\n",
    "    \n",
    "python ~/scripts/plan-b-omics/bin/write_ped.py --chunk_bs_ids ${chunk_bs_ids} --batch_id ${batch_id} --chunk_size ${chunk_size} --plink_path ${plink_path} --prefix ${prefix} --pop ${pop} --prob_threshold ${prob_threshold}\n",
    "\n",
    "echo DONE\"\"\".replace(\"ARRAY_STRING\", array_str)\n",
    "    with open(script_fn, 'w') as outf:\n",
    "        outf.write(runstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cd102e3d-e11f-4c35-8635-48a5200cb4a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/All/'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "queen_geno_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d1d8ad32-38d4-4bdd-82f8-7f35734ddb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "chunk_size = 25\n",
    "prefix = 'All'\n",
    "populations = ['Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid']\n",
    "\n",
    "# if threshold = 0, then the .prob files are not used\n",
    "prob_threshold = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "7d2f6a54-ffdf-463d-a579-c5c4d01129af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "703\n",
      "407\n"
     ]
    }
   ],
   "source": [
    "# write batch files\n",
    "for pop in populations:\n",
    "    batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)\n",
    "    bs_ids = pheno_df[pheno_df['Genetic group'] == pop]['Colony'].to_list()\n",
    "    print(len(bs_ids))\n",
    "    \n",
    "    batch_bs_ids = []\n",
    "    with open(batch_file, \"w\") as outf:\n",
    "        for i in range(0, len(bs_ids), chunk_size):\n",
    "            chunk = bs_ids[i:i+chunk_size]\n",
    "            batch_bs_ids.append(chunk)\n",
    "            outf.write(\",\".join(chunk) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76efdac3-8d56-4ef8-9fcd-b2b449a92029",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "821\n"
     ]
    }
   ],
   "source": [
    "### Sonia .bgs\n",
    "def get_bs_ids_from_bgs(bgs_file, bgs_col_idx=5):\n",
    "    with open(bgs_file, 'r') as inf:\n",
    "        header_bsids = inf.readline()\n",
    "    return header_bsids.rstrip().split(',')[bgs_col_idx:]\n",
    "\n",
    "def write_bim_like_from_bgs(bgs_file, plink_path, prefix):\n",
    "    '''\n",
    "    basically takes the first few columns of the .bgs file to create a  bim like file (that I am using to compute ped files)\n",
    "    '''\n",
    "    bim_df = pd.read_csv(bgs_file, sep=',', usecols=range(5))\n",
    "    bim_df['SNP_ID'] = bim_df['CHROM'].astype(str) + ':' + bim_df['POS'].astype(str)\n",
    "    bim_df['GEN_DIST'] = ['0'] * len(bim_df)\n",
    "    bim_df = bim_df[['CHROM', 'SNP_ID', 'GEN_DIST', 'POS', 'ALT', 'REF']]\n",
    "    bim_df.to_csv('{}{}.bim_like'.format(plink_path, prefix), index=False, header=None, sep='\\t')\n",
    "    \n",
    "chunk_size = 25\n",
    "\n",
    "prefix = 'Sonia_All'\n",
    "queen_geno_path = '{}data/queen_genotypes/{}/'.format(working_path, prefix)\n",
    "plink_path = queen_geno_path\n",
    "populations = ['Ligustica_Carnica', 'Mellifera']\n",
    "populations = ['Ligustica_Carnica']\n",
    "prob_threshold = 0\n",
    "\n",
    "# write batch files (for Sonia)\n",
    "for pop in populations:\n",
    "    batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)\n",
    "    bs_ids = get_bs_ids_from_bgs('{}{}_{}.bgs'.format(plink_path, prefix, pop), bgs_col_idx=5)\n",
    "    print(len(bs_ids))\n",
    "    \n",
    "    batch_bs_ids = []\n",
    "    with open(batch_file, \"w\") as outf:\n",
    "        for i in range(0, len(bs_ids), chunk_size):\n",
    "            chunk = bs_ids[i:i+chunk_size]\n",
    "            batch_bs_ids.append(chunk)\n",
    "            outf.write(\",\".join(chunk) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "250d47ad-d1cf-4eab-b195-d216396ad6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_bim_like_from_bgs('/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/Sonia_All/Sonia_All_Ligustica_Carnica.bgs', plink_path, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cfd82126-2442-4fb7-b99f-3c03f7c6e553",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/Sonia_All/bsid_batch_Ligustica_Carnica.txt'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop = 'Ligustica_Carnica'\n",
    "# pop = 'Caucasica'\n",
    "# pop = 'Mellifera'\n",
    "# pop = 'hybrid'\n",
    "\n",
    "batch_file = '{}bsid_batch_{}.txt'.format(plink_path, pop)\n",
    "with open(batch_file, \"r\") as inf:\n",
    "    batch_bs_ids = [x.rstrip().split(',') for x in inf.readlines()]\n",
    "batch_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "895eebd7-e16c-4a03-a6b2-d75ace9c61cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_str = '1-{}'.format(len(batch_bs_ids)) if len(batch_bs_ids) > 1 else '1'\n",
    "script_fn = '{}write_ped.run'.format(script_path)\n",
    "write_write_ped_script(script_fn, array_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "df9e5aaa-9cfd-46e5-bbb4-46535d13aad0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 1390200\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$script_fn\" \"$batch_file\" \"$chunk_size\" \"$plink_path\" \"$prefix\" \"$pop\" \"$prob_threshold\" \n",
    "cd $4\n",
    "sbatch $1 $2 $3 $4 $5 $6 $7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "325f2bb1-b447-4b78-81e9-19647c554478",
   "metadata": {},
   "source": [
    "cat the ped files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d34c9503-9bea-49c3-a46d-48b987b9bb74",
   "metadata": {},
   "source": [
    "### Write PLINK .map files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fc0666-f437-4124-b391-ca3b929179ae",
   "metadata": {},
   "source": [
    ".map file is also needed to convert ped to bed (only map + bed)\n",
    "\n",
    "CHR   SNP_ID   GEN_DIST   BP_POS\n",
    "\n",
    "basically bim wihout REF and ALT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ab5ab449-661f-462f-b1ec-929f47713227",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'All'\n",
    "prefix = 'Sonia_All'\n",
    "queen_geno_path = '{}data/queen_genotypes/{}/'.format(working_path, prefix)\n",
    "populations = ['Ligustica_Carnica']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d004a53f-cd6d-466c-8286-b35055261521",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_like = pd.read_csv('{}{}.bim_like'.format(queen_geno_path, prefix), sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "889979b2-8304-451b-b1bd-f378766bf7c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ligustica_Carnica\n"
     ]
    }
   ],
   "source": [
    "for pop in populations:\n",
    "    print(pop)\n",
    "    bim_like[[0, 1, 2, 3]].to_csv('{}{}_{}.map'.format(queen_geno_path, prefix, pop), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "745f6d19-05df-4131-9806-f110f649f672",
   "metadata": {},
   "source": [
    "convert to bed in CLI. Not sure how long it will take for convertion to bed...\n",
    "\n",
    "ml PLINK\n",
    "\n",
    "plink --file queen --make-bed --out queen --allow-extra-chr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41b26baa-caeb-48f0-8f42-ad6861dd824d",
   "metadata": {},
   "source": [
    "## GREML H2 analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3b01f831-ea9d-4772-9cb4-49a04cc2d934",
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_path = '/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/{}/'.format(plink_prefix)\n",
    "\n",
    "plink_prefix = 'hom_pop'\n",
    "\n",
    "plink_files = '{}{}'.format(plink_path, plink_prefix)\n",
    "\n",
    "gcta_path = '/cluster/raid/home/f80878961/bin/gcta-1.95.0-linux-kernel-3-x86_64/gcta64' # cluster version fails for GRM \n",
    "\n",
    "greml_path = '/cluster/raid/home/f80878961/beestrong/data/heritability_greml/{}/'.format(plink_prefix)\n",
    "\n",
    "\n",
    "lod = np.log(5e-06)\n",
    "\n",
    "populations = ['Mellifera', 'Ligustica_Carnica', 'hybrid']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3eac6a2-d2a6-4778-8304-0c736f3601a5",
   "metadata": {},
   "source": [
    "### Compute GRMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "62eae2e4-b172-4e88-81d7-3bdf078a23cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_prefix = 'hom_pop'\n",
    "plink_files = '{}{}'.format(plink_path, plink_prefix)\n",
    "\n",
    "pop = 'all_pops'\n",
    "maf = 0.01\n",
    "\n",
    "grm_path = '{}{}/GRM_maf{}/'.format(greml_path, pop, str(maf).replace('.', ''))\n",
    "if not os.path.exists(grm_path):\n",
    "    os.makedirs(grm_path)\n",
    "        \n",
    "grm_files = '{}{}'.format(grm_path, plink_prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3ba3c917-bc58-4995-8b5a-c0e3b147f90e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 14:24:25 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Options: \n",
      " \n",
      "--bfile /cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/hom_pop \n",
      "--make-grm \n",
      "--maf 0.05 \n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/all_pops/GRM_maf005/hom_pop \n",
      "\n",
      "Note: GRM is computed using the SNPs on the autosomes.\n",
      "Reading PLINK FAM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/hom_pop.fam]...\n",
      "1442 individuals to be included from FAM file.\n",
      "1442 individuals to be included. 0 males, 0 females, 1442 unknown.\n",
      "Reading PLINK BIM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/hom_pop.bim]...\n",
      "6764372 SNPs to be included from BIM file(s).\n",
      "Threshold to filter variants: MAF > 0.050000.\n",
      "Computing the genetic relationship matrix (GRM) v2 ...\n",
      "Subset 1/1, no. subject 1-1442\n",
      "  1442 samples, 6764372 markers, 1040403 GRM elements\n",
      "IDs for the GRM file have been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/all_pops/GRM_maf005/hom_pop.grm.id]\n",
      "Computing GRM...\n",
      "  100% finished in 275.8 sec\n",
      "6764372 SNPs have been processed.\n",
      "  Used 1801792 valid SNPs.\n",
      "The GRM computation is completed.\n",
      "Saving GRM...\n",
      "GRM has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/all_pops/GRM_maf005/hom_pop.grm.bin]\n",
      "Number of SNPs in each pair of individuals has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/all_pops/GRM_maf005/hom_pop.grm.N.bin]\n",
      "\n",
      "Analysis finished at 14:29:07 CET on Fri Jan 09 2026\n",
      "Overall computational time: 4 minutes 42 sec.\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$grm_path\" \"$gcta_path\" \"$plink_files\" \"$maf\" \"$grm_files\"\n",
    "mkdir -p $1\n",
    "cd $1\n",
    "$2 --bfile $3 --make-grm --maf $4 --out $5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "70ae597e-93bf-409c-90eb-197c2fa3c951",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greml_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "d4e409d1-6c83-4519-8ca3-5dfbf6c85491",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.05'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f86e379d-5199-4d51-83f0-5af3f3df85f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mellifera\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:25:05 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Options: \n",
      " \n",
      "--bfile /cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_Mellifera \n",
      "--make-grm \n",
      "--maf 0.05 \n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Mellifera/GRM_maf005/queen \n",
      "\n",
      "Note: GRM is computed using the SNPs on the autosomes.\n",
      "Reading PLINK FAM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_Mellifera.fam]...\n",
      "405 individuals to be included from FAM file.\n",
      "405 individuals to be included. 0 males, 0 females, 405 unknown.\n",
      "Reading PLINK BIM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_Mellifera.bim]...\n",
      "6764372 SNPs to be included from BIM file(s).\n",
      "Threshold to filter variants: MAF > 0.050000.\n",
      "Computing the genetic relationship matrix (GRM) v2 ...\n",
      "Subset 1/1, no. subject 1-405\n",
      "  405 samples, 6764372 markers, 82215 GRM elements\n",
      "IDs for the GRM file have been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Mellifera/GRM_maf005/queen.grm.id]\n",
      "Computing GRM...\n",
      "  100% finished in 26.2 sec\n",
      "6764372 SNPs have been processed.\n",
      "  Used 1553315 valid SNPs.\n",
      "The GRM computation is completed.\n",
      "Saving GRM...\n",
      "GRM has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Mellifera/GRM_maf005/queen.grm.bin]\n",
      "Number of SNPs in each pair of individuals has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Mellifera/GRM_maf005/queen.grm.N.bin]\n",
      "\n",
      "Analysis finished at 15:25:38 CET on Fri Jan 09 2026\n",
      "Overall computational time: 33.04 sec.\n",
      "\n",
      "Ligustica_Carnica\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:25:38 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Options: \n",
      " \n",
      "--bfile /cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_Ligustica_Carnica \n",
      "--make-grm \n",
      "--maf 0.05 \n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Ligustica_Carnica/GRM_maf005/queen \n",
      "\n",
      "Note: GRM is computed using the SNPs on the autosomes.\n",
      "Reading PLINK FAM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_Ligustica_Carnica.fam]...\n",
      "645 individuals to be included from FAM file.\n",
      "645 individuals to be included. 0 males, 0 females, 645 unknown.\n",
      "Reading PLINK BIM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_Ligustica_Carnica.bim]...\n",
      "6764372 SNPs to be included from BIM file(s).\n",
      "Threshold to filter variants: MAF > 0.050000.\n",
      "Computing the genetic relationship matrix (GRM) v2 ...\n",
      "Subset 1/1, no. subject 1-645\n",
      "  645 samples, 6764372 markers, 208335 GRM elements\n",
      "IDs for the GRM file have been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Ligustica_Carnica/GRM_maf005/queen.grm.id]\n",
      "Computing GRM...\n",
      "  100% finished in 61.5 sec\n",
      "6764372 SNPs have been processed.\n",
      "  Used 1718847 valid SNPs.\n",
      "The GRM computation is completed.\n",
      "Saving GRM...\n",
      "GRM has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Ligustica_Carnica/GRM_maf005/queen.grm.bin]\n",
      "Number of SNPs in each pair of individuals has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Ligustica_Carnica/GRM_maf005/queen.grm.N.bin]\n",
      "\n",
      "Analysis finished at 15:26:46 CET on Fri Jan 09 2026\n",
      "Overall computational time: 1 minute 8 sec.\n",
      "\n",
      "hybrid\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:26:46 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Options: \n",
      " \n",
      "--bfile /cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_hybrid \n",
      "--make-grm \n",
      "--maf 0.05 \n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen \n",
      "\n",
      "Note: GRM is computed using the SNPs on the autosomes.\n",
      "Reading PLINK FAM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_hybrid.fam]...\n",
      "372 individuals to be included from FAM file.\n",
      "372 individuals to be included. 0 males, 0 females, 372 unknown.\n",
      "Reading PLINK BIM file from [/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/queen_hybrid.bim]...\n",
      "6764372 SNPs to be included from BIM file(s).\n",
      "Threshold to filter variants: MAF > 0.050000.\n",
      "Computing the genetic relationship matrix (GRM) v2 ...\n",
      "Subset 1/1, no. subject 1-372\n",
      "  372 samples, 6764372 markers, 69378 GRM elements\n",
      "IDs for the GRM file have been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id]\n",
      "Computing GRM...\n",
      "  100% finished in 23.8 sec\n",
      "6764372 SNPs have been processed.\n",
      "  Used 1931679 valid SNPs.\n",
      "The GRM computation is completed.\n",
      "Saving GRM...\n",
      "GRM has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin]\n",
      "Number of SNPs in each pair of individuals has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.N.bin]\n",
      "\n",
      "Analysis finished at 15:27:17 CET on Fri Jan 09 2026\n",
      "Overall computational time: 30.44 sec.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "maf = 0.05\n",
    "plink_prefix = 'queen'\n",
    "for pop in populations:\n",
    "    print(pop)\n",
    "    plink_files = '{}{}_{}'.format(plink_path, plink_prefix, pop)\n",
    "    \n",
    "    grm_path = '{}{}/GRM_maf{}/'.format(greml_path, pop, str(maf).replace('.', ''))\n",
    "    if not os.path.exists(grm_path):\n",
    "        os.makedirs(grm_path)\n",
    "    grm_files = '{}{}'.format(grm_path, plink_prefix)\n",
    "    \n",
    "    command = [gcta_path, '--bfile', plink_files, '--make-grm', '--maf' , str(maf), '--out', grm_files]\n",
    "    result = subprocess.run(command, cwd=grm_path, capture_output=True, text=True)\n",
    "    print(result.stderr)\n",
    "    print(result.stdout) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c524df66-bbc8-45e5-accf-603516b1f24d",
   "metadata": {},
   "source": [
    "Create input for GREML analyses\n",
    "1. phenotype file .phen\n",
    "2. categorical variables .covar\n",
    "3. continuous variables .qcovar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "e4ed0eaf-e96a-493e-8d96-62ec7fe8346f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_pheno_file(h2_path, h2_name, gcta_df, phenotype):\n",
    "    fn = '{}{}.phen'.format(h2_path, h2_name)\n",
    "    gcta_df[['0', 'Colony', phenotype]].fillna('NA').to_csv(fn, header=None, sep='\\t', index=False)\n",
    "\n",
    "def write_covar_file(h2_path, h2_name, gcta_df, fixed_covar, tax_covar):\n",
    "\n",
    "    temp_gcta_df = gcta_df.copy()\n",
    "    \n",
    "    # replace abundance values by presence-absence for prevalence associations\n",
    "    temp_gcta_df[temp_gcta_df[tax_covar] >= lod] = int(1)\n",
    "    temp_gcta_df[temp_gcta_df[tax_covar] < lod] = int(0)\n",
    "\n",
    "    fn = '{}{}.covar'.format(h2_path, h2_name)\n",
    "    temp_gcta_df[['0', 'Colony'] + fixed_covar + tax_covar].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "\n",
    "def write_qcovar_file(h2_path, h2_name, gcta_df, fixed_qcovar, tax_qcovar):\n",
    "    fn = '{}{}.qcovar'.format(h2_path, h2_name)\n",
    "    gcta_df[['0', 'Colony'] + fixed_qcovar + tax_qcovar].to_csv(fn, header=None, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73d1d6ba-147a-43d3-8162-b3504c14a13b",
   "metadata": {},
   "source": [
    "Combine metadata and symbiosphere data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a2acee6c-ca2d-4f8e-9630-dbbf74d90dd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2233136/1970459003.py:9: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  fam_df = pd.read_csv('{}hom_pop.fam'.format(plink_path), header=None, sep='\\s')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Colony</th>\n",
       "      <th>0</th>\n",
       "      <th>date</th>\n",
       "      <th>Genetic group</th>\n",
       "      <th>percent_Ligustica_Carnica</th>\n",
       "      <th>percent_Mellifera</th>\n",
       "      <th>percent_Caucasica</th>\n",
       "      <th>bee_weigth</th>\n",
       "      <th>nbr_pho_varroa</th>\n",
       "      <th>nbr_pho_varroa_100bee</th>\n",
       "      <th>...</th>\n",
       "      <th>PC58</th>\n",
       "      <th>PC59</th>\n",
       "      <th>PC60</th>\n",
       "      <th>PC61</th>\n",
       "      <th>PC62</th>\n",
       "      <th>PC63</th>\n",
       "      <th>PC64</th>\n",
       "      <th>PC65</th>\n",
       "      <th>PC66</th>\n",
       "      <th>PC67</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>0</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.118259</td>\n",
       "      <td>1.369552e-04</td>\n",
       "      <td>0.881604</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.223545</td>\n",
       "      <td>0.611330</td>\n",
       "      <td>0.302651</td>\n",
       "      <td>0.131466</td>\n",
       "      <td>-0.060962</td>\n",
       "      <td>0.205536</td>\n",
       "      <td>-0.101299</td>\n",
       "      <td>-0.252870</td>\n",
       "      <td>0.160926</td>\n",
       "      <td>0.010749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>0</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.109495</td>\n",
       "      <td>9.200239e-05</td>\n",
       "      <td>0.890413</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.083837</td>\n",
       "      <td>0.319977</td>\n",
       "      <td>-0.277766</td>\n",
       "      <td>0.184884</td>\n",
       "      <td>-0.267314</td>\n",
       "      <td>-0.058724</td>\n",
       "      <td>0.008251</td>\n",
       "      <td>-0.107998</td>\n",
       "      <td>-0.273400</td>\n",
       "      <td>0.016922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.047143</td>\n",
       "      <td>1.376441e-05</td>\n",
       "      <td>0.952843</td>\n",
       "      <td>63.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.114503</td>\n",
       "      <td>-0.171535</td>\n",
       "      <td>-0.115155</td>\n",
       "      <td>0.327699</td>\n",
       "      <td>0.127601</td>\n",
       "      <td>-0.413191</td>\n",
       "      <td>0.045553</td>\n",
       "      <td>-0.106437</td>\n",
       "      <td>0.230031</td>\n",
       "      <td>0.392338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.015421</td>\n",
       "      <td>1.019360e-05</td>\n",
       "      <td>0.984569</td>\n",
       "      <td>52.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.256039</td>\n",
       "      <td>-0.478161</td>\n",
       "      <td>0.069915</td>\n",
       "      <td>0.570721</td>\n",
       "      <td>-0.081693</td>\n",
       "      <td>0.135746</td>\n",
       "      <td>0.021538</td>\n",
       "      <td>-0.012105</td>\n",
       "      <td>-0.083352</td>\n",
       "      <td>-0.242981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>0.005619</td>\n",
       "      <td>4.687815e-08</td>\n",
       "      <td>0.994381</td>\n",
       "      <td>60.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.114564</td>\n",
       "      <td>-0.172282</td>\n",
       "      <td>0.017520</td>\n",
       "      <td>0.432423</td>\n",
       "      <td>-0.139529</td>\n",
       "      <td>-0.082052</td>\n",
       "      <td>0.020244</td>\n",
       "      <td>-0.335893</td>\n",
       "      <td>0.113096</td>\n",
       "      <td>0.116470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.418368</td>\n",
       "      <td>5.687723e-01</td>\n",
       "      <td>0.012860</td>\n",
       "      <td>47.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.185675</td>\n",
       "      <td>-0.057137</td>\n",
       "      <td>-0.014365</td>\n",
       "      <td>0.275159</td>\n",
       "      <td>0.384569</td>\n",
       "      <td>-0.180188</td>\n",
       "      <td>0.010630</td>\n",
       "      <td>0.194236</td>\n",
       "      <td>0.054757</td>\n",
       "      <td>-0.261595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.044593</td>\n",
       "      <td>7.929777e-01</td>\n",
       "      <td>0.162429</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.049772</td>\n",
       "      <td>0.183526</td>\n",
       "      <td>-0.083507</td>\n",
       "      <td>-0.102922</td>\n",
       "      <td>-0.160323</td>\n",
       "      <td>0.213182</td>\n",
       "      <td>0.103019</td>\n",
       "      <td>-0.118458</td>\n",
       "      <td>-0.101629</td>\n",
       "      <td>-0.220722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.113869</td>\n",
       "      <td>5.741510e-01</td>\n",
       "      <td>0.311980</td>\n",
       "      <td>39.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.277892</td>\n",
       "      <td>-0.524994</td>\n",
       "      <td>0.025260</td>\n",
       "      <td>0.413383</td>\n",
       "      <td>-0.185017</td>\n",
       "      <td>-0.411371</td>\n",
       "      <td>-0.207086</td>\n",
       "      <td>0.299036</td>\n",
       "      <td>0.171580</td>\n",
       "      <td>0.535910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.554205</td>\n",
       "      <td>4.213804e-01</td>\n",
       "      <td>0.024415</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.616182</td>\n",
       "      <td>0.201392</td>\n",
       "      <td>-0.425561</td>\n",
       "      <td>0.144140</td>\n",
       "      <td>-0.430934</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>-0.186331</td>\n",
       "      <td>0.307234</td>\n",
       "      <td>0.191650</td>\n",
       "      <td>0.425309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>hybrid</td>\n",
       "      <td>0.378583</td>\n",
       "      <td>2.230246e-01</td>\n",
       "      <td>0.398393</td>\n",
       "      <td>56.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.389531</td>\n",
       "      <td>-0.027292</td>\n",
       "      <td>0.023136</td>\n",
       "      <td>0.123168</td>\n",
       "      <td>0.229125</td>\n",
       "      <td>-0.110197</td>\n",
       "      <td>0.135449</td>\n",
       "      <td>-0.078937</td>\n",
       "      <td>-0.107581</td>\n",
       "      <td>-0.052356</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 179 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Colony  0        date Genetic group  percent_Ligustica_Carnica  \\\n",
       "0     BS16_0060  0  06/07/2016     Caucasica                   0.118259   \n",
       "1     BS16_0061  0  06/07/2016     Caucasica                   0.109495   \n",
       "2     BS16_0162  0  31/08/2016     Caucasica                   0.047143   \n",
       "3     BS16_0163  0  31/08/2016     Caucasica                   0.015421   \n",
       "4     BS16_0164  0  31/08/2016     Caucasica                   0.005619   \n",
       "...         ... ..         ...           ...                        ...   \n",
       "1437  BS17_0506  0  01/08/2017        hybrid                   0.418368   \n",
       "1438  BS17_0511  0  01/08/2017        hybrid                   0.044593   \n",
       "1439  BS17_0512  0  01/08/2017        hybrid                   0.113869   \n",
       "1440  BS17_0513  0  01/08/2017        hybrid                   0.554205   \n",
       "1441  BS17_0517  0  01/08/2017        hybrid                   0.378583   \n",
       "\n",
       "      percent_Mellifera  percent_Caucasica  bee_weigth  nbr_pho_varroa  \\\n",
       "0          1.369552e-04           0.881604        51.0             0.0   \n",
       "1          9.200239e-05           0.890413        52.0             1.0   \n",
       "2          1.376441e-05           0.952843        63.0            28.0   \n",
       "3          1.019360e-05           0.984569        52.0             2.0   \n",
       "4          4.687815e-08           0.994381        60.0             9.0   \n",
       "...                 ...                ...         ...             ...   \n",
       "1437       5.687723e-01           0.012860        47.5             1.0   \n",
       "1438       7.929777e-01           0.162429        41.0             0.0   \n",
       "1439       5.741510e-01           0.311980        39.2             0.0   \n",
       "1440       4.213804e-01           0.024415        43.0             0.0   \n",
       "1441       2.230246e-01           0.398393        56.0             5.0   \n",
       "\n",
       "      nbr_pho_varroa_100bee  ...      PC58      PC59      PC60      PC61  \\\n",
       "0                       0.0  ... -0.223545  0.611330  0.302651  0.131466   \n",
       "1                       0.0  ...  1.083837  0.319977 -0.277766  0.184884   \n",
       "2                       6.0  ...  0.114503 -0.171535 -0.115155  0.327699   \n",
       "3                       1.0  ... -0.256039 -0.478161  0.069915  0.570721   \n",
       "4                       2.0  ... -0.114564 -0.172282  0.017520  0.432423   \n",
       "...                     ...  ...       ...       ...       ...       ...   \n",
       "1437                    0.0  ... -0.185675 -0.057137 -0.014365  0.275159   \n",
       "1438                    0.0  ... -0.049772  0.183526 -0.083507 -0.102922   \n",
       "1439                    0.0  ...  0.277892 -0.524994  0.025260  0.413383   \n",
       "1440                    0.0  ...  0.616182  0.201392 -0.425561  0.144140   \n",
       "1441                    1.0  ... -0.389531 -0.027292  0.023136  0.123168   \n",
       "\n",
       "          PC62      PC63      PC64      PC65      PC66      PC67  \n",
       "0    -0.060962  0.205536 -0.101299 -0.252870  0.160926  0.010749  \n",
       "1    -0.267314 -0.058724  0.008251 -0.107998 -0.273400  0.016922  \n",
       "2     0.127601 -0.413191  0.045553 -0.106437  0.230031  0.392338  \n",
       "3    -0.081693  0.135746  0.021538 -0.012105 -0.083352 -0.242981  \n",
       "4    -0.139529 -0.082052  0.020244 -0.335893  0.113096  0.116470  \n",
       "...        ...       ...       ...       ...       ...       ...  \n",
       "1437  0.384569 -0.180188  0.010630  0.194236  0.054757 -0.261595  \n",
       "1438 -0.160323  0.213182  0.103019 -0.118458 -0.101629 -0.220722  \n",
       "1439 -0.185017 -0.411371 -0.207086  0.299036  0.171580  0.535910  \n",
       "1440 -0.430934  0.012346 -0.186331  0.307234  0.191650  0.425309  \n",
       "1441  0.229125 -0.110197  0.135449 -0.078937 -0.107581 -0.052356  \n",
       "\n",
       "[1442 rows x 179 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_df = pd.read_csv('{}BeeStrong_metadata.csv'.format(tlf_path))\n",
    "abund_df = pd.read_csv('{}Core_Family_LRA.csv'.format(tlf_path))\n",
    "pca_df = pd.read_csv('{}PCA_Core_Family_LRA.csv'.format(tlf_path))\n",
    "\n",
    "gcta_df = pheno_df.merge(abund_df, on='Colony').merge(pca_df.rename(columns={'Unnamed: 0': 'Colony'}), on='Colony')\n",
    "gcta_df.insert(0, '0', [0] * len(gcta_df))\n",
    "\n",
    "# reorder pheno df according to fam_df (order disrupted due to per-population genotype inference)\n",
    "fam_df = pd.read_csv('{}hom_pop.fam'.format(plink_path), header=None, sep='\\s') \n",
    "gcta_df = gcta_df.set_index(\"Colony\").loc[fam_df[1].to_list()].reset_index()\n",
    "\n",
    "# check consistent colony order between pheno df and plink\n",
    "bs_ids = gcta_df['Colony'].to_list()\n",
    "assert fam_df[1].to_list() == bs_ids, 'messed up colony order'\n",
    "\n",
    "gcta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6bedc43f-27b7-4524-9981-a3bbedd46804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Caucasica', 'Ligustica_Carnica', 'Mellifera', 'hybrid'}"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(gcta_df['Genetic group'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "c731e14a-7d1d-4321-af20-16c3b3531faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# maf = '001'\n",
    "maf = '005'\n",
    "\n",
    "pop = 'Mellifera'\n",
    "pop = 'Ligustica_Carnica'\n",
    "pop = 'hybrid'\n",
    "# pop = 'all_pops'\n",
    "\n",
    "if pop == 'all_pops':\n",
    "    pop_gcta_df = gcta_df\n",
    "    plink_prefix = 'hom_pop'\n",
    "else:\n",
    "    pop_gcta_df = gcta_df[gcta_df['Genetic group'] == pop]\n",
    "    plink_prefix = 'queen'\n",
    "\n",
    "default_covar = ['Apiary', 'Year']\n",
    "default_qcovar = ['Day of Year']\n",
    "vaf_prev = [\n",
    "    'Oscillospiraceae',\n",
    "    'Corynebacteriaceae',\n",
    "    'Cannabaceae',\n",
    "    'Aeromonadaceae',\n",
    "    'Weeksellaceae',\n",
    "    'Araliaceae',\n",
    "    'Actinomycetaceae',\n",
    "    'Trypanosomatidae',\n",
    "    'Saccharomycodaceae', # v_pho specific\n",
    "]\n",
    "vaf_abund = [\n",
    "    'Bifidobacteriaceae',\n",
    "    'Lactobacillaceae',\n",
    "    'Pseudomonadaceae',   \n",
    "    'Debaryomycetaceae',\n",
    "    'Bartonellaceae',  \n",
    "    'Morganellaceae',\n",
    "    'Rosaceae', # v_pho specific\n",
    "    'Microbacteriaceae',  # v_pho specific\n",
    "    'Enterobacteriaceae',  # v_pho specific\n",
    "    'Nosematidae' # v_pho specific\n",
    "]\n",
    "\n",
    "top5vaf_prev = [\n",
    "   'Oscillospiraceae',\n",
    "    'Cannabaceae',\n",
    "    'Trypanosomatidae'\n",
    "]\n",
    "top5vaf_abund = [\n",
    "    'Bartonellaceae',  \n",
    "    'Morganellaceae'\n",
    "]\n",
    "\n",
    "name2settings = {\n",
    "    'default_v_meta': ('log_v_meta', default_covar, [], default_qcovar, []),\n",
    "    'default_v_pho': ('log_v_pho', default_covar, [], default_qcovar, []),\n",
    "    'default_DMR': ('DMR', default_covar, [], default_qcovar, []),\n",
    "    'default_Recapping': ('Recapping', default_covar, [], default_qcovar, []),\n",
    "    #'VAF_v_meta': ('log_v_meta', default_covar, vaf_prev, default_qcovar, vaf_abund),\n",
    "    #'VAF_v_pho': ('log_v_pho', default_covar, vaf_prev, default_qcovar, vaf_abund),\n",
    "    #'top5VAF_v_meta': ('log_v_meta', default_covar, top5vaf_prev, default_qcovar, top5vaf_abund),\n",
    "    #'top5VAF_v_pho': ('log_v_pho', default_covar, top5vaf_prev, default_qcovar, top5vaf_abund),\n",
    "    #'1PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 2)]),\n",
    "    #'1PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 2)]),\n",
    "    #'1PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 2)]),\n",
    "    #'1PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 2)]),\n",
    "    #'2PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 3)]),\n",
    "    #'2PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 3)]),\n",
    "    #'2PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 3)]),\n",
    "    #'2PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 3)]),\n",
    "    #'3PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 4)]),\n",
    "    #'3PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 4)]),\n",
    "    #'3PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 4)]),\n",
    "    #'3PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 4)]),\n",
    "    '4PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 5)]),\n",
    "    '4PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 5)]),\n",
    "    '4PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 5)]),\n",
    "    '4PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 5)]),\n",
    "    # '5PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 6)]),\n",
    "    # '5PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 6)]),\n",
    "    # '5PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 6)]),\n",
    "    # '5PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 6)]),\n",
    "    # '10PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 11)]),\n",
    "    # '10PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 11)]),\n",
    "    # '10PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 11)]),\n",
    "    # '10PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 11)]),\n",
    "    # '15PC_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 16)]),\n",
    "    # '15PC_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 16)]),\n",
    "    # '15PC_DMR': ('DMR', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 16)]),\n",
    "    # '15PC_Recapping': ('Recapping', default_covar, [], default_qcovar, ['PC{}'.format(x) for x in range(1, 16)]),\n",
    "    # 'PC5_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC5']), # 5, 1, 4, are the most associated with Varroa\n",
    "    # 'PC5_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC5']),\n",
    "    # 'PC145_v_meta': ('log_v_meta', default_covar, [], default_qcovar, ['PC1', 'PC4', 'PC5']),\n",
    "    # 'PC145_v_pho': ('log_v_pho', default_covar, [], default_qcovar, ['PC1', 'PC4', 'PC5']),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "9d707649-a150-4358-b302-c2a53aadce0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/Ligustica_Carnica/GRM_maf005/queen'"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grm_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "e7b13821-a75f-4ccc-b364-6fda41ceb17a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "default_v_meta\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:37 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.phen].\n",
      "Non-missing phenotypes of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.qcovar].\n",
      "1 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.covar].\n",
      "\n",
      "1 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "372 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "372 observations, 87 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.998379 0.903929\n",
      "logL: -223.86\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-214.18\t0.94199\t0.73502\t\n",
      "2\t-209.06\t0.91513\t0.65515\t\n",
      "3\t-207.31\t0.90230\t0.60876\t\n",
      "4\t-206.62\t0.89691\t0.57907\t\n",
      "5\t-206.32\t0.89282\t0.51541\t\n",
      "6\t-206.08\t0.91226\t0.50787\t\n",
      "7\t-206.07\t0.91943\t0.50372\t\n",
      "8\t-206.07\t0.92252\t0.50192\t\n",
      "9\t-206.07\t0.92386\t0.50114\t\n",
      "10\t-206.07\t0.92444\t0.50081\t\n",
      "11\t-206.07\t0.92469\t0.50066\t\n",
      "12\t-206.07\t0.92480\t0.50060\t\n",
      "13\t-206.07\t0.92484\t0.50057\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.35458\n",
      "logL: -238.47856\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-210.70\t1.24392\t\n",
      "2\t-208.63\t1.18271\t\n",
      "3\t-207.82\t1.14577\t\n",
      "4\t-207.48\t1.07184\t\n",
      "5\t-207.20\t1.07630\t\n",
      "6\t-207.20\t1.07632\t\n",
      "7\t-207.20\t1.07632\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.924843\t0.449088\n",
      "V(e)\t0.500575\t0.260348\n",
      "Vp\t1.425417\t0.220530\n",
      "V(G)/Vp\t0.648822\t0.225940\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "2.016796e-01\t-1.104135e-01\t\n",
      "-1.104135e-01\t6.778099e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_meta/default_v_meta.hsq].\n",
      "\n",
      "Analysis finished at 15:29:38 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.19 sec.\n",
      "\n",
      "default_v_pho\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:38 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.phen].\n",
      "Non-missing phenotypes of 349 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.qcovar].\n",
      "1 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.covar].\n",
      "\n",
      "1 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "349 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "349 observations, 80 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values:  0.74376 0.666304\n",
      "logL: -170.017\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-159.83\t0.74078\t0.50567\t\n",
      "2\t-154.20\t0.71556\t0.44700\t\n",
      "3\t-152.39\t0.69637\t0.41734\t\n",
      "4\t-151.68\t0.68305\t0.40034\t\n",
      "5\t-151.38\t0.65418\t0.36751\t\n",
      "6\t-151.14\t0.65345\t0.37084\t\n",
      "7\t-151.13\t0.65384\t0.37062\t\n",
      "8\t-151.13\t0.65380\t0.37065\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.00466\n",
      "logL: -185.81501\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-158.96\t0.92290\t\n",
      "2\t-157.02\t0.87763\t\n",
      "3\t-156.26\t0.85030\t\n",
      "4\t-155.94\t0.79559\t\n",
      "5\t-155.69\t0.79889\t\n",
      "6\t-155.69\t0.79890\t\n",
      "7\t-155.69\t0.79890\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.653797\t0.263715\n",
      "V(e)\t0.370649\t0.153433\n",
      "Vp\t1.024446\t0.139177\n",
      "V(G)/Vp\t0.638196\t0.184471\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "6.954560e-02\t-3.685858e-02\t\n",
      "-3.685858e-02\t2.354173e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_v_pho/default_v_pho.hsq].\n",
      "\n",
      "Analysis finished at 15:29:38 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.20 sec.\n",
      "\n",
      "default_DMR\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:38 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.phen].\n",
      "Non-missing phenotypes of 343 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.qcovar].\n",
      "1 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.covar].\n",
      "\n",
      "1 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "343 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "343 observations, 79 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.0195642 0.0207123\n",
      "logL: 255.019\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t256.30\t0.01280\t0.02530\t\n",
      "2\t257.51\t0.00846\t0.02845\t\n",
      "3\t258.10\t0.00599\t0.03039\t\n",
      "4\t258.35\t0.00154\t0.03418\t\n",
      "5\t258.56\t0.00175\t0.03416\t\n",
      "6\t258.56\t0.00182\t0.03412\t\n",
      "7\t258.56\t0.00184\t0.03411\t\n",
      "8\t258.56\t0.00184\t0.03410\t\n",
      "9\t258.56\t0.00184\t0.03410\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.03593\n",
      "logL: 258.22042\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t258.51\t0.03574\t\n",
      "2\t258.52\t0.03534\t\n",
      "3\t258.53\t0.03535\t\n",
      "4\t258.53\t0.03535\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.001844\t0.006261\n",
      "V(e)\t0.034103\t0.005080\n",
      "Vp\t0.035948\t0.003789\n",
      "V(G)/Vp\t0.051304\t0.171057\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "3.919742e-05\t-2.532547e-05\t\n",
      "-2.532547e-05\t2.580643e-05\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_DMR/default_DMR.hsq].\n",
      "\n",
      "Analysis finished at 15:29:38 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.10 sec.\n",
      "\n",
      "default_Recapping\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:38 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.phen].\n",
      "Non-missing phenotypes of 350 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.qcovar].\n",
      "1 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.covar].\n",
      "\n",
      "1 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "350 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "350 observations, 80 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values:  1.11273 0.959556\n",
      "logL: -215.109\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-198.31\t1.33359\t0.45096\t\n",
      "2\t-185.47\t1.30529\t0.34545\t\n",
      "3\t-182.55\t1.27521\t0.29940\t\n",
      "4\t-181.48\t1.25332\t0.27475\t\n",
      "5\t-181.04\t1.23825\t0.26022\t\n",
      "6\t-180.85\t1.20575\t0.23158\t\n",
      "7\t-180.69\t1.20612\t0.23359\t\n",
      "8\t-180.69\t1.20610\t0.23361\t\n",
      "9\t-180.69\t1.20610\t0.23361\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.37803\n",
      "logL: -233.88258\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-193.83\t1.21853\t\n",
      "2\t-190.15\t1.13838\t\n",
      "3\t-188.84\t1.09209\t\n",
      "4\t-188.30\t1.06353\t\n",
      "5\t-188.07\t1.00561\t\n",
      "6\t-187.89\t1.00859\t\n",
      "7\t-187.89\t1.00860\t\n",
      "8\t-187.89\t1.00860\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t1.206103\t0.360139\n",
      "V(e)\t0.233611\t0.193759\n",
      "Vp\t1.439714\t0.196216\n",
      "V(G)/Vp\t0.837738\t0.151002\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "1.297000e-01\t-6.437096e-02\t\n",
      "-6.437096e-02\t3.754271e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/default_Recapping/default_Recapping.hsq].\n",
      "\n",
      "Analysis finished at 15:29:38 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.14 sec.\n",
      "\n",
      "4PC_v_meta\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:39 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.phen].\n",
      "Non-missing phenotypes of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.qcovar].\n",
      "5 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.covar].\n",
      "\n",
      "5 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "372 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "372 observations, 91 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values:  1.00219 0.909588\n",
      "logL: -235.071\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-225.89\t0.96777\t0.72862\t\n",
      "2\t-220.91\t0.94980\t0.64368\t\n",
      "3\t-219.21\t0.94168\t0.59456\t\n",
      "4\t-218.53\t0.93911\t0.56324\t\n",
      "5\t-218.24\t0.94069\t0.49629\t\n",
      "6\t-218.00\t0.96004\t0.48884\t\n",
      "7\t-218.00\t0.96632\t0.48522\t\n",
      "8\t-218.00\t0.96873\t0.48383\t\n",
      "9\t-218.00\t0.96966\t0.48329\t\n",
      "10\t-218.00\t0.97001\t0.48309\t\n",
      "11\t-218.00\t0.97015\t0.48301\t\n",
      "12\t-218.00\t0.97020\t0.48298\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.37581\n",
      "logL: -249.47406\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-222.99\t1.26055\t\n",
      "2\t-220.85\t1.19716\t\n",
      "3\t-220.02\t1.15901\t\n",
      "4\t-219.67\t1.08279\t\n",
      "5\t-219.40\t1.08747\t\n",
      "6\t-219.39\t1.08749\t\n",
      "7\t-219.39\t1.08749\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.970201\t0.453567\n",
      "V(e)\t0.482977\t0.261598\n",
      "Vp\t1.453177\t0.224203\n",
      "V(G)/Vp\t0.667641\t0.220753\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "2.057227e-01\t-1.119446e-01\t\n",
      "-1.119446e-01\t6.843347e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_meta/4PC_v_meta.hsq].\n",
      "\n",
      "Analysis finished at 15:29:39 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.18 sec.\n",
      "\n",
      "4PC_v_pho\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:39 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.phen].\n",
      "Non-missing phenotypes of 349 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.qcovar].\n",
      "5 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.covar].\n",
      "\n",
      "5 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "349 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "349 observations, 84 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.746717 0.672319\n",
      "logL: -181.98\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-172.45\t0.72971\t0.52240\t\n",
      "2\t-167.14\t0.70164\t0.46556\t\n",
      "3\t-165.40\t0.68157\t0.43637\t\n",
      "4\t-164.71\t0.66794\t0.41949\t\n",
      "5\t-164.42\t0.63874\t0.38667\t\n",
      "6\t-164.19\t0.63836\t0.38979\t\n",
      "7\t-164.19\t0.63869\t0.38961\t\n",
      "8\t-164.19\t0.63865\t0.38963\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.02112\n",
      "logL: -197.21990\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-171.64\t0.93552\t\n",
      "2\t-169.62\t0.88845\t\n",
      "3\t-168.84\t0.86013\t\n",
      "4\t-168.51\t0.80353\t\n",
      "5\t-168.25\t0.80701\t\n",
      "6\t-168.25\t0.80703\t\n",
      "7\t-168.25\t0.80703\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.638647\t0.269962\n",
      "V(e)\t0.389628\t0.158129\n",
      "Vp\t1.028276\t0.141498\n",
      "V(G)/Vp\t0.621086\t0.190653\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "7.287946e-02\t-3.893127e-02\t\n",
      "-3.893127e-02\t2.500488e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_v_pho/4PC_v_pho.hsq].\n",
      "\n",
      "Analysis finished at 15:29:39 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.20 sec.\n",
      "\n",
      "4PC_DMR\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:39 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.phen].\n",
      "Non-missing phenotypes of 343 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.qcovar].\n",
      "5 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.covar].\n",
      "\n",
      "5 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "343 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "343 observations, 83 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.0195906 0.0207001\n",
      "logL: 237.199\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t238.48\t0.01323\t0.02511\t\n",
      "2\t239.57\t0.00923\t0.02808\t\n",
      "3\t240.10\t0.00695\t0.02992\t\n",
      "4\t240.32\t0.00283\t0.03353\t\n",
      "5\t240.49\t0.00330\t0.03333\t\n",
      "6\t240.50\t0.00336\t0.03329\t\n",
      "7\t240.50\t0.00336\t0.03329\t\n",
      "8\t240.50\t0.00336\t0.03329\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.03611\n",
      "logL: 240.11155\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t240.35\t0.03593\t\n",
      "2\t240.36\t0.03554\t\n",
      "3\t240.36\t0.03555\t\n",
      "4\t240.36\t0.03555\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.003361\t0.007070\n",
      "V(e)\t0.033289\t0.005427\n",
      "Vp\t0.036650\t0.004066\n",
      "V(G)/Vp\t0.091707\t0.186516\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "4.998908e-05\t-3.145425e-05\t\n",
      "-3.145425e-05\t2.945029e-05\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_DMR/4PC_DMR.hsq].\n",
      "\n",
      "Analysis finished at 15:29:39 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.10 sec.\n",
      "\n",
      "4PC_Recapping\n",
      "\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:29:40 CET on Fri Jan 09 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "372 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "GRM for 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/GRM_maf005/queen.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.phen].\n",
      "Non-missing phenotypes of 350 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.qcovar].\n",
      "5 quantitative covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.covar].\n",
      "2 discrete covariate(s) of 372 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.covar].\n",
      "\n",
      "5 quantitative variable(s) included as covariate(s).\n",
      "2 discrete variable(s) included as covariate(s).\n",
      "350 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "350 observations, 84 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values:  1.11528 0.962868\n",
      "logL: -225.137\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-208.65\t1.35869\t0.43006\t\n",
      "2\t-195.43\t1.32651\t0.32695\t\n",
      "3\t-192.53\t1.29365\t0.28265\t\n",
      "4\t-191.47\t1.26995\t0.25916\t\n",
      "5\t-191.03\t1.25364\t0.24540\t\n",
      "6\t-190.84\t1.21847\t0.21843\t\n",
      "7\t-190.69\t1.21856\t0.22059\t\n",
      "8\t-190.69\t1.21859\t0.22058\t\n",
      "9\t-190.69\t1.21859\t0.22058\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.39637\n",
      "logL: -243.94485\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-204.95\t1.22662\t\n",
      "2\t-201.01\t1.14275\t\n",
      "3\t-199.61\t1.09465\t\n",
      "4\t-199.05\t1.06507\t\n",
      "5\t-198.81\t1.00524\t\n",
      "6\t-198.62\t1.00842\t\n",
      "7\t-198.62\t1.00843\t\n",
      "8\t-198.62\t1.00843\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t1.218591\t0.360156\n",
      "V(e)\t0.220582\t0.193205\n",
      "Vp\t1.439173\t0.196705\n",
      "V(G)/Vp\t0.846730\t0.149781\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "1.297124e-01\t-6.417383e-02\t\n",
      "-6.417383e-02\t3.732800e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/hybrid/005/4PC_Recapping/4PC_Recapping.hsq].\n",
      "\n",
      "Analysis finished at 15:29:40 CET on Fri Jan 09 2026\n",
      "Overall computational time: 0.14 sec.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "grm_path = '{}{}/GRM_maf{}/'.format(greml_path, pop, maf)\n",
    "grm_files = '{}{}'.format(grm_path, plink_prefix)\n",
    "\n",
    "for h2_name, (phenotype, fixed_covar, tax_covar, fixed_qcovar, tax_qcovar) in name2settings.items():\n",
    "    print(h2_name)\n",
    "    h2_path = '{}{}/{}/{}/'.format(greml_path, pop, maf, h2_name)\n",
    "\n",
    "    if not os.path.exists(h2_path):\n",
    "        os.makedirs(h2_path)\n",
    "        \n",
    "    write_pheno_file(h2_path, h2_name, pop_gcta_df, phenotype)\n",
    "    write_covar_file(h2_path, h2_name, pop_gcta_df, fixed_covar, tax_covar)\n",
    "    write_qcovar_file(h2_path, h2_name, pop_gcta_df, fixed_qcovar, tax_qcovar)\n",
    "\n",
    "    command = [gcta_path, '--reml', '--grm',  grm_files , '--pheno', '{}{}.phen'.format(h2_path, h2_name), '--covar', '{}{}.covar'.format(h2_path, h2_name), \n",
    "           '--qcovar', '{}{}.qcovar'.format(h2_path, h2_name), '--out', '{}{}'.format(h2_path, h2_name)]\n",
    "    result = subprocess.run(command, cwd=greml_path, capture_output=True, text=True)\n",
    "    print(result.stderr)\n",
    "    print(result.stdout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "191d6b4c-803d-49a9-b90f-4e4aa5eb3ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "maf = '001'\n",
    "maf = '005'\n",
    "\n",
    "pop = 'Mellifera'\n",
    "# pop = 'Ligustica_Carnica'\n",
    "#pop = 'hybrid'\n",
    "# pop = 'all_pops'\n",
    "\n",
    "phenotypes = ['v_meta', 'v_pho', 'Recapping', 'DMR']\n",
    "\n",
    "# settings = ['default', 'VAF', 'top5VAF', '1PC', '2PC', '3PC', '4PC', '5PC', '10PC', '15PC', 'PC5', 'PC145']\n",
    "settings = ['default', '4PC']\n",
    "\n",
    "pheno2h2_list = defaultdict(list)\n",
    "pheno2se_list = defaultdict(list)\n",
    "pheno2pv_list = defaultdict(list)\n",
    "\n",
    "for pheno in phenotypes:\n",
    "    for sett in settings:\n",
    "        h2_name = '{}_{}'.format(sett, pheno)\n",
    "        h2_path = '{}{}/{}/{}/'.format(greml_path, pop, maf, h2_name)\n",
    "        hsq_fn = '{}{}.hsq'.format(h2_path, h2_name)\n",
    "        if not os.path.exists(hsq_fn): \n",
    "            h2 = np.nan\n",
    "            se = np.nan\n",
    "            pv = np.nan\n",
    "        else:\n",
    "            h2 = pd.read_table(hsq_fn).iloc[3, 1]\n",
    "            se = pd.read_table(hsq_fn).iloc[3, 2]\n",
    "            pv = pd.read_table(hsq_fn).iloc[8, 1]\n",
    "            \n",
    "        pheno2h2_list[pheno].append(h2)\n",
    "        pheno2se_list[pheno].append(se)\n",
    "        pheno2pv_list[pheno].append(pv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "2abc0127-da5e-4b70-a3d9-6abfc00c6788",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>setting</th>\n",
       "      <th>v_meta</th>\n",
       "      <th>v_pho</th>\n",
       "      <th>Recapping</th>\n",
       "      <th>DMR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>default</td>\n",
       "      <td>0.669</td>\n",
       "      <td>0.762</td>\n",
       "      <td>0.673</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4PC</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.703</td>\n",
       "      <td>0.629</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2_df = pd.DataFrame(pheno2h2_list)\n",
    "h2_df.insert(0, 'setting', settings)\n",
    "HTML(h2_df.round(3).to_html(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "482121b8-0952-4ef6-a4cf-9cfba4eebc01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>setting</th>\n",
       "      <th>v_meta</th>\n",
       "      <th>v_pho</th>\n",
       "      <th>Recapping</th>\n",
       "      <th>DMR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>default</td>\n",
       "      <td>0.159</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4PC</td>\n",
       "      <td>0.170</td>\n",
       "      <td>0.164</td>\n",
       "      <td>0.174</td>\n",
       "      <td>0.138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_df = pd.DataFrame(pheno2se_list)\n",
    "se_df.insert(0, 'setting', settings)\n",
    "HTML(se_df.round(3).to_html(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c064ad61-531a-4387-804c-2b4d70a5c55e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>setting</th>\n",
       "      <th>v_meta</th>\n",
       "      <th>v_pho</th>\n",
       "      <th>Recapping</th>\n",
       "      <th>DMR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>default</td>\n",
       "      <td>0.006356</td>\n",
       "      <td>0.000906</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.33356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5PC</td>\n",
       "      <td>0.005243</td>\n",
       "      <td>0.001364</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.30143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pv_df = pd.DataFrame(pheno2pv_list)\n",
    "pv_df.insert(0, 'setting', settings)\n",
    "HTML(pv_df.to_html(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "b6b18972-f528-416c-8f65-c2f3c3d9c8b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>setting</th>\n",
       "      <th>v_meta</th>\n",
       "      <th>v_pho</th>\n",
       "      <th>Recapping</th>\n",
       "      <th>DMR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>default</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>3.039900e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.35711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VAF</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>2.402300e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>top5VAF</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>2.408800e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5PC</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>7.916900e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.35697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10PC</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>1.099400e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.36479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15PC</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>1.302400e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PC5</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>2.914100e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PC145</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>1.559800e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   setting    v_meta         v_pho Recapping      DMR\n",
       "0  default  0.000009  3.039900e-06       0.0  0.35711\n",
       "1      VAF  0.000067  2.402300e-05        NA       NA\n",
       "2  top5VAF  0.000021  2.408800e-06        NA       NA\n",
       "3      5PC  0.000005  7.916900e-07       0.0  0.35697\n",
       "4     10PC  0.000007  1.099400e-06       0.0  0.36479\n",
       "5     15PC  0.000008  1.302400e-06       0.0   0.3471\n",
       "6      PC5  0.000009  2.914100e-06        NA       NA\n",
       "7    PC145  0.000008  1.559800e-06        NA       NA"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pv_df = pd.DataFrame(pheno2pv_list)\n",
    "pv_df.insert(0, 'setting', settings)\n",
    "pv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af68d1a7-dd12-40d5-b8ae-96e322399e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_df = pd.DataFrame(pheno2h2_list)\n",
    "res_df.insert(0, 'setting', settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "73273c55-e52b-41de-8a6b-c2c8a1671416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Variance</th>\n",
       "      <th>SE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V(G)</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>V(e)</td>\n",
       "      <td>0.036814</td>\n",
       "      <td>0.002078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vp</td>\n",
       "      <td>0.037358</td>\n",
       "      <td>0.001625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V(G)/Vp</td>\n",
       "      <td>0.014569</td>\n",
       "      <td>0.048732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>logL</td>\n",
       "      <td>1159.177000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>logL0</td>\n",
       "      <td>1159.109000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LRT</td>\n",
       "      <td>0.134000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>df</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pval</td>\n",
       "      <td>0.357110</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>n</td>\n",
       "      <td>1280.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Source     Variance        SE\n",
       "0     V(G)     0.000544  0.001827\n",
       "1     V(e)     0.036814  0.002078\n",
       "2       Vp     0.037358  0.001625\n",
       "3  V(G)/Vp     0.014569  0.048732\n",
       "4     logL  1159.177000       NaN\n",
       "5    logL0  1159.109000       NaN\n",
       "6      LRT     0.134000       NaN\n",
       "7       df     1.000000       NaN\n",
       "8     Pval     0.357110       NaN\n",
       "9        n  1280.000000       NaN"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_table(hsq_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "21b4cd6c-9d52-419c-938a-081812325fb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PC1', 'PC2', 'PC3', 'PC4', 'PC5']"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['PC{}'.format(x) for x in range(1, 6)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "2d8c55f0-98e4-4856-89f1-427edd35eeff",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype = 'log_v_meta'\n",
    "fixed_covar = ['Apiary', 'Year']\n",
    "fixed_qcovar = ['Day of Year']\n",
    "tax_covar = []\n",
    "tax_qcovar = []\n",
    "tax_covar = [\n",
    "    'Oscillospiraceae',\n",
    "    'Corynebacteriaceae',\n",
    "    'Cannabaceae',\n",
    "    'Aeromonadaceae',\n",
    "    'Weeksellaceae',\n",
    "    'Araliaceae',\n",
    "    'Actinomycetaceae',\n",
    "    'Trypanosomatidae',\n",
    "    'Saccharomycodaceae', # v_pho specific\n",
    "]\n",
    "tax_qcovar = [\n",
    "    'Bifidobacteriaceae',\n",
    "    'Lactobacillaceae',\n",
    "    'Pseudomonadaceae',   \n",
    "    'Debaryomycetaceae',\n",
    "    'Bartonellaceae',  \n",
    "    'Morganellaceae',\n",
    "    'Rosaceae', # v_pho specific\n",
    "    'Microbacteriaceae',  # v_pho specific\n",
    "    'Enterobacteriaceae',  # v_pho specific\n",
    "    'Nosematidae' # v_pho specific\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "7af35feb-57ea-46a7-b781-78497c460010",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(h2_path):\n",
    "    os.makedirs(h2_path)\n",
    "\n",
    "write_pheno_file(gcta_df, h2_name, phenotype)\n",
    "write_covar_file(gcta_df, h2_name, phenotype, fixed_covar, tax_covar)\n",
    "write_qcovar_file(gcta_df, h2_name, phenotype, fixed_qcovar, tax_qcovar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e22ddc-048d-415f-af29-2a64350bced5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "194d0778-4413-44c9-a2f6-c3fa198100c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Linux\n",
      "* Built at Jul 21 2025 17:30:18, by GCC 8.4\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 15:42:04 CET on Tue Jan 06 2026.\n",
      "Hostname: node07\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--grm /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/GRM/hom_pop\n",
      "--pheno /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.phen\n",
      "--covar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.covar\n",
      "--qcovar /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.qcovar\n",
      "--out /cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF\n",
      "\n",
      "Note: This is a multi-thread program. You could specify the number of threads by the --thread-num option to speed up the computation if there are multiple processors in your machine.\n",
      "\n",
      "Reading IDs of the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/GRM/hom_pop.grm.id].\n",
      "1442 IDs are read from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/GRM/hom_pop.grm.id].\n",
      "Reading the GRM from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/GRM/hom_pop.grm.bin].\n",
      "GRM for 1442 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/GRM/hom_pop.grm.bin].\n",
      "Reading phenotypes from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.phen].\n",
      "Non-missing phenotypes of 1442 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.phen].\n",
      "Reading quantitative covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.qcovar].\n",
      "11 quantitative covariate(s) of 1442 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.qcovar].\n",
      "Reading discrete covariate(s) from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.covar].\n",
      "11 discrete covariate(s) of 1442 individuals are included from [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.covar].\n",
      "\n",
      "11 quantitative variable(s) included as covariate(s).\n",
      "11 discrete variable(s) included as covariate(s).\n",
      "1442 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "1442 observations, 171 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.11729  1.0341\n",
      "logL: -1100.61\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G)\tV(e)\t\n",
      "1\t-1057.62\t0.76511\t1.05891\t\n",
      "2\t-1038.89\t0.62594\t1.04975\t\n",
      "3\t-1032.44\t0.55394\t1.03941\t\n",
      "4\t-1029.85\t0.51186\t1.03160\t\n",
      "5\t-1028.74\t0.48554\t1.02623\t\n",
      "6\t-1028.25\t0.46839\t1.02267\t\n",
      "7\t-1028.03\t0.43207\t1.01538\t\n",
      "8\t-1027.84\t0.43123\t1.01677\t\n",
      "9\t-1027.84\t0.43112\t1.01685\t\n",
      "10\t-1027.84\t0.43111\t1.01686\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Calculating the logLikelihood for the reduced model ...\n",
      "(variance component 1 is dropped from the model)\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 1.49274\n",
      "logL: -1149.54615\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(e)\t\n",
      "1\t-1038.62\t1.43974\t\n",
      "2\t-1036.68\t1.40659\t\n",
      "3\t-1035.83\t1.38518\t\n",
      "4\t-1035.46\t1.34056\t\n",
      "5\t-1035.14\t1.34195\t\n",
      "6\t-1035.14\t1.34196\t\n",
      "7\t-1035.14\t1.34196\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G)\t0.431106\t0.158422\n",
      "V(e)\t1.016856\t0.116236\n",
      "Vp\t1.447962\t0.075165\n",
      "V(G)/Vp\t0.297733\t0.098801\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "2.509737e-02\t-1.647915e-02\t\n",
      "-1.647915e-02\t1.351071e-02\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/cluster/raid/home/f80878961/beestrong/data/heritability_greml/hom_pop/VAF/VAF.hsq].\n",
      "\n",
      "Analysis finished at 15:42:07 CET on Tue Jan 06 2026\n",
      "Overall computational time: 2.97 sec.\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1455e53-1623-4008-9bd1-a2a97104971d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58c64f08-4d82-4061-a0a2-10a049ef2a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gg in set(pheno_df['Genetic group']):\n",
    "    command = ['sbatch', script_fn, dir_in, 'depth_{}.txt'.format(gg), 'count_ref_{}.txt'.format(gg), 'queen_{}'.format(gg)]\n",
    "    result = subprocess.run(command, cwd=dir_in, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35dcbf3e-c22b-4405-af6e-017ecf7fc49e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "cc74acf6-5acb-4998-bd28-d09255716635",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['Colony', 'log_v_pho', 'log_v_meta', 'Apiary', 'date', 'Year', 'Day of Year']\n",
    "\n",
    "# Varroa-Associated families\n",
    "vaf_abund = [\n",
    "    'Bifidobacteriaceae',\n",
    "    'Lactobacillaceae',\n",
    "    'Pseudomonadaceae',   \n",
    "    'Debaryomycetaceae',\n",
    "    'Bartonellaceae',  \n",
    "    'Morganellaceae'\n",
    "]\n",
    "vaf_prev = [\n",
    "    'Oscillospiraceae',\n",
    "    'Corynebacteriaceae',\n",
    "    'Cannabaceae',\n",
    "    'Aeromonadaceae',\n",
    "    'Weeksellaceae',\n",
    "    'Araliaceae',\n",
    "    'Actinomycetaceae',\n",
    "    'Trypanosomatidae'\n",
    "]\n",
    "\n",
    "fn = '{}tlf/Core_Family_LRA.csv'.format(working_path)\n",
    "abund_df = pd.read_csv(fn) #, index_col='Colony').reset_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "ce3b5f9d-b606-47c4-8ba9-c6932fb7f95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gcta_pheno_df = pd.DataFrame(\n",
    "    {\n",
    "        '0' : [0] * len(bs_ids),\n",
    "        # '0': bs_ids,\n",
    "        'Colony': bs_ids\n",
    "    }\n",
    ")\n",
    "gcta_pheno_df = gcta_pheno_df.merge(pheno_df[columns], on='Colony', how='left')\n",
    "gcta_pheno_df = gcta_pheno_df.merge(abund_df[['Colony'] + vaf_abund + vaf_prev], on='Colony', how='left')\n",
    "\n",
    "# replace abundance values by presence-absence for prevalence associations\n",
    "lod = np.log(5e-06)\n",
    "gcta_pheno_df[gcta_pheno_df[vaf_prev] >= lod] = int(1)\n",
    "gcta_pheno_df[gcta_pheno_df[vaf_prev] < lod] = int(0)\n",
    "\n",
    "gcta_pheno_df = gcta_pheno_df.fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "f198551c-3e89-4031-8732-1b9934f15edd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>Colony</th>\n",
       "      <th>log_v_pho</th>\n",
       "      <th>log_v_meta</th>\n",
       "      <th>Apiary</th>\n",
       "      <th>date</th>\n",
       "      <th>Year</th>\n",
       "      <th>Day of Year</th>\n",
       "      <th>Bifidobacteriaceae</th>\n",
       "      <th>Lactobacillaceae</th>\n",
       "      <th>...</th>\n",
       "      <th>Bartonellaceae</th>\n",
       "      <th>Morganellaceae</th>\n",
       "      <th>Oscillospiraceae</th>\n",
       "      <th>Corynebacteriaceae</th>\n",
       "      <th>Cannabaceae</th>\n",
       "      <th>Aeromonadaceae</th>\n",
       "      <th>Weeksellaceae</th>\n",
       "      <th>Araliaceae</th>\n",
       "      <th>Actinomycetaceae</th>\n",
       "      <th>Trypanosomatidae</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>88.0</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>188</td>\n",
       "      <td>-6.146180</td>\n",
       "      <td>-4.997377</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.744082</td>\n",
       "      <td>-9.018644</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>-1.312186</td>\n",
       "      <td>-12.003170</td>\n",
       "      <td>88.0</td>\n",
       "      <td>06/07/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>188</td>\n",
       "      <td>-6.114353</td>\n",
       "      <td>-4.710165</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.098915</td>\n",
       "      <td>-10.080948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>1.828127</td>\n",
       "      <td>-9.786422</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>244</td>\n",
       "      <td>-5.950136</td>\n",
       "      <td>-4.598848</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.069882</td>\n",
       "      <td>-7.210088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>-0.619039</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>244</td>\n",
       "      <td>-5.707229</td>\n",
       "      <td>-4.439819</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.104517</td>\n",
       "      <td>-10.880938</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>0.741937</td>\n",
       "      <td>-10.537386</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31/08/2016</td>\n",
       "      <td>2016</td>\n",
       "      <td>244</td>\n",
       "      <td>-5.405451</td>\n",
       "      <td>-4.150527</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.050514</td>\n",
       "      <td>-9.954003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>-1.221672</td>\n",
       "      <td>-11.682297</td>\n",
       "      <td>128.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.955760</td>\n",
       "      <td>-4.546061</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.303314</td>\n",
       "      <td>-5.188458</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-14.661200</td>\n",
       "      <td>128.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.556695</td>\n",
       "      <td>-3.947906</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.801414</td>\n",
       "      <td>-8.585859</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-15.168735</td>\n",
       "      <td>97.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.081466</td>\n",
       "      <td>-4.056083</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.312203</td>\n",
       "      <td>-9.878038</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>-1.987004</td>\n",
       "      <td>-10.623075</td>\n",
       "      <td>97.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-5.038656</td>\n",
       "      <td>-4.301295</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.182466</td>\n",
       "      <td>-4.987978</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>0.223144</td>\n",
       "      <td>-11.524928</td>\n",
       "      <td>97.0</td>\n",
       "      <td>01/08/2017</td>\n",
       "      <td>2017</td>\n",
       "      <td>213</td>\n",
       "      <td>-6.144559</td>\n",
       "      <td>-4.356450</td>\n",
       "      <td>...</td>\n",
       "      <td>-8.342813</td>\n",
       "      <td>-5.329028</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0     Colony log_v_pho  log_v_meta  Apiary        date  Year  \\\n",
       "0     0  BS16_0060 -1.987004  -15.168735    88.0  06/07/2016  2016   \n",
       "1     0  BS16_0061 -1.312186  -12.003170    88.0  06/07/2016  2016   \n",
       "2     0  BS16_0162  1.828127   -9.786422    88.0  31/08/2016  2016   \n",
       "3     0  BS16_0163 -0.619039  -15.168735    88.0  31/08/2016  2016   \n",
       "4     0  BS16_0164  0.741937  -10.537386    88.0  31/08/2016  2016   \n",
       "...  ..        ...       ...         ...     ...         ...   ...   \n",
       "1437  0  BS17_0506 -1.221672  -11.682297   128.0  01/08/2017  2017   \n",
       "1438  0  BS17_0511 -1.987004  -14.661200   128.0  01/08/2017  2017   \n",
       "1439  0  BS17_0512 -1.987004  -15.168735    97.0  01/08/2017  2017   \n",
       "1440  0  BS17_0513 -1.987004  -10.623075    97.0  01/08/2017  2017   \n",
       "1441  0  BS17_0517  0.223144  -11.524928    97.0  01/08/2017  2017   \n",
       "\n",
       "      Day of Year  Bifidobacteriaceae  Lactobacillaceae  ...  Bartonellaceae  \\\n",
       "0             188           -6.146180         -4.997377  ...       -6.744082   \n",
       "1             188           -6.114353         -4.710165  ...       -6.098915   \n",
       "2             244           -5.950136         -4.598848  ...       -7.069882   \n",
       "3             244           -5.707229         -4.439819  ...       -6.104517   \n",
       "4             244           -5.405451         -4.150527  ...       -7.050514   \n",
       "...           ...                 ...               ...  ...             ...   \n",
       "1437          213           -5.955760         -4.546061  ...       -7.303314   \n",
       "1438          213           -5.556695         -3.947906  ...       -5.801414   \n",
       "1439          213           -5.081466         -4.056083  ...       -9.312203   \n",
       "1440          213           -5.038656         -4.301295  ...       -7.182466   \n",
       "1441          213           -6.144559         -4.356450  ...       -8.342813   \n",
       "\n",
       "      Morganellaceae  Oscillospiraceae  Corynebacteriaceae  Cannabaceae  \\\n",
       "0          -9.018644               0.0                 0.0          0.0   \n",
       "1         -10.080948               0.0                 0.0          0.0   \n",
       "2          -7.210088               0.0                 0.0          1.0   \n",
       "3         -10.880938               0.0                 1.0          0.0   \n",
       "4          -9.954003               0.0                 0.0          1.0   \n",
       "...              ...               ...                 ...          ...   \n",
       "1437       -5.188458               0.0                 0.0          0.0   \n",
       "1438       -8.585859               0.0                 0.0          0.0   \n",
       "1439       -9.878038               0.0                 1.0          1.0   \n",
       "1440       -4.987978               0.0                 1.0          0.0   \n",
       "1441       -5.329028               0.0                 0.0          1.0   \n",
       "\n",
       "      Aeromonadaceae  Weeksellaceae  Araliaceae  Actinomycetaceae  \\\n",
       "0                1.0            1.0         1.0               0.0   \n",
       "1                1.0            1.0         1.0               0.0   \n",
       "2                0.0            1.0         1.0               0.0   \n",
       "3                1.0            1.0         1.0               0.0   \n",
       "4                0.0            1.0         0.0               0.0   \n",
       "...              ...            ...         ...               ...   \n",
       "1437             0.0            1.0         0.0               1.0   \n",
       "1438             0.0            1.0         0.0               1.0   \n",
       "1439             0.0            1.0         0.0               0.0   \n",
       "1440             1.0            1.0         0.0               1.0   \n",
       "1441             0.0            1.0         0.0               0.0   \n",
       "\n",
       "      Trypanosomatidae  \n",
       "0                  1.0  \n",
       "1                  1.0  \n",
       "2                  1.0  \n",
       "3                  0.0  \n",
       "4                  0.0  \n",
       "...                ...  \n",
       "1437               0.0  \n",
       "1438               1.0  \n",
       "1439               0.0  \n",
       "1440               0.0  \n",
       "1441               0.0  \n",
       "\n",
       "[1442 rows x 22 columns]"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gcta_pheno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "4f9e8843-f508-4f65-a3c8-38f7440251f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>-15.168735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>-12.003170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>-9.786422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>-15.168735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>-10.537386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>-11.682297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>-14.661200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>-15.168735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>-10.623075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>-11.524928</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2\n",
       "0     0  BS16_0060 -15.168735\n",
       "1     0  BS16_0061 -12.003170\n",
       "2     0  BS16_0162  -9.786422\n",
       "3     0  BS16_0163 -15.168735\n",
       "4     0  BS16_0164 -10.537386\n",
       "...  ..        ...        ...\n",
       "1437  0  BS17_0506 -11.682297\n",
       "1438  0  BS17_0511 -14.661200\n",
       "1439  0  BS17_0512 -15.168735\n",
       "1440  0  BS17_0513 -10.623075\n",
       "1441  0  BS17_0517 -11.524928\n",
       "\n",
       "[1442 rows x 3 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# phenotypes\n",
    "fn = '{}{}.v_pho'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'log_v_pho']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "\n",
    "fn = '{}{}.v_meta'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'log_v_meta']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "\n",
    "pd.read_table(fn, header=None) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "8f8cdf31-0bc3-4151-bedd-53067916c2d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1      2\n",
       "0     0  BS16_0060   88.0\n",
       "1     0  BS16_0061   88.0\n",
       "2     0  BS16_0162   88.0\n",
       "3     0  BS16_0163   88.0\n",
       "4     0  BS16_0164   88.0\n",
       "...  ..        ...    ...\n",
       "1437  0  BS17_0506  128.0\n",
       "1438  0  BS17_0511  128.0\n",
       "1439  0  BS17_0512   97.0\n",
       "1440  0  BS17_0513   97.0\n",
       "1441  0  BS17_0517   97.0\n",
       "\n",
       "[1442 rows x 3 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# categorical covariable\n",
    "fn = '{}{}.covar'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'Apiary']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony', 'Apiary', 'Year'] + vaf_prev].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony'] + vaf_prev].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony', 'Apiary']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "pd.read_table(fn, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "d7e9c698-ca53-4bf0-9ed7-b5aabc61181e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0060</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0061</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0162</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0163</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>BS16_0164</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0506</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0511</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0512</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0513</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441</th>\n",
       "      <td>0</td>\n",
       "      <td>BS17_0517</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1442 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1    2\n",
       "0     0  BS16_0060  188\n",
       "1     0  BS16_0061  188\n",
       "2     0  BS16_0162  244\n",
       "3     0  BS16_0163  244\n",
       "4     0  BS16_0164  244\n",
       "...  ..        ...  ...\n",
       "1437  0  BS17_0506  213\n",
       "1438  0  BS17_0511  213\n",
       "1439  0  BS17_0512  213\n",
       "1440  0  BS17_0513  213\n",
       "1441  0  BS17_0517  213\n",
       "\n",
       "[1442 rows x 3 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# continuous covariable\n",
    "fn = '{}{}.qcovar'.format(out_path, prefix)\n",
    "gcta_pheno_df[['0', 'Colony', 'Day of Year']].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "# gcta_pheno_df[['0', 'Colony', 'Day of Year'] + vaf_abund].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "#gcta_pheno_df[['0', 'Colony'] + vaf_abund].to_csv(fn, header=None, sep='\\t', index=False)\n",
    "pd.read_table(fn, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8404592f-a58f-49ab-bb3a-2f9f22b8fc7d",
   "metadata": {},
   "source": [
    "# Old"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2608e02b-38da-4d45-a6bc-83adfb21a8e2",
   "metadata": {},
   "source": [
    "### .bim_like"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e99323-d78a-4133-b5fe-059e78b7a1e1",
   "metadata": {},
   "source": [
    "I need to create bim-like files but be careful they are overwritten with make-bed command\n",
    "\n",
    "1. Chromosome code\n",
    "2. Variant ID\n",
    "3. Position in centimorgans (safe to use dummy value of '0')\n",
    "4. Base-pair coordinate (1-based; limited to 231-2)\n",
    "5. ALT ('A1' in PLINK 1.x) allele code\n",
    "6. REF ('A2' in PLINK 1.x) allele code\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "e78177aa-b0f1-4f67-942b-a5d4c9ab5d1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "81.028587136\n",
      "CPU times: user 0 ns, sys: 1min 33s, total: 1min 33s\n",
      "Wall time: 1min 52s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "with open('{}count_mat.pkl'.format(allele_count_path), 'rb') as inf:\n",
    "    count_mat = pkl.load(inf)\n",
    "print(count_mat.nbytes/1000000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "ab217aa9-7920-483c-a524-5a30cdf3091c",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_snp_nr = 6764372\n",
    "snp_nr = 7023976\n",
    "assert snp_nr == count_mat.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "6ca86c4d-f3aa-4dfd-ac6d-e9f957856237",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM POS REF ALT BS17-0071\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7023976it [00:05, 1403874.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.31 s, sys: 685 ms, total: 5 s\n",
      "Wall time: 5.08 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# load original SNP ids\n",
    "chr_ids = [''] * snp_nr\n",
    "snp_pos = [''] * snp_nr\n",
    "with open('{}data/gwas/BS17_0071.count'.format(working_path), 'r') as inf:\n",
    "    print(inf.readline())\n",
    "    i = 0\n",
    "    for l in tqdm(inf):\n",
    "        ch = l.split()[0]\n",
    "        pos = l.split()[1]\n",
    "        chr_ids[i] = ch\n",
    "        snp_pos[i] = pos\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "616695fb-3100-4116-8871-d96002a6bd5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6764372it [00:04, 1543827.37it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.7 s, sys: 676 ms, total: 4.37 s\n",
      "Wall time: 4.45 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# load filtered SNPs\n",
    "f_chr_ids = [''] * f_snp_nr\n",
    "f_snp_pos = [''] * f_snp_nr\n",
    "with open('{}filtered_snps.txt'.format(allele_count_path), 'r') as inf:\n",
    "    inf.readline()\n",
    "    i = 0\n",
    "    for l in tqdm(inf):\n",
    "        ch = l.split()[0]\n",
    "        pos = l.split()[1]\n",
    "        f_chr_ids[i] = ch\n",
    "        f_snp_pos[i] = pos\n",
    "        i += 1\n",
    "assert f_snp_nr == i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "49518cfa-38ec-48e7-845d-8c82f7939c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_path = '{}data/queen_genotypes/hom_pop/'.format(working_path)\n",
    "bim = pd.read_csv('{}{}.bim'.format(plink_path, 'queen'), sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "a0c2aa05-01df-433d-9807-37607ae8b59c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# good\n",
    "bim[3].to_list() == [int(x) for x in f_snp_pos]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "c96187ca-d725-44bd-a177-f3ae3f0b014b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get back mask by comparing original snps and filtered snps. TO DO: if rerun filtering, include this part directly\n",
    "snp_ids = ['{}_{}'.format(x, snp_pos[i]) for i, x in enumerate(chr_ids)]\n",
    "f_snp_ids = ['{}_{}'.format(x, f_snp_pos[i]) for i, x in enumerate(f_chr_ids)]\n",
    "\n",
    "assert len(snp_ids) == len(set(snp_ids))\n",
    "assert len(f_snp_ids) == len(set(f_snp_ids))\n",
    "\n",
    "snp_mask = ~np.isin(np.array(snp_ids), np.array(f_snp_ids))\n",
    "\n",
    "assert np.sum(~snp_mask) == f_snp_nr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "5c85e1db-7dcf-4236-979a-e4b88319de33",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6764372/6764372 [03:30<00:00, 32146.91it/s]\n"
     ]
    }
   ],
   "source": [
    "# find reference and alternative allele per SNP (REF: the one with the most counts. not sure if correct but probably REF and ALT are interchangeable) \n",
    "jj2allele = np.array(['A', 'T', 'C', 'G'])\n",
    "ref_alt = np.full((snp_nr, 2), '.')\n",
    "for i in tqdm(np.argwhere(~snp_mask).flatten()):\n",
    "    ref_alt[i] = jj2allele[np.argsort(np.sum(count_mat[i, :, :], axis=0))[::-1][:2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "d0630f4a-293d-40ca-b525-4eb52f7355ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_like = pd.DataFrame({\n",
    "    'CHR': f_chr_ids,\n",
    "    'SNP_ID': f_snp_ids,\n",
    "    'GEN_DIST': ['0'] * f_snp_nr,\n",
    "    'BP_POS': f_snp_pos,\n",
    "    'ALT': ref_alt[~snp_mask][:, 1],\n",
    "    'REF': ref_alt[~snp_mask][:, 0]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "22e8c407-a0ec-483b-8cb4-9d21008fe934",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3354437/2140746017.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  bim_like['CHR'] = bim_like['CHR'].replace(chr2int)\n"
     ]
    }
   ],
   "source": [
    "# recode snp id\n",
    "chr2int = dict(zip(list(dict.fromkeys(bim_like['CHR'])), range(1, 1 + len(set(bim_like['CHR'])))))\n",
    "bim_like['CHR'] = bim_like['CHR'].replace(chr2int)\n",
    "\n",
    "def recode_SNP_ids(old_id):\n",
    "    chrom, pos = old_id.split('.1_')\n",
    "    return '{}:{}'.format(chr2int['{}.1'.format(chrom)], pos)\n",
    "\n",
    "bim_like['SNP_ID'] = bim_like['SNP_ID'].map(recode_SNP_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "c260a07c-ffec-4348-8417-06c6b55296c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GEN_DIST</th>\n",
       "      <th>BP_POS</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5671</td>\n",
       "      <td>0</td>\n",
       "      <td>5671</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:5698</td>\n",
       "      <td>0</td>\n",
       "      <td>5698</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6621</td>\n",
       "      <td>0</td>\n",
       "      <td>6621</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6717</td>\n",
       "      <td>0</td>\n",
       "      <td>6717</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6823</td>\n",
       "      <td>0</td>\n",
       "      <td>6823</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764367</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225908</td>\n",
       "      <td>0</td>\n",
       "      <td>7225908</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764368</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7225911</td>\n",
       "      <td>0</td>\n",
       "      <td>7225911</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764369</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226186</td>\n",
       "      <td>0</td>\n",
       "      <td>7226186</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764370</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7226346</td>\n",
       "      <td>0</td>\n",
       "      <td>7226346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6764371</th>\n",
       "      <td>16</td>\n",
       "      <td>16:7227034</td>\n",
       "      <td>0</td>\n",
       "      <td>7227034</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6764372 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR      SNP_ID GEN_DIST   BP_POS ALT REF\n",
       "0          1      1:5671        0     5671   G   T\n",
       "1          1      1:5698        0     5698   C   T\n",
       "2          1      1:6621        0     6621   G   A\n",
       "3          1      1:6717        0     6717   T   C\n",
       "4          1      1:6823        0     6823   C   G\n",
       "...      ...         ...      ...      ...  ..  ..\n",
       "6764367   16  16:7225908        0  7225908   C   T\n",
       "6764368   16  16:7225911        0  7225911   T   C\n",
       "6764369   16  16:7226186        0  7226186   A   G\n",
       "6764370   16  16:7226346        0  7226346   A   G\n",
       "6764371   16  16:7227034        0  7227034   A   G\n",
       "\n",
       "[6764372 rows x 6 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "8853381a-a668-4b1b-a21b-55aa4ee0c057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# good\n",
    "bim[1].to_list() == bim_like['SNP_ID'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "f398c788-f3cf-4d90-bd85-78ca6fcf559d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.47 s, sys: 509 ms, total: 6.98 s\n",
      "Wall time: 7.11 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "bim_like.to_csv('{}queen.bim_like'.format(hom_pop_path), header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "0b0b63a9-5207-4e62-be80-91410803f736",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cluster/raid/home/f80878961/beestrong/data/queen_genotypes/hom_pop/'"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hom_pop_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc9171c7-aeac-45e2-a254-4f4fbfb174b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
